%% Easel miniapps manpage chapter automatically generated. Do not edit.
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-afetch}}{esl-afetch} - retrieve alignments from a multi-MSA database  }
\subsection*{Synopsis}



\noindent
\monob{esl-afetch} [\monoi{options}] \monoi{msafile key}

  (single MSA retrieval)

\noindent
\monob{esl-afetch -f} [\monoi{options}] \monoi{msafile keyfile}

  (multiple MSA retrieval, from a file of keys)

\noindent
\monob{esl-afetch {-}{-}index }\monoi{msafile}

  (index an MSA file for retrieval)


\subsection*{Description}
 \mono{esl-afetch} retrieves the alignment named \monoi{key} from an alignment
database in file \monoi{msafile.} The \monoi{msafile} is a "multiple multiple alignment"
file in Stockholm (e.g. native Pfam or Rfam) format. The  \monoi{key} is either the
name (ID) of the alignment, or its accession number (AC).  

Alternatively,
 \mono{esl-afetch -f} provides the ability to fetch many alignments at once. The
 \mono{-f } option has it interpret the second argument as a \monoi{keyfile}, a file consisting
of one name or accession per line.  

The \monoi{msafile} should first be SSI indexed
with \mono{esl-afetch {-}{-}index} for efficient retrieval. An SSI index is not required,
but without one alignment retrieval may be painfully slow.  
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Print
brief help; includes version number and summary of all options, including
expert options.  
\item [\monob{-f} ] Interpret the second argument as a  \monoi{keyfile} instead of
as just one \monoi{key}.\monoi{} The \monoi{keyfile} contains one name or accession per line. This
option doesn't work with the \mono{{-}{-}index} option.   
\item [\monob{-o}\monoi{ $<$f$>$} ] Output retrieved alignments
to a file  \monoi{$<$f$>$} instead of to stdout.  
\item [\monob{-O} ] Output retrieved alignment to a file
named \monoi{key}. This is a convenience for saving some typing: instead of  

 \user{\% esl-afetch -o RRM\_1 msafile RRM\_1}

you can just type 

 \user{\% esl-afetch -O msafile RRM\_1}

The \mono{-O } option only works if you're retrieving a single alignment; it is
incompatible with  \mono{-f.}  
\item [\monob{{-}{-}index} ] Instead of retrieving a \monoi{key,} the special command
\mono{esl-afetch {-}{-}index} \monoi{msafile} produces an SSI index of the names and accessions
of the alignments in the  \monoi{msafile.} Indexing should be done once on the \monoi{msafile}
to prepare it for all future fetches.   
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-alimanip}}{esl-alimanip} - manipulate a multiple sequence alignment  }
\subsection*{Synopsis}
\noindent
 \monob{esl-alimanip}
[\monoi{options}] \monoi{msafile}  
\subsection*{Description}
 \mono{esl-alimanip} can manipulate the multiple
sequence alignment(s) in  \monoi{msafile} in various ways. Options exist to remove
specific sequences, reorder sequences, designate reference columns using
Stockholm "\#=GC RF" markup, and add annotation that numbers columns.   

The
alignments can be of protein or DNA/RNA sequences. All alignments in the
same  \monoi{msafile} must be either protein or DNA/RNA. The alphabet will be autodetected
unless one of the options  \mono{{-}{-}amino,} \mono{{-}{-}dna,} or  \mono{{-}{-}rna } are given.     
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h
} ] Print brief help;  includes version number and summary of all options,
including expert options.  
\item [\monob{-o}\monoi{ $<$f$>$} ] Save the resulting, modified alignment in
Stockholm format to a file \monoi{$<$f$>$.} The default is to write it to standard output.
 
\item [\monob{{-}{-}informat}\monoi{ $<$s$>$} ] Assert that  \monoi{msafile} is in alignment format \monoi{$<$s$>$}, bypassing format
autodetection. Common choices for  \monoi{$<$s$>$}  include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast},
\mono{clustal}, \mono{phylip}. For more information, and for codes for some less common
formats, see main documentation. The string \monoi{$<$s$>$} is case-insensitive (\mono{a2m} or
\mono{A2M} both work).   
\item [\monob{{-}{-}outformat}\monoi{ $<$s$>$} ] Write the output in alignment format \monoi{$<$s$>$}. Common
choices for  \monoi{$<$s$>$}  include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal}, \mono{phylip}.
The string \monoi{$<$s$>$} is case-insensitive (\mono{a2m} or \mono{A2M} both work). Default is \mono{stockholm}.
 
\item [\monob{{-}{-}devhelp} ] Print help, as with   \mono{-h,} but also include undocumented developer
options. These options are not listed below, are under development or experimental,
and are not guaranteed to even work correctly. Use developer options at
your own risk. The only resources for understanding what they actually do
are the brief one-line description printed when \mono{{-}{-}devhelp} is enabled, and
the source code.  
\end{wideitem}

\subsection*{Expert Options}
 \begin{wideitem}
\item [\monob{{-}{-}lnfract}\monoi{ $<$x$>$} ] Remove any sequences with length
less than  \monoi{$<$x$>$} fraction the length of the median length sequence in the alignment.
 
\item [\monob{{-}{-}lxfract}\monoi{ $<$x$>$} ] Remove any sequences with length more than  \monoi{$<$x$>$} fraction the length
of the median length sequence in the alignment.  
\item [\monob{{-}{-}lmin}\monoi{ $<$n$>$} ] Remove any sequences
with length less than  \monoi{$<$n$>$} residues.  
\item [\monob{{-}{-}lmax}\monoi{ $<$n$>$} ] Remove any sequences with length
more than  \monoi{$<$n$>$} residues.  
\item [\monob{{-}{-}rfnfract}\monoi{ $<$x$>$} ] Remove any sequences with nongap RF length
less than  \monoi{$<$x$>$} fraction the nongap RF length of the alignment.  
\item [\monob{{-}{-}detrunc}\monoi{ $<$n$>$}
] Remove any sequences that have all gaps in the first  \monoi{$<$n$>$} non-gap \#=GC RF
columns or the last  \monoi{$<$n$>$} non-gap \#=GC RF columns.  
\item [\monob{{-}{-}xambig}\monoi{ $<$n$>$} ] Remove any sequences
that has more than \monoi{$<$n$>$} ambiguous (degenerate) residues.  
\item [\monob{{-}{-}seq-r}\monoi{ $<$f$>$} ] Remove any
sequences with names listed in file  \monoi{$<$f$>$.} Sequence names listed in  \monoi{$<$f$>$} can
be separated by tabs, new lines, or spaces. The file must be in Stockholm
format for this option to work.   
\item [\monob{{-}{-}seq-k}\monoi{ $<$f$>$} ] Keep only sequences with names
listed in file  \monoi{$<$f$>$.} Sequence names listed in  \monoi{$<$f$>$} can be separated by tabs,
new lines, or spaces. By default, the kept sequences will remain in the
original order they appeared in  \monoi{msafile,} but the order from  \monoi{$<$f$>$}  will be
used if the  \mono{{-}{-}k-reorder} option is enabled. The file must be in Stockholm format
for this option to work.   
\item [\monob{{-}{-}small} ] With \mono{{-}{-}seq-k } or \mono{{-}{-}seq-r,} operate in small memory
mode.  The alignment(s) will not be stored in memory, thus \mono{{-}{-}seq-k } and \mono{{-}{-}seq-r}
will be able to work on very large alignments regardless of the amount
of available RAM. The alignment file must be in Pfam format and  \mono{{-}{-}informat
pfam} and one of \mono{{-}{-}amino,} \mono{{-}{-}dna,} or \mono{{-}{-}rna} must be given as well.  
\item [\monob{{-}{-}k-reorder} ] With
\mono{{-}{-}seq-k}\monoi{ $<$f$>$}\mono{,} reorder the kept sequences in the output alignment to the order
from the list file \monoi{$<$f$>$.}  
\item [\monob{{-}{-}seq-ins}\monoi{ $<$n$>$} ] Keep only sequences that have at least 1
inserted residue after  nongap RF position  \monoi{$<$n$>$.}  
\item [\monob{{-}{-}seq-ni}\monoi{ $<$n$>$} ] With  \mono{{-}{-}seq-ins} require
at least  \monoi{$<$n$>$}  inserted residues in a sequence for it to be kept.  
\item [\monob{{-}{-}seq-xi}\monoi{ $<$n$>$}
] With  \mono{{-}{-}seq-ins} allow at most \monoi{$<$n$>$}  inserted residues in a sequence for it to
be kept.  
\item [\monob{{-}{-}trim}\monoi{ $<$f$>$} ] File  \monoi{$<$f$>$} is an unaligned FASTA file containing truncated
versions of each sequence in the  \monoi{msafile.}  Trim the sequences in the alignment
to match their truncated versions in  \monoi{$<$f$>$.} If the alignment output format
is Stockholm (the default output format), all per-column (GC) and per-residue
(GR) annotation will be removed from the alignment when \mono{{-}{-}trim} is used. However,
if  \mono{{-}{-}t-keeprf } is also used, the reference annotation (GC RF) will be kept.
 
\item [\monob{{-}{-}t-keeprf} ] Specify that the 'trimmed' alignment maintain the original reference
(GC RF) annotation. Only works in combination with  \mono{{-}{-}trim.}  
\item [\monob{{-}{-}minpp}\monoi{ $<$x$>$} ] Replace
all residues in the alignments for which the posterior probability annotation
(\#=GR PP) is less than  \monoi{$<$x$>$} with gaps. The PP annotation for these residues
is also converted to gaps.  \monoi{$<$x$>$} must be greater than 0.0 and less than or equal
to 0.95.  
\item [\monob{{-}{-}tree}\monoi{ $<$f$>$} ] Reorder sequences by tree order.  Perform single linkage
clustering on the sequences in the alignment based on sequence identity
given the alignment to define a 'tree'  of the sequences. The sequences in
the alignment are reordered according to the tree, which groups similar
sequences together. The tree is output in Newick format to  \monoi{$<$f$>$.}  
\item [\monob{{-}{-}reorder}\monoi{ $<$f$>$}
] Reorder sequences to the order listed in file  \monoi{$<$f$>$.} Each sequence in the alignment
must be listed in  \monoi{$<$f$>$.} Use \mono{{-}{-}k-reorder} to reorder only a subset of sequences
to a subset alignment file.  The file must be in Stockholm format for this
option to work.   
\item [\monob{{-}{-}mask2rf}\monoi{ $<$f$>$} ] Read in the 'mask' file  \monoi{$<$f$>$} and use it to define
new \#=GC RF annotation for the  alignment. \monoi{$<$f$>$} must be a single line, with
exactly  \monoi{$<$alen$>$}  or  \monoi{$<$rflen$>$} characters, either the full alignment length or
the number of nongap \#=GC RF characters, respectively. Each character must
be either a '1' or a '0'. The new \#=GC RF markup will contain an 'x' for each
column that is a '1' in lane mask file, and a '.' for each column that is a
'0'.  If the mask is of length \monoi{$<$rflen$>$} then it is interpreted as applying to
only nongap RF characters in the existing RF annotation, all gap RF characters
will remain gaps and nongap RF characters will be redefined as above.  
\item [\monob{{-}{-}m-keeprf}
] With  \mono{{-}{-}mask2rf,} do not overwrite existing nongap RF characters that are
included by the input mask as 'x', leave them as the character they are. 

\item [\monob{{-}{-}num-all}\monoi{} ] Add annotation to the alignment numbering all of the columns in
the alignment.   
\item [\monob{{-}{-}num-rf}\monoi{} ] Add annotation to the alignment numbering the non-gap
(non '.') \#=GC RF columns of the alignment.   
\item [\monob{{-}{-}rm-gc}\monoi{ $<$s$>$} ] Remove certain types of
\#=GC annotation from the alignment.  \monoi{$<$s$>$}  must be one of: \mono{RF}, \mono{SS\_cons}, \mono{SA\_cons},
\mono{PP\_cons}.  
\item [\monob{{-}{-}sindi}\monoi{} ] Annotate individual secondary structures for each sequence
by imposing the consensus secondary structure defined by the \#=GC SS\_cons
annotation.   
\item [\monob{{-}{-}post2pp}\monoi{} ] Update Infernal's cmalign 0.72-1.0.2 posterior probability
"POST" annotation to "PP" annotation, which is read by other miniapps,
including  \mono{esl-alimask} and  \mono{esl-alistat.}  
\item [\monob{{-}{-}amino} ] Assert that the  \monoi{msafile} 
contains protein sequences.   
\item [\monob{{-}{-}dna} ] Assert that the  \monoi{msafile}  contains DNA
sequences.   
\item [\monob{{-}{-}rna} ] Assert that the  \monoi{msafile}  contains RNA sequences.     
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-alimap}}{esl-alimap} - map two alignments to each other  }
\subsection*{Synopsis}
\noindent
\monob{esl-alimap} [\monoi{options}]
\monoi{msafile1} \monoi{msafile2}  
\subsection*{Description}
 \mono{esl-alimap} is a highly specialized application
that determines the optimal alignment mapping of columns between two alignments
of the same sequences. An alignment mapping defines for each column in alignment
1 a matching column in alignment 2. The number of residues in the aligned
sequences that are in common between the two matched columns are considered
'shared' by those two columns.  

For example, if the nth residue of sequence
i occurs in alignment 1 column x and alignment 2 column y, then only a
mapping of alignment 1 and 2 that includes column x mapping to column y
would correctly map and share the residue.   

The optimal mapping of the
two alignments is the mapping which maximizes the sum of shared residues
between all pairs of matching columns. The fraction of total residues that
are shared is reported as the coverage in the  \mono{esl-alimap} output.  

Only the
first alignments in  \monoi{msafile1}  and \monoi{msafile2} will be mapped to each other.
If the files contain more than one alignment, all alignments after the
first will be ignored.  

The two alignments (one from each file) must contain
exactly the same sequences (if they were unaligned, they'd be identical)
in precisely the same order. They must also be in Stockholm format.  

The
output of  \mono{esl-alimap} differs depending on whether one or both of the alignments
 contain reference (\#=GC RF) annotation. If so, the coverage for residues
from nongap RF positions will be reported separately from the total coverage.
 

\mono{esl-alimap} uses a dynamic programming algorithm to compute the optimal
mapping. The algorithm is similar to the Needleman-Wunsch-Sellers algorithm
but the scores used at each step of the recursion are not residue-residue
comparison scores but rather the number of residues shared between two
columns.   The \mono{{-}{-}mask-a2a}\monoi{ $<$f$>$}\mono{,} \mono{{-}{-}mask-a2rf}\monoi{ $<$f$>$}\mono{,} \mono{{-}{-}mask-rf2a}\monoi{ $<$f$>$}\mono{,} and \mono{{-}{-}mask-rf2rf}\monoi{ $<$f$>$} options
create 'mask' files that pertain to the optimal mapping in slightly different
ways. A mask file consists of a single line, of only '0' and '1' characters.
These denote which positions of the alignment from  \monoi{msafile1} map to positions
of the alignment from  \monoi{msafile2} as described below for each of the four
respective masking options. These masks can be used to extract only those
columns of the  \monoi{msafile1} alignment  that optimally map to columns of the
 \monoi{msafile2} alignment using the  \mono{esl-alimask} miniapp. To extract the corresponding
set of columns  from  \monoi{msafile2} (that optimally map to columns of the alignment
from \monoi{msafile1}), it is necessary to rerun the program with the order of
the two  msafiles reversed, save new masks, and use \mono{esl-alimask} again.  
\subsection*{Options}

\begin{wideitem}
\item [\monob{-h} ] Print brief help; includes version number and summary of all options.
 
\item [\monob{-q} ] Be quiet; don't print information the optimal mapping of each column,
only report coverage and potentially save masks to optional output files.
  
\item [\monob{{-}{-}mask-a2a}\monoi{ $<$f$>$} ] Save a mask of '0's and '1's to file \monoi{$<$f$>$.} A '1' at position x means
that position x of the alignment from \monoi{msafile1} maps to an alignment position
in the alignment from \monoi{msafile2} in the optimal map.  
\item [\monob{{-}{-}mask-a2rf}\monoi{ $<$f$>$} ] Save a mask
of '0's and '1's to file \monoi{$<$f$>$.} A '1' at position x means that position x of the alignment
from \monoi{msafile1} maps to a nongap RF position in the alignment from  \monoi{msafile2}
in the optimal map.  
\item [\monob{{-}{-}mask-rf2a}\monoi{ $<$f$>$} ] Save a mask of '0's and '1's to file \monoi{$<$f$>$.} A '1' at
position x means that nongap RF position x of the alignment from \monoi{msafile1}
maps to an alignment position in the alignment from  \monoi{msafile2} in the optimal
map.  
\item [\monob{{-}{-}mask-rf2rf}\monoi{ $<$f$>$} ] Save a mask of '0's and '1's to file \monoi{$<$f$>$.} A '1' at position x means
that nongap RF position x of the alignment from \monoi{msafile1} maps to a nongap
RF position in the alignment from  \monoi{msafile2} in the optimal map.  
\item [\monob{{-}{-}submap}\monoi{
$<$f$>$} ] Specify that all of the columns from the alignment from  \monoi{msafile1} exist
identically (contain the same residues from all sequences) in the alignment
from  \monoi{msafile2.}  This makes the task of mapping trivial. However, not all
columns of  \monoi{msafile1}  must exist in  \monoi{msafile2.} Save the mask to file \monoi{$<$f$>$.} A
'1' at position x of the mask means that position x of the alignment from
\monoi{msafile1} is the same as position y of \monoi{msafile2,} where y is the number of
'1's that occur at positions $<$= x in the mask.  
\item [\monob{{-}{-}amino} ] Assert that  \monoi{msafile1}
and  \monoi{msafile2} contain protein sequences.   
\item [\monob{{-}{-}dna} ] Assert that  \monoi{msafile1} and
 \monoi{msafile2} contain DNA sequences.   
\item [\monob{{-}{-}rna} ] Assert that the  \monoi{msafile1} and  \monoi{msafile2}
contain RNA sequences.    
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-alimask}}{esl-alimask} - remove columns from a multiple sequence alignment  }
\subsection*{Synopsis}



\noindent
\monob{esl-alimask }[\monoi{options}] \monoi{msafile maskfile}

  (remove columns based on a mask in an input file)

\noindent
\monob{esl-alimask -t }[\monoi{options}] \monoi{msafile coords}

  (remove a contiguous set of columns at the start and end of an alignment)

\noindent
\monob{esl-alimask -g }[\monoi{options}] \monoi{msafile}

  (remove columns based on their frequency of gaps)

\noindent
\monob{esl-alimask -p }[\monoi{options}] \monoi{msafile}

  (remove columns based on their posterior probability annotation)

\noindent
\monob{esl-alimask {-}{-}rf-is-mask }[\monoi{options}] \monoi{msafile}

  (only remove columns that are gaps in the RF annotation)

The \mono{-g} and \mono{-p} options may be used in combination. 

  
\subsection*{Description}
 \mono{esl-alimask} reads a single input alignment, removes some columns
from it (i.e. masks it), and outputs the masked alignment.  

\mono{esl-alimask } can
be run in several different modes.  

\mono{esl-alimask } runs in "mask file mode"
by default when two command-line arguments (\monoi{msafile} and \monoi{maskfile}) are supplied.
In this mode, a bit-vector mask in the  \monoi{maskfile} defines which columns to
keep/remove.  The mask is a string that may only contain the characters
'0' and '1'. A '0' at position x of the mask indicates that column x is excluded
by the mask and should be removed during masking.  A '1' at position x of
the mask indicates that column x is included by the mask and should not
be removed during masking.  All lines in the \monoi{maskfile} that begin with '\#'
are considered comment lines and are ignored.  All non-whitespace characters
in non-comment lines are considered to be part of the mask. The length of
the mask must equal either the total number of columns in the (first) alignment
in \monoi{msafile,} or the number of columns that are not gaps in the RF annotation
of that alignment. The latter case is only valid if \monoi{msafile} is in Stockholm
format and contains '\#=GC RF' annotation.  If the mask length is equal to
the non-gap RF length, all gap RF columns will automatically be removed.
 

\mono{esl-alimask } runs in "truncation mode" if the  \mono{-t } option is used along
with two command line arguments (\monoi{msafile} and \monoi{coords}). In this mode, the
alignment will be truncated by removing a contiguous set of columns from
the beginning and end of the alignment. The second command line argument
is the  \monoi{coords} string, that specifies what range of columns to keep in
the alignment, all columns outside of this range will be removed. The \monoi{coords}
string consists of start and end coordinates separated by any nonnumeric,
nonwhitespace character or characters you like; for example, \mono{23..100}, \mono{23/100},
or \mono{23-100} all work. To keep all alignment columns beginning at 23 until the
end of the alignment, you  can omit the end; for example, \mono{23:} would work.
If the  \mono{{-}{-}t-rf } option is used in combination with  \mono{-t,} the coordinates in
 \monoi{coords} are interpreted as non-gap RF column coordinates. For example, with
 \mono{{-}{-}t-rf, } a  \monoi{coords}  string of \mono{23-100 } would remove all columns before the
23rd non-gap residue in the "\#=GC RF" annotation and after the 100th non-gap
RF residue.  

\mono{esl-alimask } runs in "RF mask" mode if the \mono{{-}{-}rf-is-mask} option is
enabled. In this mode, the alignment must be in Stockholm format and contain
'\#=GC RF' annotation.  \mono{esl-alimask} will simply remove all columns that are
gaps in the RF annotation.  

\mono{esl-alimask} runs in "gap frequency mode" if 
\mono{-g } is enabled. In this mode columns for which greater than  \monoi{$<$f$>$} fraction of
the aligned sequences have gap residues will be removed.  By default,  \monoi{$<$f$>$}
is 0.5, but this value can be changed to  \monoi{$<$f$>$} with the  \mono{{-}{-}gapthresh}\monoi{ $<$f$>$}\mono{} option.
In this mode, if the alignment is in Stockholm format and has RF annotation,
then all columns that are gaps in the RF annotation will automatically
be removed, unless \mono{{-}{-}saveins} is enabled.  

\mono{esl-alimask} runs in "posterior probability
mode" if  \mono{-p } is enabled. In this mode,  masking is based on posterior probability
annotation, and the input alignment must be in Stockholm format and contain
'\#=GR PP' (posterior probability) annotation for all sequences. As a special
case, if  \mono{-p } is used in combination with  \mono{{-}{-}ppcons,} then the input alignment
need not have '\#=GR PP' annotation, but must contain '\#=GC PP\_cons' (posterior
probability consensus) annotation.  

Characters in Stockholm alignment posterior
probability annotation (both '\#=GR PP' and '\#=GC PP\_cons') can have 12 possible
values: the ten digits '0-9', '*', and '.'. If '.', the position must correspond to
a gap in the sequence (for '\#=GR PP') or in the RF annotation (for '\#=GC PP\_cons').
 A value of '0' indicates a posterior probability of between 0.0 and 0.05,
'1' indicates between 0.05 and 0.15, '2' indicates between 0.15 and 0.25 and so
on up to '9' which indicates between 0.85 and 0.95. A value of '*' indicates a
posterior probability of between 0.95 and 1.0. Higher posterior probabilities
correspond to greater confidence that the aligned residue belongs where
it appears in the alignment.  

When \mono{-p } is enabled with  \mono{{-}{-}ppcons}\monoi{ $<$x$>$}\mono{,} columns
which have a consensus posterior probability of less than \monoi{$<$x$>$} will be removed
during masking, and all other columns will not be removed.  

When \mono{-p } is enabled
without \mono{{-}{-}ppcons,} the number of each possible PP value in each column is
counted.  If  \monoi{$<$x$>$} fraction of the sequences that contain aligned residues
(i.e. do not contain gaps) in a column have a posterior probability  greater
than or equal to  \monoi{$<$y$>$,} then that column will not be removed during masking.
All columns that do not meet this criterion will be removed. By default,
the values of both \monoi{$<$x$>$} and  \monoi{$<$y$>$} are 0.95, but they can be changed with the 
\mono{{-}{-}pfract}\monoi{ $<$x$>$} and  \mono{{-}{-}pthresh}\monoi{ $<$y$>$}\mono{} options, respectively.  

In posterior probability
mode, all columns that have 0 residues (i.e. that are 100\% gaps) will be
automatically removed, unless the  \mono{{-}{-}pallgapok} option is enabled, in which
case such columns will not be removed.  

Importantly, during posterior probability
masking, unless \mono{{-}{-}pavg } is used, PP annotation values are always considered
to be the minimum numerical value in their corresponding range. For example,
a PP '9' character is converted to a numerical posterior probability of 0.85.
If \mono{{-}{-}pavg } is used, PP annotation values are considered to be the average
numerical value in their range. For example, a PP '9' character is converted
to a numerical posterior probability of 0.90.  

In posterior probability mode,
if the alignment is in Stockholm format and has RF annotation, then all
columns that are gaps in the RF annotation will automatically be removed,
unless \mono{{-}{-}saveins} is enabled.  

A single run of \mono{esl-alimask} can perform both
gap frequency-based masking and posterior probability-based masking if both
the  \mono{-g} and \mono{-p} options are enabled. In this case, a gap frequency-based mask
and a posterior probability-based mask are independently computed.  These
two masks are combined to create the final mask using a logical 'and' operation.
Any column that is to be removed by either the gap or PP mask will be removed
by the final mask.  

With the \mono{{-}{-}small} option,  \mono{esl-alimask} will operate in memory
saving mode and the required RAM for the masking will be minimal (usually
less than a Mb) and independent of the alignment size. To use  \mono{{-}{-}small}, the
alignment alphabet must be specified with either \mono{{-}{-}amino}, \mono{{-}{-}dna},\mono{} or  \mono{{-}{-}rna}, and
the alignment must be in Pfam format (non-interleaved, 1 line/sequence Stockholm
format). Pfam format is the default output format of INFERNAL's \mono{cmalign }
program. Without  \mono{{-}{-}small} the required RAM will be equal to roughly the size
of the first input alignment (the size of the alignment file itself if
it only contains one alignment).   
\subsection*{Output}
 By default,  \mono{esl-alimask} will print
only the masked alignment to stdout and then exit. If the \mono{-o}\monoi{ $<$f$>$} option is
used, the alignment will be saved to file  \monoi{$<$f$>$} , and information on the number
of columns kept and removed will be printed to stdout. If  \mono{-q} is used in
combination with  \mono{-o}, nothing is printed to stdout.  

The mask(s) computed
by  \mono{esl-alimask} when the  \mono{-t}, \mono{-p}, \mono{-g}, or \mono{{-}{-}rf-is-mask} options are used can be saved
to output files using the options \mono{{-}{-}fmask-rf}\monoi{ $<$f$>$}, \mono{{-}{-}fmask-all}\monoi{ $<$f$>$}, \mono{{-}{-}gmask-rf}\monoi{ $<$f$>$}, \mono{{-}{-}gmask-all}\monoi{
$<$f$>$}, \mono{{-}{-}pmask-rf}\monoi{ $<$f$>$}, and  \mono{{-}{-}pmask-all}\monoi{ $<$f$>$}. In all cases,  \monoi{$<$f$>$}  will contain a single line,
a bit vector of length \monoi{$<$n$>$,} where  \monoi{$<$n$>$}  is the either the total number of columns
in the alignment (for the options suffixed with 'all') or the number of non-gap
columns in the RF annotation (for the options suffixed with 'rf'). The mask
will be a string of '0' and '1' characters: a '0' at position x in the mask indicates
column x was removed (excluded) by the mask, and a '1' at position x indicates
column x was kept (included) by the mask. For the 'rf' suffixed options, the
mask only applies to non-gap RF columns.  The options beginning with 'f' will
save the 'final' mask used to keep/remove columns from the alignment. The
options beginning with 'g' save the masks based on gap frequency and require
\mono{-g}. The options beginning with 'p' save the masks based on posterior probabilities
and require  \mono{-p}.   
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Print brief help; includes version number and
summary of all options, including expert options.  
\item [\monob{-o}\monoi{ $<$f$>$} ] Output the final,
masked alignment to file  \monoi{$<$f$>$} instead of to stdout. When this option is used,
information about the number of columns kept/removed is printed to stdout.
 
\item [\monob{-q} ] Be quiet; do not print anything to stdout.  This option can only be used
in combination with the \mono{-o } option.  
\item [\monob{{-}{-}small} ] Operate in memory saving mode.
Required RAM will be independent of the size of the input alignment to
mask, instead of roughly the size of the input alignment. When enabled,
the alignment must be in Pfam Stockholm (non-interleaved 1 line/seq) format
(see \mono{esl-reformat}) and the output alignment will be in Pfam format.  
\item [\monob{{-}{-}informat}\monoi{
$<$s$>$} ] Assert that input \monoi{msafile} is in alignment format \monoi{$<$s$>$}. Common choices for
 \monoi{$<$s$>$}  include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal}, \mono{phylip}. For more information,
and for codes for some less common formats, see main documentation. The
string \monoi{$<$s$>$} is case-insensitive (\mono{a2m} or \mono{A2M} both work). Default is  \mono{stockholm}
format, unless \mono{{-}{-}small} is used, in which case \mono{pfam} format (non-interleaved
Stockholm) is assumed.  
\item [\monob{{-}{-}outformat}\monoi{ $<$s$>$} ] Write the output \monoi{msafile} in alignment
format \monoi{$<$s$>$}. Common choices for  \monoi{$<$s$>$}  include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast},
\mono{clustal}, \mono{phylip}. The string \monoi{$<$s$>$} is case-insensitive (\mono{a2m} or \mono{A2M} both work).
Default is \mono{stockholm}, unless \mono{{-}{-}small} is enabled, in which case \mono{pfam} (noninterleaved
Stockholm) is the default output format.   
\item [\monob{{-}{-}fmask-rf}\monoi{ $<$f$>$} ] Save the non-gap RF-length
final mask used to mask the alignment to file \monoi{$<$f$>$}. The input alignment must
be in Stockholm format and contain '\#=GC RF' annotation for this option to
be valid. See the OUTPUT section above for more details on output mask files.
 
\item [\monob{{-}{-}fmask-all}\monoi{ $<$f$>$} ] Save the full alignment-length final mask used to mask the alignment
to file \monoi{$<$f$>$}. See the OUTPUT section above for more details on output mask
files.  
\item [\monob{{-}{-}amino} ] Specify that the input alignment is a protein alignment. By
default, \mono{esl-alimask} will try to autodetect the alphabet, but if the alignment
is sufficiently small it may be ambiguous. This option defines the alphabet
as protein. Importantly, if  \mono{{-}{-}small} is enabled, the alphabet must be specified
with either \mono{{-}{-}amino}, \mono{{-}{-}dna}, or  \mono{{-}{-}rna}.  
\item [\monob{{-}{-}dna} ] Specify that the input alignment is
a DNA alignment.  
\item [\monob{{-}{-}rna} ] Specify that the input alignment is an RNA alignment.
  
\item [\monob{{-}{-}t-rf} ] With \mono{-t}, specify that the start and end coordinates defined in the
second command line argument  \monoi{coords} correspond to non-gap RF coordinates.
To use this option, the alignment must be in Stockholm format and have
"\#=GC RF" annotation. See the DESCRIPTION section for an example of using
the \mono{{-}{-}t-rf} option.  
\item [\monob{{-}{-}t-rmins} ] With \mono{-t}, specify that all columns that are gaps in
the reference (RF) annotation in between the specified start and end coordinates
be removed. By default, these columns will be kept. To use this option, the
alignment must be in  Stockholm format and have "\#=GC RF" annotation.  

\item [\monob{{-}{-}gapthresh}\monoi{ $<$x$>$} ] With \mono{-g}, specify that a column is kept (included by mask) if
no more than  \monoi{$<$f$>$} fraction of sequences in the alignment have a gap ('.', '-',
or '\_') at that position. All other columns are removed (excluded by mask).
By default,  \monoi{$<$x$>$} is 0.5.  
\item [\monob{{-}{-}gmask-rf}\monoi{ $<$f$>$} ] Save the non-gap RF-length gap frequency-based
mask used to mask the alignment to file \monoi{$<$f$>$}. The input alignment must be in
Stockholm format and contain '\#=GC RF' annotation for this option to be valid.
See the OUTPUT section above for more details on output mask files.  
\item [\monob{{-}{-}gmask-all}\monoi{
$<$f$>$} ] Save the full alignment-length gap frequency-based mask used to mask the
alignment to file \monoi{$<$f$>$}. See the OUTPUT section above for more details on output
mask files.   
\item [\monob{{-}{-}pfract}\monoi{ $<$x$>$} ] With \mono{-p}, specify that a column is kept (included by
mask) if the fraction of sequences with a non-gap residue in that column
with a  posterior probability of at least  \monoi{$<$y$>$} (from \mono{{-}{-}pthresh}\monoi{ $<$y$>$}) is \monoi{$<$x$>$} or greater.
All other columns are removed (excluded by mask). By default  \monoi{$<$x$>$}  is 0.95.
  
\item [\monob{{-}{-}pthresh}\monoi{ $<$y$>$} ] With \mono{-p}, specify that a column is kept (included by mask) if
 \monoi{$<$x$>$} (from \mono{{-}{-}pfract }\monoi{$<$x$>$}) fraction of sequences with a non-gap residue in that
column have a  posterior probability of at least  \monoi{$<$y$>$}.\monoi{} All other columns are
removed (excluded by mask). By default  \monoi{$<$y$>$}  is 0.95. See the DESCRIPTION section
for more on posterior probability (PP) masking.  Due to the granularity
of the PP annotation, different  \monoi{$<$y$>$} values within a range covered by a single
PP character will be have the same effect on masking. For example, using
 \mono{{-}{-}pthresh 0.86 } will have the same effect as using \mono{{-}{-}pthresh 0.94}.  
\item [\monob{{-}{-}pavg}\monoi{ $<$x$>$} ] With
\mono{-p}, specify that a column is kept (included by mask) if  the average posterior
probability of non-gap residues in that column is at least \monoi{$<$x$>$}. See the DESCRIPTION
section for more on posterior probability (PP) masking.   
\item [\monob{{-}{-}ppcons}\monoi{ $<$x$>$} ] With
\mono{-p}, use the '\#=GC PP\_cons' annotation to define which columns to keep/remove.
A column is kept (included by mask) if the PP\_cons value for that column
is  \monoi{$<$x$>$} or greater. Otherwise it is removed.  
\item [\monob{{-}{-}pallgapok} ] With \mono{-p}, do not automatically
remove any columns that are 100\% gaps (i.e. contain 0 aligned residues). By
default, such columns will be removed.  
\item [\monob{{-}{-}pmask-rf}\monoi{ $<$f$>$} ] Save the non-gap RF-length
posterior probability-based mask used to mask the alignment to file \monoi{$<$f$>$}. The
input alignment must be in Stockholm format and contain '\#=GC RF' annotation
for this option to be valid. See the OUTPUT section above for more details
on output mask files.  
\item [\monob{{-}{-}pmask-all}\monoi{ $<$f$>$} ] Save the full alignment-length posterior
probability-based mask used to mask the alignment to file \monoi{$<$f$>$}. See the OUTPUT
section above for more details on output mask files.   
\item [\monob{{-}{-}keepins } ] If  \mono{-p } and/or
\mono{-g} is enabled and the alignment is in Stockholm or Pfam format and has '\#=GC
RF' annotation, then allow columns that are gaps in the RF annotation to
possibly be kept. By default, all gap RF columns would be removed automatically,
but with this option enabled gap and non-gap RF columns are treated identically.
 To automatically remove all gap RF columns when using a  \monoi{maskfile}  , then
define the mask in  \monoi{maskfile} as having length equal to the non-gap RF length
in the alignment. To automatically remove all gap RF columns when using
 \mono{-t,} use the \mono{{-}{-}t-rmins} option.         
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-alimerge}}{esl-alimerge} - merge alignments based on their reference (RF) annotation}
 
\subsection*{Synopsis}
 

\noindent
\monob{esl-alimerge }[\monoi{options}] \monoi{alifile1 alifile2}

  (merge two alignment files)

\noindent
\monob{esl-alimerge {-}{-}list }[\monoi{options}] \monoi{listfile}

  (merge many alignment files listed in a file)


\subsection*{Description}
 

\mono{esl-alimerge} reads more than one input alignments, merges them
into a single alignment and outputs it.  

The input alignments must all be
in Stockholm format.  All alignments must have reference ('\#=GC RF') annotation.
Further, the RF annotation must be identical in all alignments once gap
characters in the RF annotation ('.','-','\_') have been removed.  This requirement
allows alignments with different numbers of total columns to be merged
together based on consistent RF annotation, such as alignments created
by successive runs of the \mono{cmalign} program of the INFERNAL package using
the same CM.  Columns which have a gap character in the RF annotation are
called 'insert' columns.  

All sequence data in all input alignments will be
included in the output alignment regardless of the output format (see \mono{{-}{-}outformat
} option below). However, sequences in the merged alignment will usually
contain more gaps ('.') than they did in their respective input alignments.
This is because  \mono{esl-alimerge} must add 100\% gap columns to each individual
input alignment so that insert columns in the other input alignments can
be accomodated in the merged alignment.  

If the output format is Stockholm
or Pfam, annotation will be transferred from the input alignments to the
merged alignment as follows. All per-sequence ('\#=GS') and per-residue ('\#=GR')
annotation is transferred.  Per-file ('\#=GF') annotation is transferred if
it is present and identical in all alignments.  Per-column ('\#=GC') annotation
is transferred if it is present and identical in all alignments once all
insert positions have been removed and  the '\#=GC' annotation includes zero
non-gap characters in insert columns.  

With the  \mono{{-}{-}list}\monoi{ $<$f$>$} option,  \monoi{$<$f$>$} is a file
listing alignment files to merge. In the list file, blank lines and lines
that start with '\#' (comments) are ignored. Each data line contains a single
word: the name of an alignment file to be merged. All alignments in each
file will be merged.  

With the \mono{{-}{-}small} option,  \mono{esl-alimerge} will operate in
memory saving mode and the required RAM for the merge will be minimal (should
be only a few Mb) and independent of the alignment sizes. To use  \mono{{-}{-}small},
all alignments must be in Pfam format (non-interleaved, 1 line/sequence
Stockholm format). You can reformat alignments to Pfam using the \mono{esl-reformat}
Easel miniapp. Without  \mono{{-}{-}small} the required RAM will be equal to roughly
the size of the final merged alignment file which will necessarily be at
least the summed size of all of the input alignment files to be merged
and sometimes several times larger. If you're merging large alignments or
you're experiencing very slow performance of \mono{esl-alimerge}, try reformatting
to Pfam and using \mono{{-}{-}small}.    
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Print brief help; includes version
number and summary of all options, including expert options.  
\item [\monob{-o}\monoi{ $<$f$>$} ] Output
merged alignment to file  \monoi{$<$f$>$} instead of to stdout.  
\item [\monob{-v} ] Be verbose; print information
on the size of the alignments being merged, and the annotation transferred
to the merged alignment to stdout. This option can only be used in combination
with the \mono{-o } option (so that the printed info doesn't corrupt the output
alignment file).  
\item [\monob{{-}{-}small} ] Operate in memory saving mode. Required RAM will
be independent of the sizes of the alignments to merge, instead of roughly
the size of the eventual merged alignment. When enabled, all alignments
must be in Pfam Stockholm (non-interleaved 1 line/seq) format; see \textsf{\mono{esl-reformat}(1)}.
The output alignment will be in Pfam format.  
\item [\monob{{-}{-}rfonly} ] Only include columns
that are not gaps in the GC RF annotation in the merged alignment.   
\item [\monob{{-}{-}outformat}\monoi{
$<$s$>$} ] Write the output alignment in format \monoi{$<$s$>$}. Common choices for  \monoi{$<$s$>$}  include:
\mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal}, \mono{phylip}. The string \monoi{$<$s$>$} is case-insensitive
(\mono{a2m} or \mono{A2M} both work). Default is \mono{stockholm}.   
\item [\monob{{-}{-}rna} ] Specify that the input
alignments are RNA alignments. By default \mono{esl-alimerge} will try to autodetect
the alphabet, but if the alignment is sufficiently small it may be ambiguous.
This option defines the alphabet as RNA.  
\item [\monob{{-}{-}dna} ] Specify that the input alignments
are DNA alignments.  
\item [\monob{{-}{-}amino} ] Specify that the input alignments are protein
alignments.    
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-alipid}}{esl-alipid} - calculate pairwise percent identities for all sequence}
pairs in an MSA  
\subsection*{Synopsis}
\noindent
\monob{esl-alipid} [\monoi{options}] \monoi{msafile}   
\subsection*{Description}
 

\mono{esl-alistat
} calculates the pairwise percent identity of each sequence pair in in the
MSA(s) in  \monoi{msafile.} For each sequence pair, it outputs a line of  \monoi{$<$sqname1$>$}
$<$sqname2$>$ $<$pid$>$ $<$nid$>$ $<$n$>$ where  \monoi{$<$pid$>$}  is the percent identity, \monoi{$<$nid$>$} is the number
of identical aligned pairs, and  \monoi{$<$n$>$}  is the denominator used for the calculation:
the shorter of the two (unaligned) sequence lengths.  

If \monoi{msafile}  is - (a
single dash), alignment input is read from  stdin.  

Only canonical residues
are counted toward \monoi{$<$nid$>$}  and  \monoi{$<$n$>$.} Degenerate residue codes are not counted.
 
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h } ] Print brief help;  includes version number and summary of all
options, including expert options.  
\item [\monob{{-}{-}informat}\monoi{ $<$s$>$} ] Assert that input \monoi{msafile}
is in alignment format \monoi{$<$s$>$}, bypassing format autodetection. Common choices
for  \monoi{$<$s$>$}  include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal}, \mono{phylip}. For more
information, and for codes for some less common formats, see main documentation.
The string \monoi{$<$s$>$} is case-insensitive (\mono{a2m} or \mono{A2M} both work).  
\item [\monob{{-}{-}amino} ] Assert that
the  \monoi{msafile}  contains protein sequences.   
\item [\monob{{-}{-}dna} ] Assert that the  \monoi{msafile}
 contains DNA sequences.   
\item [\monob{{-}{-}rna} ] Assert that the  \monoi{msafile}  contains RNA sequences.
    
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-alirev}}{esl-alirev} - reverse complement a multiple alignment    }
\subsection*{Synopsis}
\noindent
\monob{esl-alirev}
[\monoi{options}] \monoi{msafile}   
\subsection*{Description}
 

\mono{esl-alirev} reads the multiple alignment
in \monoi{msafile} and outputs its reverse complement to stdout.  

An example of
where you might need to do this is when you've downloaded a chunk of multiway
genomic alignment from one of the genome browsers, but your RNA of interest
is on the opposite strand.  

Any per-column and per-residue annotation lines
are reversed as well, including Stockholm format and old SELEX format annotations.
Annotations that Easel recognizes as secondary structure annotation (a
consensus structure line, individual secondary structure lines) will be
"reverse complemented" to preserve proper bracketing orders: for example,
...$<$$<$$<$...$>$$>$$>$ is reverse complemented to $<$$<$$<$...$>$$>$$>$..., not simply reversed to $>$$>$$>$...$<$$<$$<$..., which would be wrong.
 

If \monoi{msafile}  is - (a single dash), alignment input is read from stdin.  


By default the output alignment is written in the same format as the input
alignment. See the \mono{{-}{-}outformat} option to use a different output format.  

Because
the alignment is parsed into Easel's digital internal representation, the
output alignment may differ in certain details from the original alignment;
these details should be inconsequential but may catch your eye. One is that
if you have a reference annotation line, Easel's output will put consensus
residues in upper case, nonconsensus (inserted) residues in lower case.
Another is that the headers for some formats, such as Clustal format, are
written with an arbitrary version number - so you may find yourself revcomping
an alignment in "MUSCLE (3.7) multiple sequence alignment" format and it
could come out claiming to be a "CLUSTAL 2.1 multiple sequence alignment",
just because Easel writes all of its Clustal format alignment files with
that header.  

The \monoi{msafile} must contain nucleic acid sequences (DNA or RNA).
The alphabet will be autodetected by default. See the \mono{{-}{-}dna} or \mono{{-}{-}rna } options
to assert an alphabet.    
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h } ] Print brief help;  includes version
number and summary of all options, including expert options.  
\item [\monob{{-}{-}informat}\monoi{ $<$s$>$}
] Assert that input \monoi{msafile} is in alignment format \monoi{$<$s$>$}, bypassing format autodetection.
Common choices for  \monoi{$<$s$>$}  include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal},
\mono{phylip}. For more information, and for codes for some less common formats,
see main documentation. The string \monoi{$<$s$>$} is case-insensitive (\mono{a2m} or \mono{A2M} both
work).  
\item [\monob{{-}{-}outformat}\monoi{ $<$s$>$} ] Write the output alignment in alignment format \monoi{$<$s$>$}. Common
choices for  \monoi{$<$s$>$}  include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal}, \mono{phylip}.
The string \monoi{$<$s$>$} is case-insensitive (\mono{a2m} or \mono{A2M} both work). Default is to use
same format as the input \monoi{msafile}.  
\item [\monob{{-}{-}dna} ] Assert that the  \monoi{msafile}  contains
DNA sequences.   
\item [\monob{{-}{-}rna} ] Assert that the  \monoi{msafile}  contains RNA sequences.  
 
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-alistat}}{esl-alistat} - summarize a multiple sequence alignment file  }
\subsection*{Synopsis}
\noindent
\monob{esl-alistat}
[\monoi{options}] \monoi{msafile}  
\subsection*{Description}
 

\mono{esl-alistat } summarizes the contents of the
multiple sequence alignment(s) in  \monoi{msafile,}  such as the alignment name,
format, alignment length (number of aligned columns), number of sequences,
average pairwise \% identity, and mean, smallest, and largest raw (unaligned)
lengths of the sequences.  

If  \monoi{msafile} is - (a single dash), multiple alignment
input is read from stdin.    

The  \mono{{-}{-}list}, \mono{{-}{-}icinfo}, \mono{{-}{-}rinfo}, \mono{{-}{-}pcinfo}, \mono{{-}{-}psinfo}, \mono{{-}{-}cinfo},
\mono{{-}{-}bpinfo}, and \mono{{-}{-}iinfo} options allow dumping various statistics on the alignment
to optional output files as described for each of those options below. 


The  \mono{{-}{-}small} option allows summarizing alignments without storing them in
memory and can be useful for large alignment files with sizes that approach
or exceed the amount of available RAM.  When \mono{{-}{-}small} is used,  \mono{esl-alistat}
will print fewer statistics on the alignment, omitting data on the smallest
and largest sequences and the average identity of the alignment. \mono{{-}{-}small} only
works on Pfam formatted alignments (a special type of non-interleaved Stockholm
alignment in which each sequence occurs on a single line) and  \mono{{-}{-}informat
pfam} must be given with \mono{{-}{-}small}. Further, when  \mono{{-}{-}small} is used, the alphabet
must be specified with \mono{{-}{-}amino}, \mono{{-}{-}dna}, or  \mono{{-}{-}rna}.    
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h } ] Print brief help;
 includes version number and summary of all options, including expert options.
 
\item [\monob{-1} ] Use a tabular output format with one line of statistics per alignment
in  \monoi{msafile.} This is most useful when \monoi{msafile} contains many different alignments
(such as a Pfam database in Stockholm format).   
\end{wideitem}

\subsection*{Expert Options}
 \begin{wideitem}
\item [\monob{{-}{-}informat}\monoi{
$<$s$>$} ] Assert that input \monoi{msafile} is in alignment format \monoi{$<$s$>$}, bypassing format
autodetection. Common choices for  \monoi{$<$s$>$}  include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast},
\mono{clustal}, \mono{phylip}. For more information, and for codes for some less common
formats, see main documentation. The string \monoi{$<$s$>$} is case-insensitive (\mono{a2m} or
\mono{A2M} both work).   
\item [\monob{{-}{-}amino} ] Assert that the  \monoi{msafile}  contains protein sequences.
  
\item [\monob{{-}{-}dna} ] Assert that the  \monoi{msafile}  contains DNA sequences.   
\item [\monob{{-}{-}rna} ] Assert that
the  \monoi{msafile}  contains RNA sequences.   
\item [\monob{{-}{-}small} ] Operate in small memory mode
for Pfam formatted alignments. \mono{{-}{-}informat pfam} and one of \mono{{-}{-}amino}, \mono{{-}{-}dna}, or \mono{{-}{-}rna}
must be given as well.  
\item [\monob{{-}{-}list}\monoi{ $<$f$>$} ] List the names of all sequences in all alignments
in  \mono{msafile} to file \monoi{$<$f$>$}. Each sequence name is written on its own line.   
\item [\monob{{-}{-}icinfo}\monoi{
$<$f$>$} ] Dump the information content per position in tabular format to file \monoi{$<$f$>$}.
Lines prefixed with "\#" are comment lines, which explain the meanings of
each of the tab-delimited fields.  
\item [\monob{{-}{-}rinfo}\monoi{ $<$f$>$} ] Dump information on the frequency
of gaps versus nongap residues per position in tabular format to file \monoi{$<$f$>$}.
Lines prefixed with "\#" are comment lines, which explain the meanings of
each of the tab-delimited fields.  
\item [\monob{{-}{-}pcinfo}\monoi{ $<$f$>$} ] Dump per column information on
posterior probabilities in tabular format to file \monoi{$<$f$>$}. Lines prefixed with
"\#" are comment lines, which explain the meanings of each of the tab-delimited
fields.  
\item [\monob{{-}{-}psinfo}\monoi{ $<$f$>$} ] Dump per sequence information on posterior probabilities
in tabular format to file \monoi{$<$f$>$}. Lines prefixed with "\#" are comment lines,
which explain the meanings of each of the tab-delimited fields.  
\item [\monob{{-}{-}iinfo}\monoi{ $<$f$>$}
] Dump information on inserted residues in tabular format to file \monoi{$<$f$>$}. Insert
columns of the alignment are those that are gaps in the reference (\#=GC
RF) annotation. This option only works if the input file is in Stockholm
format with reference annotation. Lines prefixed with "\#" are comment lines,
which explain the meanings of each of the tab-delimited fields.   
\item [\monob{{-}{-}cinfo}\monoi{ $<$f$>$}
] Dump per-column residue counts to file \monoi{$<$f$>$}. If used in combination with \mono{{-}{-}noambig}
ambiguous (degenerate) residues will be ignored and not counted. Otherwise,
they will be marginalized. For example, in an RNA sequence file, a 'N' will
be counted as 0.25 'A', 0.25 'C', 0.25 'G', and 0.25 'U'.  
\item [\monob{{-}{-}noambig} ] With  \mono{{-}{-}cinfo}, do not
count ambiguous (degenerate) residues.   
\item [\monob{{-}{-}bpinfo} ] Dump per-column basepair
counts to file \monoi{$<$f$>$}. Counts appear for each basepair in the consensus secondary
structure (annotated as "\#=GC SS\_cons"). Only basepairs from sequences for
which both paired positions are canonical residues will be counted. That
is, any basepair that is a gap or an ambiguous (degenerate) residue at
either position of the pair is ignored and not counted.   
\item [\monob{{-}{-}weight} ] With  \mono{{-}{-}icinfo},
\mono{{-}{-}rinfo}, \mono{{-}{-}pcinfo}, \mono{{-}{-}iinfo}, \mono{{-}{-}cinfo}, and \mono{{-}{-}bpinfo}, weight counts based on \#=GS WT
annotation in the input  \monoi{msafile}. A residue or basepair from a sequence
with a weight of  \monoi{$<$x$>$} will be considered  \monoi{$<$x$>$} counts.  By default, raw, unweighted
counts are reported; corresponding to each sequence having an equal weight
of 1.     
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-compalign}}{esl-compalign} - compare two multiple sequence alignments  }
\subsection*{Synopsis}
\noindent
\monob{esl-compalign}
[\monoi{options}] \monoi{trusted\_file} \monoi{test\_file}    
\subsection*{Description}
 

\mono{esl-compalign} evaluates
the accuracy of a predicted multiple sequence alignment with respect to
a trusted alignment of the same sequences.   

The  \monoi{trusted\_file}  and \monoi{test\_file}
must contain the same number of alignments. Each predicted alignment in
the  \monoi{test\_file}  will be compared against a single trusted alignment from
the \monoi{trusted\_file.} The first alignments in each file correspond to each other
and will be compared, the second alignment in each file correspond to each
other and will be compared, and so on.  Each corresponding pair of alignments
must contain the same sequences (i.e. if they were unaligned they would be
identical) in the same order in both files. Further, both alignment files
must be in Stockholm format and contain 'reference' annotation, which appears
as "\#=GC RF" per-column markup for each alignment. The number of nongap (non
'.' characters) in the reference (RF) annotation must be identical between
all corresponding alignments in the two files.  

\mono{esl-compalign} reads an alignment
from each file, and compares them based on their 'reference' annotation. 
The number of correctly predicted residues for each sequence is computed
as follows. A residue that is in the Nth nongap RF column in the trusted
alignment must also appear in the Nth nongap RF column in the predicted
alignment to be counted as 'correct', otherwise it is 'incorrect'. A residue
that appears in a gap RF column in the trusted alignment between nongap
RF columns N and N+1 must also appear in a nongap RF column in the predicted
alignment between nongap RF columns N and N+1 to be counted as 'correct',
otherwise it is incorrect.  

The default output of \mono{esl-compalign} lists each
sequence and the number of correctly and incorrectly predicted residues
for that sequence. These counts are broken down into counts for residues
in the predicted alignments that occur in 'match' columns and 'insert' columns.
A 'match' column is one for which the RF annotation does not contain a gap.
An 'insert' column is one for which the RF annotation does contain a gap.
   
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Print brief help; includes version number and summary of all
options.  
\item [\monob{-c} ] Print per-column statistics instead of per-sequence statistics.
 
\item [\monob{-p } ] Print statistics on accuracy versus posterior probability values. The
 \monoi{test\_file} must be annotated with posterior probabilities (\#=GR PP) for
this option to work.   
\end{wideitem}

\subsection*{Expert Options}
 \begin{wideitem}
\item [\monob{{-}{-}p-mask}\monoi{ $<$f$>$} ] This option may only be used
in combination with the  \mono{-p} option. Read a "mask" from file  \monoi{$<$f$>$}. The mask file
must consist of a single line, of only '0' and '1' characters. There must be
exactly RFLEN characters where RFLEN is the number of nongap characters
in the RF annotation of all alignments in both  \monoi{trusted\_file} and \monoi{test\_file}.
Positions of the mask that are '1' characters indicate that the corresponding
nongap RF position is included by the mask. The posterior probability accuracy
statistics for match columns will only pertain to positions that are included
by the mask, those that are excluded will be ignored from the accuracy
calculation.  \mono{{-}{-}c2dfile}\monoi{ $<$f$>$} Save a 'draw file' to file  \monoi{$<$f$>$} which can be read into
the  \mono{esl-ssdraw} miniapp. This draw file will define two postscript pages
for  \mono{esl-ssdraw}. The first page will depict the frequency of errors per match
position and frequency of gaps per match position, indicated by magenta
and yellow, respectively. The darker magenta, the more errors and the darker
yellow, the more gaps. The second page will depict the frequency of errors
in insert positions in shades of magenta, the darker the magenta the more
errors in inserts after each position. See \mono{esl-ssdraw} documentation for more
information on these diagrams.   
\item [\monob{{-}{-}amino} ] Assert that  \monoi{trusted\_file} and  \monoi{test\_file}
contain protein sequences.   
\item [\monob{{-}{-}dna} ] Assert that  \monoi{trusted\_file} and  \monoi{test\_file}
contain DNA sequences.   
\item [\monob{{-}{-}rna} ] Assert that the  \monoi{trusted\_file} and  \monoi{test\_file}
contain RNA sequences.     
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-compstruct}}{esl-compstruct} - calculate accuracy of RNA secondary structure predictions}
  
\subsection*{Synopsis}
\noindent
\monob{esl-compstruct} [\monoi{options}] \monoi{trusted\_file} \monoi{test\_file}    
\subsection*{Description}



\mono{esl-compstruct } evaluates the accuracy of RNA secondary structure predictions
on a per-base-pair basis. The  \monoi{trusted\_file}  contains one or more sequences
with trusted (known) RNA secondary structure annotation. The \monoi{test\_file} contains
the same sequences, in the same order, with predicted RNA secondary structure
annotation. \mono{esl-compstruct } reads the structures and compares them, and calculates
both the sensitivity (the number of true base pairs that are correctly
predicted) and the positive predictive value (PPV; the number of predicted
base pairs that are true). Results are reported for each individual sequence,
and in summary for all sequences together.  

Both files must contain secondary
structure annotation in WUSS notation. Only SELEX and Stockholm formats
support  structure markup at present.  

The default definition of a correctly
predicted base pair is that a true pair (i,j) must exactly match a predicted
pair (i,j).  

Mathews and colleagues (Mathews et al., JMB 288:911-940, 1999)
use a more relaxed definition. Mathews defines "correct" as follows: a true
pair (i,j) is correctly predicted if any of the following pairs are predicted:
(i,j), (i+1,j), (i-1,j), (i,j+1), or (i,j-1). This rule allows for "slipped
helices" off by one base.  The \mono{-m} option activates this rule for both sensitivity
and for specificity. For specificity, the rule is reversed: predicted pair
(i,j) is considered to be true if the true structure contains one of the
five pairs (i,j), (i+1,j), (i-1,j), (i,j+1), or (i,j-1).     
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Print
brief help; includes version number and summary of all options, including
expert options.  
\item [\monob{-m} ] Use the Mathews relaxed accuracy rule (see above), instead
of requiring exact prediction of base pairs.  
\item [\monob{-p } ] Count pseudoknotted base
pairs towards the accuracy, in either trusted or predicted structures. By
default, pseudoknots are ignored.  
\item [Normally, only the  ] \monoi{trusted\_file}  would
have pseudoknot annotation, since most RNA secondary structure prediction
programs do not predict pseudoknots. Using the \mono{-p} option allows you to penalize
the prediction program for not predicting known pseudoknots. In a case where
both the \monoi{trusted\_file}  and the \monoi{test\_file}  have pseudoknot annotation, 
the \mono{-p} option lets you count pseudoknots in evaluating the prediction accuracy.
Beware, however, the case where you use a pseudoknot-capable prediction
program to generate the \monoi{test\_file}, but the  \monoi{trusted\_file}  does not have
pseudoknot annotation; in this case, \mono{-p} will penalize any predicted pseudoknots
when it calculates specificity, even if they're right, because they don't
appear in the trusted annotation. This is probably not what you'd want to
do.    
\end{wideitem}

\subsection*{Expert Options}
 \begin{wideitem}
\item [\monob{{-}{-}quiet} ] Don't print any verbose header information. (Used
by regression test scripts, for example, to suppress version/date information.)
   
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-construct}}{esl-construct} - describe or create a consensus secondary structure }

\subsection*{Synopsis}
\noindent
\monob{esl-construct} [\monoi{options}] \monoi{msafile}  
\subsection*{Description}
 \mono{esl-construct } reports
information on existing consensus secondary structure annotation of an
alignment or derives new consensus secondary structures based on  structure
annotation for individual aligned sequences.  

The alignment file must contain
either individual sequence secondary structure annotation (Stockholm \#=GR
SS), consensus secondary structure annotation (Stockohlm \#=GC SS\_cons),
or both. All structure annotation must be in WUSS notation (Vienna dot paranetheses
notation will be correctly interpreted). At present, the alignment file
must be in Stockholm format and contain RNA or DNA sequences.  

By default,
 \mono{esl-construct} generates lists the sequences in the alignment that have
structure annotation and the number of basepairs in those structures. If
the alignment also contains consensus structure annotation, the default
output will list how many of the individual basepairs overlap with the
consensus basepairs and how many conflict with a consensus basepair.  

For
the purposes of this miniapp, a basepair 'conflict' exists between two basepairs
in different structures, one between columns i and j and the other between
columns k and l, if (i == k and j != l) or (j == l and i != k).  

\mono{esl-construct}
can also be used to derive a new consensus structure based on structure
annotation for individual sequences in the alignment by using any of the
following options:  \mono{-x}, \mono{-r}, \mono{-c}, \mono{{-}{-}indi }\monoi{$<$s$>$}, \mono{{-}{-}ffreq }\monoi{$<$x$>$}, \mono{{-}{-}fmin}. These are described
below. All of these options require the  \mono{-o}\monoi{ $<$f$>$} option be used as well to specify
that a new alignment file  \monoi{$<$f$>$}  be created. Differences between the new alignment(s)
and the input alignment(s) will be limited to the the consensus secondary
structure (\#=GC SS\_cons) annotation and possibly reference (\#=GC RF) annotation.
 
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Print brief help; includes version number and summary of all
options, including expert options.  
\item [\monob{-a} ] List all alignment positions that
are involved in at least one conflicting basepair in at least one sequence
to the screen, and then exit.  
\item [\monob{-v} ] Be verbose; with no other options, list
individual sequence basepair conflicts as well as summary statistics.  
\item [\monob{-x}
] Compute a new consensus structure as the maximally sized set of basepairs
(greatest number of basepairs) chosen from all individual structures that
contains 0 conflicts. Output the alignment with the new SS\_cons annotation.
This option must be used in combination with the \mono{-o } option.   
\item [\monob{-r} ] Remove any
consensus basepairs that conflict with $>$= 1 individual basepair and output
the alignment with the new SS\_cons annotation.  This option must be used
in combination with the  \mono{-o } option.   
\item [\monob{-c} ] Define a new consensus secondary
structure as the individual structure annotation that has the maximum number
of consistent basepairs with the existing consensus secondary structure
annotation. This option must be used in combination with the  \mono{-o } option.
  
\item [\monob{{-}{-}rfc} ] With  \mono{-c},\mono{} set the reference annotation (\#=GC RF) as the sequence whose
 individual structure becomes the consensus structure.   
\item [\monob{{-}{-}indi}\monoi{ $<$s$>$} ] Define a
new consensus secondary structure as the individual structure annotation
from sequence named \monoi{$<$s$>$}. This option must be used in combination with the
\mono{-o } option.   
\item [\monob{{-}{-}rfindi} ] With  \mono{{-}{-}indi }\monoi{$<$s$>$},  set the reference annotation (\#=GC RF)
as the sequence named \mono{$<$s$>$}.  
\item [\monob{{-}{-}ffreq}\monoi{ $<$x$>$} ] Define a new consensus structure as the
set of basepairs between columns i:j that are paired in more than  \monoi{$<$x$>$} fraction
of the individual sequence structures. This option must be used in combination
with the \mono{-o } option.   
\item [\monob{{-}{-}fmin} ] Same as \mono{{-}{-}ffreq}\monoi{ $<$x$>$}\mono{} except find the maximal  \monoi{$<$x$>$} that
gives a consistent consensus structure. A consistent structure has each
base (alignment position) as a member of at most 1 basepair.    
\item [\monob{-o}\monoi{ $<$s$>$}\mono{,}\monoi{} ] Output
the alignment(s) with new consensus structure annotation to file \monoi{$<$f$>$}.  
\item [\monob{{-}{-}pfam
} ] With \mono{-o}, specify that the alignment output format be Pfam format, a special
type of non-interleaved Stockholm on which each sequence appears on a single
line.  
\item [\monob{-l}\monoi{ $<$f$>$} ] Create a new file  \monoi{$<$f$>$}  that lists the sequences that have at least
one basepair that conflicts with a consensus basepair.  
\item [\monob{{-}{-}lmax}\monoi{ $<$n$>$} ] With \mono{-l}, only
list sequences that have more than  \monoi{$<$n$>$} basepairs that conflict with the
consensus structure to the list file.   
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-histplot}}{esl-histplot} - collate data histogram, output xmgrace datafile  }
\subsection*{Synopsis}
\noindent
\monob{esl-histplot}
[\monoi{options}] \monoi{datafile}   
\subsection*{Description}
 

\mono{esl-histplot} summarizes numerical data
in the input file \monoi{datafile.}  

One real-numbered value is taken from each line
of the input file.  Each line is split into whitespace-delimited fields,
and one field is converted to data. By default this is the first field;
this can be  changed by the  \mono{-f } option.  

Default output is a survival plot
(Prob(value $>$ x)) in xmgrace XY data format, to stdout. Output may be directed
to a file with the \mono{-o} option.  

If  \monoi{datafile} is - (a single dash), input lines
are read from stdin instead of opening a file.     
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-f}\monoi{ $<$n$>$} ] Read data
from whitespace-delimited field  \monoi{$<$n$>$} on each line, instead of the first field.
Fields are numbered starting from 1.  
\item [\monob{-h } ] Print brief help; includes version
number and summary of all options, including expert options.  
\item [\monob{-o}\monoi{ $<$f$>$} ] Send output
to file \monoi{$<$f$>$} instead of stdout.    
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-mask}}{esl-mask} - mask sequence residues with X's (or other characters)  }
\subsection*{Synopsis}
\noindent
\monob{esl-mask}
[\monoi{options}] \monoi{seqfile} \monoi{maskfile}   
\subsection*{Description}
 

\mono{esl-mask} reads lines from  \monoi{maskfile}
 that give start/end coordinates for regions in each sequence in  \monoi{seqfile},
masks these residues (changes them to X's), and outputs the masked sequence.
 

The \monoi{maskfile} is a space-delimited file. Blank lines and lines that start
with '\#' (comments) are ignored. Each data line contains at least three fields:
\monoi{seqname}, \monoi{start}, and \monoi{end}.\monoi{} The  \monoi{seqname}  is the name of a sequence in the
 \monoi{seqfile}, and  \monoi{start} and  \monoi{end}  are coordinates defining a region in that
sequence. The coordinates are indexed $<$1..L$>$ with respect to a sequence of length
$<$L$>$.  

By default, the sequence names must appear in exactly the same order
and number as the sequences in the \monoi{seqfile.} This is easy to enforce, because
the format of \monoi{maskfile}  is also legal as a list of names for  \mono{esl-sfetch},\mono{}
so you can always fetch a temporary sequence file with  \mono{esl-sfetch } and
pipe that to  \mono{esl-mask}. (Alternatively, see the  \mono{-R} option for fetching from
an SSI-indexed \monoi{seqfile}.)  

The default is to mask the region indicated by
 \monoi{$<$start$>$}..\monoi{$<$end$>$}. Alternatively, everything but this region can be masked; see
the \mono{-r} reverse masking option.  

The default is to mask residues by converting
them to X's. Any other masking character can be chosen (see \mono{-m} option), or
alternatively, masked residues can be lowercased (see \mono{-l} option).    
\subsection*{Options}

\begin{wideitem}
\item [\monob{-h} ] Print brief help; includes version number and summary of all options,
including expert options.  
\item [\monob{-l} ] Lowercase; mask by converting masked characters
to lower case and unmasked characters to upper case.  
\item [\monob{-m}\monoi{ $<$c$>$} ] Mask by converting
masked residues to  \monoi{$<$c$>$} instead of the default X.  
\item [\monob{-o}\monoi{ $<$f$>$} ] Send output to file
\monoi{$<$f$>$} instead of stdout.  
\item [\monob{-r} ] Reverse mask; mask everything outside the region
\monoi{start..end,}  as opposed to the default of masking that region.  
\item [\monob{-R} ] Random access;
 fetch sequences from  \monoi{seqfile} rather than requiring that sequence names
in \monoi{maskfile} and \monoi{seqfile} come in exactly the same order and number.  The
\monoi{seqfile} must be SSI indexed (see \mono{esl-sfetch {-}{-}index}.)  
\item [\monob{-x}\monoi{ $<$n$>$} ] Extend all masked
regions by up to $<$n$>$ residues on each side.  For normal masking, this means
masking \monoi{$<$start$>$}-\monoi{$<$n$>$}..\monoi{$<$end$>$}+\monoi{$<$n$>$}.  For reverse masking, this means masking 1..\monoi{$<$start$>$}-1+\monoi{$<$n$>$}
and \monoi{$<$end$>$}+1-\monoi{$<$n$>$}..L in a sequence of length L.   
\item [\monob{{-}{-}informat}\monoi{ $<$s$>$} ] Assert that input \monoi{seqfile}
is in format \monoi{$<$s$>$}, bypassing format autodetection. Common choices for  \monoi{$<$s$>$}  include:
\mono{fasta}, \mono{embl}, \mono{genbank.} Alignment formats also work; common choices include:
\mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal}, \mono{phylip}. For more information, and
for codes for some less common formats, see main documentation. The string
\monoi{$<$s$>$} is case-insensitive (\mono{fasta} or \mono{FASTA} both work).     
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-mixdchlet}}{esl-mixdchlet} - fitting mixture Dirichlets to count data  }
\subsection*{Synopsis}
 

\noindent
\monob{esl-mixdchlet fit} [\monoi{options}] \monoi{Q K in\_countfile out\_mixchlet}

  (train a new mixture Dirichlet)

\noindent
\monob{esl-mixdchlet score} [\monoi{options}] \monoi{mixdchlet\_file counts\_file}

  (calculate log likelihood of count data, given mixture Dirichlet)

\noindent
\monob{esl-mixdchlet gen }[\monoi{options}] \monoi{mixdchlet\_file}

  (generate synthetic count data from mixture Dirichlet)

\noindent
\monob{esl-mixdchlet sample }[\monoi{options}]

  (sample a random mixture Dirichlet for testing)

  
\subsection*{Description}
 

The \mono{esl-mixdchlet} miniapp is for training mixture Dirichlet
priors, such as the priors used in HMMER and Infernal. It has four subcommands:
\mono{fit,} \mono{score,} \mono{gen,} and \mono{sample.} The most important subcommand is \mono{fit,} which
is the subcommand for fitting a new mixture Dirichlet distribution to a
collection of count vectors (for example, emission or transition count
vectors from Pfam or Rfam training sets).  

Specifically, \mono{esl-mixdchlet fit}
fits a new mixture Dirichlet distribution with \monoi{Q} mixture components to
the count vectors (of alphabet size \monoi{K} ) in input file \monoi{in\_countfile,} and
saves the mixture Dirichlet into output file \monoi{out\_mixdchlet.}  

The input count
vector file \monoi{in\_countfile} contains one count vector of length \monoi{K} fields per
line, for any number of lines. Blank lines and lines starting in \# (comments)
are ignored. Fields are nonnegative real values; they do not have to be
integers, because they can be weighted counts.  

The format of a mixture
Dirichlet file \monoi{out\_mixdchlet} is as follows. The first line has two fields,
\monoi{K} Q, where \monoi{K} is the alphabet size and  \monoi{Q} is the number of mixture components.
The next \monoi{Q} lines consist of \monoi{K+1} fields. The first field is the mixture coefficient
\monoi{q\_k,} followed by \monoi{K} fields with the Dirichlet alpha[k][a] parameters for
this component.  

The \mono{esl-mixdchlet score} subcommand calculates the log likelihood
of the count vector data in \monoi{counts\_file,} given the mixture Dirichlet in
\monoi{mixdchlet\_file.}  

The \mono{esl-mixdchlet gen} subcommand generates synthetic count
data, given a mixture Dirichlet.  

The \mono{esl-mixdchlet sample} subcommand creates
a random mixture Dirichlet distribution  and outputs it to standard output.
  
\subsection*{Options for Fit Subcommand}
 \begin{wideitem}
\item [\monob{-h} ] Print brief help specific to the \mono{fit} subcommand.
 
\item [\monob{-s}\monoi{ $<$seed$>$} ] Set random number generator seed to nonnegative integer \monoi{$<$seed$>$.} Default
is 0, which means to use a quasirandom arbitrary seed. Values $>$0 give reproducible
results.     
\end{wideitem}

\subsection*{Options for Score Subcommand}
 \begin{wideitem}
\item [\monob{-h} ] Print brief help specific to
the \mono{score} subcommand.    
\end{wideitem}

\subsection*{Options for Gen Subcommand}
 \begin{wideitem}
\item [\monob{-h} ] Print brief help specific
to the \mono{gen} subcommand.  
\item [\monob{-s}\monoi{ $<$seed$>$} ] Set random number generator seed to nonnegative
integer \monoi{$<$seed$>$.} Default is 0, which means to use a quasirandom arbitrary seed.
Values $>$0 give reproducible results.   
\item [\monob{-M}\monoi{ $<$M$>$} ] Generate \monoi{$<$M$>$} counts per sampled
vector. (Default 100.)  
\item [\monob{-N}\monoi{ $<$N$>$} ] Generate \monoi{$<$N$>$} count vectors. (Default 1000.)   
\end{wideitem}

\subsection*{Options
for Sample Subcommand}
 \begin{wideitem}
\item [\monob{-h} ] Print brief help specific to the \mono{sample} subcommand.
 
\item [\monob{-s}\monoi{ $<$seed$>$} ] Set random number generator seed to nonnegative integer \monoi{$<$seed$>$.} Default
is 0, which means to use a quasirandom arbitrary seed. Values $>$0 give reproducible
results.   
\item [\monob{-K}\monoi{ $<$K$>$} ] Set the alphabet size to  \monoi{$<$K$>$.} (Default is 20, for amino acids.)
 
\item [\monob{-Q}\monoi{ $<$Q$>$} ] Set the number of mixture components to \monoi{$<$Q$>$.} (Default is 9.)     
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-reformat}}{esl-reformat} - convert sequence file formats  }
\subsection*{Synopsis}
\noindent
\monob{esl-reformat} [\monoi{options}]
\monoi{format} \monoi{seqfile}   
\subsection*{Description}
 

\mono{esl-reformat} reads the sequence file \monoi{seqfile}
in any supported format, reformats it into a new format specified by  \monoi{format},
then outputs the reformatted text.  

The  \monoi{format} argument must (case-insensitively)
match a supported sequence file format. Common choices for  \monoi{format} include:
\mono{fasta}, \mono{embl}, \mono{genbank.} If \monoi{seqfile} is an alignment file, alignment output
formats also work. Common choices include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast},
\mono{clustal}, \mono{phylip}. For more information, and for codes for some less common
formats, see main documentation. The string \monoi{$<$s$>$} is case-insensitive (\mono{fasta}
or \mono{FASTA} both work).  

Unaligned format files cannot be reformatted to aligned
formats. However, aligned formats can be reformatted to unaligned formats,
in which case gap characters are  simply stripped out.  
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-d } ] DNA;
convert U's to T's, to make sure a nucleic acid sequence is shown as DNA
not RNA. See \mono{-r.}   
\item [\monob{-h} ] Print brief help; includes version number and summary
of all options, including expert options.   
\item [\monob{-l} ] Lowercase; convert all sequence
residues to lower case. See \mono{-u}.   
\item [\monob{-n} ] For DNA/RNA sequences, converts any character
that's not unambiguous RNA/DNA (e.g. ACGTU/acgtu) to an N. Used to convert
IUPAC ambiguity codes to N's, for software that can't handle all IUPAC codes
(some public RNA folding codes, for example). If the file is an alignment,
gap characters are also left unchanged. If sequences are not nucleic acid
sequences, this option will corrupt the data in a predictable fashion. 
 
\item [\monob{-o}\monoi{ $<$f$>$} ] Send output to file \monoi{$<$f$>$} instead of stdout.   
\item [\monob{-r } ] RNA; convert T's to U's,
to make sure a nucleic acid sequence is shown as RNA not DNA. See \mono{-d}.   
\item [\monob{-u}
] Uppercase; convert all sequence residues to upper case. See \mono{-l}.   
\item [\monob{-x} ] For DNA
sequences, convert non-IUPAC characters (such as X's) to N's. This is for compatibility
with benighted people who insist on using X instead of the IUPAC ambiguity
character N. (X is for ambiguity in an amino acid residue).  
\item [Warning: like
the ] \mono{-n} option, the code doesn't check that you are actually giving it DNA.
It simply literally just converts non-IUPAC DNA symbols to N. So if you accidentally
give it protein sequence, it will happily convert most every amino acid
residue to an N.     
\end{wideitem}

\subsection*{Expert Options}
  \begin{wideitem}
\item [\monob{{-}{-}gapsym}\monoi{ $<$c$>$} ] Convert all gap characters
to  \monoi{$<$c$>$}. Used to prepare alignment files for programs with strict requirements
for gap symbols. Only makes sense if the input  \monoi{seqfile} is an alignment.
 
\item [\monob{{-}{-}informat}\monoi{ $<$s$>$} ] Assert that input \monoi{seqfile} is in format \monoi{$<$s$>$}, bypassing format
autodetection. Common choices for  \monoi{$<$s$>$}  include: \mono{fasta}, \mono{embl}, \mono{genbank.} Alignment
formats also work; common choices include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast},
\mono{clustal}, \mono{phylip}. For more information, and for codes for some less common
formats, see main documentation. The string \monoi{$<$s$>$} is case-insensitive (\mono{fasta}
or \mono{FASTA} both work).  
\item [\monob{{-}{-}mingap} ] If  \monoi{seqfile} is an alignment, remove any columns
that contain 100\% gap or missing data characters, minimizing the overall
length of the alignment. (Often useful if you've extracted a subset of aligned
sequences from a larger alignment.)  
\item [\monob{{-}{-}keeprf} ] When used in combination with
\mono{{-}{-}mingap}, never remove a column that is not a gap in the reference (\#=GC
RF)  annotation, even if the column contains 100\% gap characters in  all
aligned sequences. By default with \mono{{-}{-}mingap}, nongap RF columns that are 100\%
gaps in all sequences are removed.  
\item [\monob{{-}{-}nogap} ] Remove any aligned columns that
contain any gap or missing data symbols at all. Useful as a prelude to phylogenetic
analyses, where you only want to analyze columns containing 100\% residues,
so you want to strip out any columns with gaps in them.  Only makes sense
if the file is an alignment file.  
\item [\monob{{-}{-}wussify} ] Convert RNA secondary structure
annotation strings (both consensus and individual) from old "KHS" format,
$>$$<$, to the new WUSS notation, $<$$>$. If the notation is already in WUSS format,
this option will screw it up, without warning. Only SELEX and Stockholm
format files have secondary structure markup at present.  
\item [\monob{{-}{-}dewuss} ] Convert
RNA secondary structure annotation strings from the new WUSS notation,
$<$$>$, back to the old KHS format, $>$$<$. If the annotation is already in KHS, this
option will corrupt it, without warning. Only SELEX and Stockholm format
files have secondary structure markup.  
\item [\monob{{-}{-}fullwuss} ] Convert RNA secondary structure
annotation strings from simple (input) WUSS notation to full (output) WUSS
notation.  
\item [\monob{{-}{-}replace}\monoi{ $<$s$>$} ] \monoi{$<$s$>$} must be in the format \monoi{$<$s1$>$:$<$s2$>$} with equal numbers of
characters in  \monoi{$<$s1$>$} and  \monoi{$<$s2$>$} separated by a ":" symbol. Each character from
\monoi{$<$s1$>$} in the input file will be replaced by its counterpart (at the same position)
from \monoi{$<$s2$>$}. Note that special characters in  \monoi{$<$s$>$} (such as "~") may need to be
prefixed by a "$\backslash$" character.   
\item [\monob{{-}{-}small} ] Operate in small memory mode for input
alignment files in  Pfam format. If not used, each alignment is stored in
memory so the required memory will be roughly the size of the largest alignment
in the input file. With  \mono{{-}{-}small},\mono{} input alignments are not stored in memory.
 This option only works in combination with  \mono{{-}{-}informat pfam} and output format
 \monoi{pfam} or \monoi{afa}.\monoi{}    
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-selectn}}{esl-selectn} - select random subset of lines from file  }
\subsection*{Synopsis}
\noindent
\monob{esl-selectn}
[\monoi{options}] \monoi{nlines} \monoi{filename}   
\subsection*{Description}
 

\mono{esl-selectn} selects  \monoi{nlines} lines
at random from file  \monoi{filename} and outputs them on  \monoi{stdout.}  

If  \monoi{filename}
is - (a single dash), input is read from stdin.   

Uses an efficient reservoir
sampling algorithm that only requires only a single pass through \monoi{filename,}
and memory storage proportional to  \monoi{nlines} (and importantly, not to the
size of the file \monoi{filename} itself). \mono{esl-selectn } can therefore be used to
create large scale statistical sampling  experiments, especially in combination
with other Easel miniapplications.   
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Print brief help; includes
version number and summary of all options, including expert options.   
\item [\monob{{-}{-}seed}\monoi{
$<$d$>$} ] Set the random number seed to \monoi{$<$d$>$,} an integer $>$= 0. The default is 0, which
means to use a randomly selected seed. A seed $>$ 0 results in reproducible
identical samples from different runs of the same command.   
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-seqrange}}{esl-seqrange} - determine a range of sequences for one of many parallel}
processes  
\subsection*{Synopsis}
\noindent
\monob{esl-sfetch} [\monoi{options}] \monoi{seqfile} \monoi{procidx} \monoi{nproc}  
\subsection*{Description}



\mono{esl-seqrange} reads an SSI-indexed  \monoi{seqfile} and determines the range of sequence
indices in that file that process number  \monoi{procidx} out of \monoi{nproc} total processes
should operate on during a parallel processing of  \monoi{seqfile}.  

The  \monoi{seqfile}
 must be indexed first using  \mono{esl-sfetch {-}{-}index} \monoi{seqfile}. This creates an SSI
index file \monoi{seqfile}.ssi. An SSI file is required in order for \mono{esl-seqrange}
to work.  

Sequence index ranges are calculated using a simple rule: the
number of sequences for each process should be identical, or as close as
possible to identical, across all processes. The lengths of the sequences
are not considered (even though they probably should be).  
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Print
brief help; includes version number and summary of all options, including
expert options.  
\item [\monob{{-}{-}informat}\monoi{ $<$s$>$} ] Assert that input \monoi{seqfile} is in format \monoi{$<$s$>$}, bypassing
format autodetection. Common choices for  \monoi{$<$s$>$}  include: \mono{fasta}, \mono{embl}, \mono{genbank.}
Alignment formats also work; common choices include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa},
\mono{psiblast}, \mono{clustal}, \mono{phylip}. For more information, and for codes for some
less common formats, see main documentation. The string \monoi{$<$s$>$} is case-insensitive
(\mono{fasta} or \mono{FASTA} both work).    
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-seqstat}}{esl-seqstat} - summarize contents of a sequence file  }
\subsection*{Synopsis}
\noindent
\monob{esl-seqstat}
[\monoi{options}] \monoi{seqfile}  
\subsection*{Description}
 

\mono{esl-seqstat } summarizes the contents of the
\monoi{seqfile}. It prints the format, alphabet type, number of sequences, total
number of residues, and the mean, smallest, and largest sequence length.
 

If  \monoi{seqfile} is - (a single dash), sequence input is read from stdin.   
 
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h } ] Print brief help;  includes version number and summary of all
options, including expert options.  
\item [\monob{-a} ] Additionally show a summary statistic
line showing the name, length, and description of each individual sequence.
Each of these lines is prefixed by an = character, in order to allow these
lines to be easily grepped out of the output.  
\item [\monob{-c} ] Additionally print the
residue composition of the sequence file.    
\end{wideitem}

\subsection*{Expert Options}
 \begin{wideitem}
\item [\monob{{-}{-}informat}\monoi{ $<$s$>$} ] Assert
that input \monoi{seqfile} is in format \monoi{$<$s$>$}, bypassing format autodetection. Common
choices for  \monoi{$<$s$>$}  include: \mono{fasta}, \mono{embl}, \mono{genbank.} Alignment formats also work;
common choices include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal}, \mono{phylip}.
For more information, and for codes for some less common formats, see main
documentation. The string \monoi{$<$s$>$} is case-insensitive (\mono{fasta} or \mono{FASTA} both work).
  
\item [\monob{{-}{-}amino} ] Assert that the  \monoi{seqfile}  contains protein sequences.   
\item [\monob{{-}{-}dna} ] Assert
that the  \monoi{seqfile}  contains DNA sequences.   
\item [\monob{{-}{-}rna} ] Assert that the  \monoi{seqfile}
 contains RNA sequences.     
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-sfetch}}{esl-sfetch} - retrieve (sub-)sequences from a sequence file  }
\subsection*{Synopsis}



\noindent
\monob{esl-sfetch} [\monoi{options}] \monoi{seqfile key}

  (retrieve a single sequence by key)

\noindent
\monob{esl-sfetch -c }\monoi{from}\monob{..}\monoi{to }[\monoi{options}]\monoi{ seqfile key}

  (retrieve a single subsequence by key and coords)

\noindent
\monob{esl-sfetch -f }[\monoi{options}] \monoi{seqfile keyfile}

  (retrieve multiple sequences using a file of keys)

\noindent
\monob{esl-sfetch -Cf }[\monoi{options}] \monoi{seqfile subseq-coord-file}

  (retrieve multiple subsequences using file of keys and coords)

\noindent
\monob{esl-sfetch {-}{-}index}\monoi{ msafile}

  (index a sequence file for retrievals)

  
\subsection*{Description}
 

\mono{esl-sfetch} retrieves one or more sequences or subsequences
from \monoi{seqfile}.  

The  \monoi{seqfile}  must be indexed using \mono{esl-sfetch {-}{-}index}\monoi{ seqfile}.
This creates an SSI index file \monoi{seqfile}.ssi.  

To retrieve a single complete
sequence, do \mono{esl-sfetch}\monoi{ seqfile key}, where  \monoi{key} is the name or accession
of the desired sequence.  

To retrieve a single subsequence rather than a
complete sequence, use the  \mono{-c }\monoi{start}..\monoi{end} option to provide \monoi{start} and \monoi{end}
coordinates. The \monoi{start} and \monoi{end} coordinates are provided as one string, separated
by any nonnumeric, nonwhitespace character or characters you like; see
the \mono{-c} option below for more details.  

To retrieve more than one complete
sequence at once, you may use the  \mono{-f} option, and the second command line
argument will specify the name of a  \monoi{keyfile} that contains a list of names
or accessions, one per line; the first whitespace-delimited field on each
line of this file is parsed as the name/accession.  

To retrieve more than
one subsequence at once, use the \mono{-C} option in addition to \mono{-f}, and now the
second argument is parsed as a list of subsequence coordinate lines. See
the \mono{-C} option below for more details, including the format of these lines.
  

 

In DNA/RNA files, you may extract (sub-)sequences in reverse complement
orientation in two different ways: either by providing a  \monoi{from} coordinate
that is greater than  \monoi{to},\monoi{} or by providing the  \monoi{-r} option.  

When the \mono{-f } option
is used to do multiple (sub-)sequence retrieval, the file argument may be
- (a single dash), in which case the list of names/accessions (or subsequence
coordinate lines) is read from standard input. However, because a standard
input stream can't be SSI indexed, (sub-)sequence retrieval from stdin may
be slow.   
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Print brief help; includes version number and summary
of all options, including expert options.  
\item [\monob{-c}\monoi{ coords} ] Retrieve a subsequence
with start and end coordinates specified by the  \monoi{coords} string. This string
consists of start  and end coordinates separated by any nonnumeric, nonwhitespace
character or characters you like; for example,  \mono{-c 23..100}, \mono{-c 23/100}, or \mono{-c
23-100} all work. To retrieve a suffix of a subsequence, you can omit the
 \monoi{end} ; for example, \mono{-c 23:} would work. To specify reverse complement (for
DNA/RNA sequence), you can specify  \monoi{from} greater than \monoi{to}; for example,
\mono{-c 100..23} retrieves the reverse complement strand from 100 to 23.  
\item [\monob{-f} ] Interpret
the second argument as a  \monoi{keyfile} instead of as just one \monoi{key.}  The first
whitespace-limited field on each line of  \monoi{keyfile} is interpreted as a name
or accession to be fetched. This option doesn't work with the \mono{{-}{-}index} option.
 Any other fields on a line after the first one are ignored. Blank lines
and lines beginning with \# are ignored.  
\item [\monob{-o}\monoi{ $<$f$>$} ] Output retrieved sequences
to a file  \monoi{$<$f$>$} instead of to stdout.   
\item [\monob{-n}\monoi{ $<$s$>$} ] Rename the retrieved (sub-)sequence
 \monoi{$<$s$>$}. Incompatible with  \mono{-f}.  
\item [\monob{-r} ] Reverse complement the retrieved (sub-)sequence.
Only accepted for DNA/RNA sequences.  
\item [\monob{-C} ] Multiple subsequence retrieval mode,
with  \mono{-f} option (required). Specifies that the second command line argument
is to be parsed as a subsequence coordinate file, consisting of lines containing
four whitespace-delimited fields: \monoi{new\_name}, \monoi{from}, \monoi{to}, \monoi{name/accession}. For
each such line, sequence \monoi{name/accession} is found, a subsequence \monoi{from}..\monoi{to}
is extracted, and the subsequence is renamed  \monoi{new\_name}  before being output.
 Any other fields after the first four are ignored. Blank lines and lines
beginning with \# are ignored.   
\item [\monob{-O} ] Output retrieved sequence to a file named
\monoi{key}. This is a convenience for saving some typing: instead of  

  \user{\% esl-sfetch -o SRPA\_HUMAN swissprot SRPA\_HUMAN}

you can just type 

  \user{\% esl-sfetch -O swissprot SRPA\_HUMAN}

The \mono{-O } option only works if you're retrieving a single alignment; it is
incompatible with  \mono{-f}.  
\item [\monob{{-}{-}index} ] Instead of retrieving a \monoi{key,} the special command
\mono{esl-sfetch {-}{-}index} \monoi{seqfile} produces an SSI index of the names and accessions
of the alignments in the  \monoi{seqfile.} Indexing should be done once on the \monoi{seqfile}
to prepare it for all future fetches.   
\end{wideitem}

\subsection*{Expert Options}
 \begin{wideitem}
\item [\monob{{-}{-}informat}\monoi{ $<$s$>$} ] Assert
that  \monoi{seqfile} is in format \monoi{$<$s$>$}, bypassing format autodetection. Common choices
for  \monoi{$<$s$>$}  include: \mono{fasta}, \mono{embl}, \mono{genbank.} Alignment formats also work; common
choices include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal}, \mono{phylip}. For more
information, and for codes for some less common formats, see main documentation.
The string \monoi{$<$s$>$} is case-insensitive (\mono{fasta} or \mono{FASTA} both work).    
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-shuffle}}{esl-shuffle} - shuffling sequences or generating random ones  }
\subsection*{Synopsis}



\noindent
\monob{esl-shuffle }[\monoi{options}] \monoi{seqfile}

  (shuffle sequences)

\noindent
\monob{esl-shuffle -G }[\monoi{options}]

  (generate random sequences)

\noindent
\monob{esl-shuffle -A }[\monoi{options}] \monoi{msafile}

  (shuffle multiple sequence alignments)

 
\subsection*{Description}
 

\mono{esl-shuffle} has three different modes of operation.  

By default,
 \mono{esl-shuffle} reads individual sequences from  \monoi{seqfile}, shuffles them, and
outputs the shuffled sequences. By default, shuffling is done by preserving
monoresidue composition; other options are listed below.  

With the  \mono{-G } option,
\mono{esl-shuffle} generates some number of random sequences of some length in
some alphabet. The \mono{-N} option controls the number (default is 1), the \mono{-L} option
controls the length (default is 0),  and the  \mono{{-}{-}amino}, \mono{{-}{-}dna}, and  \mono{{-}{-}rna} options
control the alphabet.  

With the  \mono{-A} option,  \mono{esl-shuffle} reads one or more
multiple alignments from \monoi{msafile} shuffles them, and outputs the shuffled
alignments. By default, the alignment is shuffled columnwise (i.e. column
order is permuted). Other options are listed below.   
\subsection*{General Options}
 \begin{wideitem}
\item [\monob{-h }
] Print brief help;  includes version number and summary of all options,
including expert options.  
\item [\monob{-o}\monoi{ $<$f$>$} ] Direct output to a file named \monoi{$<$f$>$} rather than
to stdout.  
\item [\monob{-N}\monoi{ $<$n$>$} ] Generate  \monoi{$<$n$>$} sequences, or \monoi{$<$n$>$}  perform independent shuffles
per input sequence or alignment.  
\item [\monob{-L}\monoi{ $<$n$>$} ] Generate sequences of length \monoi{$<$n$>$}, or
truncate output shuffled sequences or alignments to a length of \monoi{$<$n$>$}.     
\end{wideitem}

\subsection*{Sequence
Shuffling Options}
 These options only apply in default (sequence shuffling)
mode.  They are mutually exclusive.  \begin{wideitem}
\item [\monob{-m} ] Monoresidue shuffling (the default):
preserve monoresidue composition exactly. Uses the Fisher/Yates algorithm
(aka Knuth's "Algorithm P").  
\item [\monob{-d} ] Diresidue shuffling; preserve diresidue composition
exactly.  Uses the Altschul/Erickson algorithm (Altschul and Erickson, 1986).
A more efficient algorithm (Kandel and Winkler 1996) is known but has not
yet been implemented in Easel.  
\item [\monob{-0} ] 0th order Markov generation: generate
a sequence of the same length with the same 0th order Markov frequencies.
Such a sequence will approximately preserve the monoresidue composition
of the input.  
\item [\monob{-1} ] 1st order Markov generation: generate a sequence of the
same length with the same 1st order Markov frequencies. Such a sequence
will  approximately preserve the diresidue composition of the input.  
\item [\monob{-r}
] Reversal; reverse each input.  
\item [\monob{-w}\monoi{ $<$n$>$} ] Regionally shuffle the input in nonoverlapping
windows of size  \monoi{$<$n$>$}  residues, preserving exact monoresidue composition
in each window.  

   
\end{wideitem}

\subsection*{Multiple Alignment Shuffling Options}
 \begin{wideitem}
\item [\monob{-b} ] Sample columns with replacement,
in order to generate a bootstrap-resampled alignment dataset.   
\item [\monob{-v} ] Shuffle
residues with each column independently; i.e., permute residue order in each
column ("vertical" shuffling).   
\end{wideitem}

\subsection*{Sequence Generation Options}
 One of these
must be selected, if \mono{-G} is used.  \begin{wideitem}
\item [\monob{{-}{-}amino} ] Generate amino acid sequences.  
\item [\monob{{-}{-}dna}
] Generate DNA sequences.  
\item [\monob{{-}{-}rna} ] Generate RNA sequences.    
\end{wideitem}

\subsection*{Expert Options}
 \begin{wideitem}
\item [\monob{{-}{-}informat}\monoi{
$<$s$>$} ] Assert that input \monoi{seqfile} is in format \monoi{$<$s$>$}, bypassing format autodetection.
Common choices for  \monoi{$<$s$>$}  include: \mono{fasta}, \mono{embl}, \mono{genbank.} Alignment formats
also work; common choices include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal},
\mono{phylip}. For more information, and for codes for some less common formats,
see main documentation. The string \monoi{$<$s$>$} is case-insensitive (\mono{fasta} or \mono{FASTA}
both work).   
\item [\monob{{-}{-}seed}\monoi{ $<$n$>$} ] Specify the seed for the random number generator, where
the seed \monoi{$<$n$>$} is an integer greater than zero. This can be used to make the
results of  \mono{esl-shuffle } reproducible. If  \monoi{$<$n$>$} is 0, the random number generator
is seeded arbitrarily and stochastic simulations will vary from run to
run. Arbitrary seeding (0) is the default.     
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-ssdraw}}{esl-ssdraw} - create postscript secondary structure diagrams  }
\subsection*{Synopsis}
\noindent
\monob{esl-ssdraw}
[\monoi{options}] \monoi{msafile} \monoi{postscript\_template} \monoi{postscript\_output\_file}  
\subsection*{Description}



\mono{esl-ssdraw} reads an existing template consensus secondary structure diagram
from \monoi{postscript\_template} and creates new postscript diagrams including
the template structure but with positions colored differently based on
alignment statistics such as frequency of gaps per position, average posterior
probability per position or information content per position. Additionally,
all or some of the aligned sequences can be drawn separately, with nucleotides
or posterior probabilities mapped onto the corresponding positions of the
consensus structure.  

The alignment must be in Stockholm format with per-column
reference annotation (\#=GC RF). The sequences in the alignment must be RNA
or DNA sequences. The \monoi{postscript\_template} file must contain one page that
includes $<$rflen$>$ consensus nucleotides (positions), where $<$rflen$>$ is the number
of nongap characters in the reference (RF) annotation of the first alignment
in \monoi{msafile}. The specific format required in the  \monoi{postscript\_template} is
described below in the INPUT section. Postscript diagrams will only be created
for the first alignment in \monoi{msafile}.\monoi{}   
\subsection*{Output}
 

By default (if run with zero
command line options), \mono{esl-ssdraw} will create a six or seven page  \monoi{postscript\_output\_file},\monoi{}
with each page displaying a different alignment statistic. These pages display
the alignment consensus sequence, information content per position, mutual
information per position, frequency of inserts per position, average length
of inserts per position, frequency of deletions (gaps) per position, and
average posterior probability per position (if posterior probabilites exist
in the alignment) If  \mono{-d } is enabled, all of these pages plus additional
ones, such as individual sequences (see discussion of  \mono{{-}{-}indi } below) will
be drawn. These pages can be selected to be drawn individually by using
 the command line options \mono{{-}{-}cons}, \mono{{-}{-}info}, \mono{{-}{-}mutinfo}, \mono{{-}{-}ifreq}, \mono{{-}{-}iavglen}, \mono{{-}{-}dall}, and
\mono{{-}{-}prob}. The calculation of the statistics for each of these options is discussed
below in the description for each option. Importantly, only so-called 'consensus'
positions of the alignment will be drawn. A consensus position is one that
is a nongap nucleotide in the 'reference' annotation of the Stockholm alignment
(\#=GC RF) read from \monoi{msafile}.  

By default, a consensus sequence for the input
alignment will be calculated and displayed on the alignment statistic diagrams.
The consensus sequence is defined as the most common nucleotide at each
 consensus position of the alignment. The consensus sequence will not be
displayed if the  \mono{{-}{-}no-cnt} option is used. The  \mono{{-}{-}cthresh}, \mono{{-}{-}cambig},\mono{} and  \mono{{-}{-}athresh}
options affect the definition of the consensus sequence as explained below
in the descriptions for those options.  

If the  \mono{{-}{-}tabfile}\monoi{ $<$f$>$} option is used,
a tab-delimited text file  \monoi{$<$f$>$} will be created that includes per-position lists
of the numerical values for each of the calculated statistics that were
drawn to  \monoi{postscript\_output\_file}. Comment lines in \monoi{$<$f$>$} are prefixed with a
'\#' character and explain the meaning of each of the tab-delimited columns
and how each of the statistics was calculated.  If  \mono{{-}{-}indi} is used, \mono{esl-ssdraw}
will create diagrams showing each sequence in the alignment on a separate
page, with aligned nucleotides in their corresponding position in the structure
diagram.  By default, basepaired nucleotides will be colored based on their
basepair type: either Watson-Crick (A:U, U:A, C:G, or G:C), G:U or U:G,
or non-canonical (the other ten possible basepairs). This coloring can be
turned off with the \mono{{-}{-}no-bp} option. Also by default, nucleotides that differ
from the most common nucleotide at each aligned consensus position will
be outlined. If the most common nucleotide occurs in more than 75\% of sequences
that do not have a gap at that position, the outline will be bold. Outlining
can be turned off with the  \mono{{-}{-}no-ol } option.  

With  \mono{{-}{-}indi}, if the alignment
contains posterior probability annotation (\#=GR PP), the  \monoi{postscript\_output\_file}
will contain an additional page for each sequence drawn with positions
colored by the posterior probability of each aligned nucleotide. No posterior
probability pages will be drawn if the  \mono{{-}{-}no-pp} option is used.   

\mono{esl-ssdraw}
can also be used to draw 'mask' diagrams which color positions of the structure
one of two colors depending on if they are included or excluded by a mask.
This is enabled with the  \mono{{-}{-}mask-col}\monoi{ $<$f$>$} option.  \monoi{$<$f$>$}  must contain a single line
of $<$rflen$>$ characters, where $<$rflen$>$ is the the number of nongap RF characters
in the alignment. The line must contain only '0' and '1' characters. A '0' at position
$<$x$>$ of the string indicates position $<$x$>$ is excluded from the mask, and a '1'
indicates position $<$x$>$ is included by the mask. A page comparing the overlap
of the  \monoi{$<$f$>$}  mask from  \mono{{-}{-}mask-col} and another mask in  \monoi{$<$f2$>$}  will be created
if the  \mono{{-}{-}mask-diff}\monoi{ $<$f2$>$}\mono{} option is used.  

If the  \mono{{-}{-}mask}\monoi{ $<$f$>$} option is used, positions
excluded by the mask in  \monoi{$<$f$>$} will be drawn differently (as open circles by
default) than positions included by the mask. The style of the masked positions
can be modified with the  \mono{{-}{-}mask-u}, \mono{{-}{-}mask-x}, and  \mono{{-}{-}mask-a} options.   

Finally, two
different types of input files can be used to customize output diagrams
using the \mono{{-}{-}dfile} and \mono{{-}{-}efile} options, as described below.    
\subsection*{Input}
 

The  \monoi{postscript\_template\_file}
is a postscript file that must be in a very specific format in order for
\mono{esl-ssdraw } to work. The specifics of the format, described below, are likely
to change in future versions of  \mono{esl-ssdraw}. The  \monoi{postscript\_output\_file}
files generated by  \mono{esl-ssdraw} will not be valid  \monoi{postscript\_template\_file}
format (i.e. an output file from  \mono{esl-ssdraw} cannot be used as an  \monoi{postscript\_template\_file}
in a subsequent run of the program).  

An example  \monoi{postscript\_template\_file}
('trna-ssdraw.ps') is included with the Easel distribution in the 'testsuite/'
subdirectory of the top-level 'easel' directory.  

The \monoi{postscript\_template\_file}
is a valid postscript file. It includes postscript commands for drawing
a secondary structure. The commands specify x and y coordinates for placing
each nucleotide on the page. The  \monoi{postscript\_template\_file} might also contain
commands for drawing lines connecting basepaired positions and tick marks
indicating every tenth position, though these are not required, as explained
below.   

If you are unfamiliar with the postscript language, it may be useful
for you to know that a postscript page is, by default, 612 points wide
and 792 points tall. The (0,0) coordinate of a postscript file is at the
bottom left corner of the page, (0,792) is the top left, (612,0) is the
bottom right, and (612,792) is the top right.  \mono{esl-ssdraw} uses 8 point by
8 point cells for drawing positions of the consensus secondary structure.
The 'scale' section of the \monoi{postscript\_template\_file} allows for different
'zoom levels', as described below. Also, it is important to know that postscript
lines beginning with '\%' are considered comments and do not include postscript
commands.  

An  \mono{esl-ssdraw} \monoi{postscript\_template\_file} contains n $>$= 1 pages,
each specifying a consensus secondary structure diagram. Each page is delimited
by a 'showpage' line in an 'ignore' section (as described below). \mono{esl-ssdraw}
will read all pages of the  \monoi{postscript\_template\_file} and then choose the
appropriate one that corresponds with the alignment in  \monoi{msafile}  based
on the consensus (nongap RF) length of the alignment.  For an alignment
of consensus length $<$rflen$>$, the first page of \monoi{postscript\_template\_file} that
has a structure diagram with consensus length $<$rflen$>$ will be used as the
template structure for the alignment.  

Each page of  \monoi{postscript\_template\_file}
contains blocks of text organized into seven different possible sections.
Each section must begin with a single line '\% begin $<$sectionname$>$' and end
with a single line '\% end $<$sectionname$>$' and have n $>$= 1 lines in between. On
the begin and end lines, there must be at least one space between the '\%'
and the 'begin' or 'end'. $<$sectionname$>$ must be one of the following: 'modelname',
'legend', 'scale', 'regurgitate', 'ignore', 'text positiontext', 'text nucleotides',
'lines positionticks', or 'lines bpconnects'. The n $>$=1 lines in between the
begin and end lines of each section must be in a specific format that differs
for each section as described below.  

Importantly, each page must end with
an 'ignore' section that includes a single line 'showpage' between the begin
and end lines. This lets  \mono{esl-ssdraw} know that a page has ended and another
might follow.  

Each page of a  \monoi{postscript\_template\_file} must include a single
'modelname' section. This section  must include exactly one line in between
its begin and end lines. This line must begin with a '\%' character followed
by a single space. The remainder of the line will be parsed as the model
name and will appear on each page of  \mono{postscript\_output\_file} in the header
section. If the name is more than 16 characters, it will be truncated in
the output.  

Each page of a  \monoi{postscript\_template\_file} must include a single
'legend' section.  This section must include exactly one line in between its
begin and end lines. This line must be formatted as '\% $<$d1$>$ $<$f1$>$ $<$f2$>$ $<$d2$>$ $<$f3$>$', where
$<$d1$>$ is an integer specifying the consensus position with relation to which
the legend will be placed; $<$f1$>$ and $<$f2$>$ specify the x and y axis offsets for
the top left corner of the legend relative to the x and y position of consensus
position $<$d1$>$; $<$d2$>$ specifies the size of a cell in the legend and $<$f3$>$ specifies
how many extra points should be between the right hand edge of the legend
and the end of the page. the offset of the right hand end of the legend
. For example, the line '\% 34 -40. -30. 12 0.' specfies that the legend be placed
40 points to the left and 30 points below the 34th consensus position,
that cells appearing in the legend be squares of size 12 points by 12 points,
and that the right hand side of the legend flush against the right hand
edge of the printable page.   

Each page of a  \monoi{postscript\_template\_file} must
include a single 'scale' section.  This section must include exactly one line
in between its begin and end lines. This line must be formatted as '$<$f1$>$ $<$f2$>$
scale', where $<$f1$>$ and $<$f2$>$ are both positive real numbers that are identical,
for example '1.7 1.7 scale' is valid, but '1.7 2.7 scale' is not. This line is a
valid postscript command which specifies the scale or zoom level on the
pages in the output. If $<$f1$>$ and $<$f2$>$ are '1.0' the default scale is used for which
the total size of the page is 612 points wide and 792 points tall. A scale
of 2.0 will reduce this to 306 points wide by 396 points tall. A scale of
0.5 will increase it to 1224 points wide by 1584 points tall. A single cell
corresponding to one position of the secondary structure is 8 points by
8 points. For larger RNAs, a scale of less than 1.0 is appropriate (for example,
SSU rRNA models (about 1500 nt) use a scale of about 0.6), and for smaller
RNAs, a scale of more than 1.0 might be desirable (tRNA (about 70 nt) uses
a scale of 1.7). The best way to determine the exact scale to use is trial
and error.  

Each page of a  \monoi{postscript\_template\_file} can include n $>$= 0 'regurgitate'
sections. These sections can include any number of lines.  The text in this
section will not be parsed by \mono{esl-ssdraw} but will be included in each page
of  \monoi{postscript\_output\_file.} The format of the lines in this section must
therefore be valid postscript commands. An example of content that might
be in a  regurgitate section are commands to draw lines and text annotating
the anticodon on a tRNA secondary structure diagram.  

Each page of a  \monoi{postscript\_template\_file}
must include at least 1 'ignore' section. One of these sections must include
a single line that reads 'showpage'. This section should be placed at the
end of each page of the template file.   Other ignore sections can include
any number of lines.  The text in these section will not be parsed by \mono{esl-ssdraw}
nor will it be included in each page of  \monoi{postscript\_output\_file}. An ignore
section can contain comments or postscript commands that draw features
of the \monoi{postscript\_template\_file} that are  unwanted in the  \monoi{postscript\_output\_file}.
 

Each page of a  \monoi{postscript\_template\_file} must include a single 'text nucleotides'
section. This section must include exactly $<$rflen$>$ lines, indicating that
the consensus secondary structure has exactly $<$rflen$>$ nucleotide positions.
Each line must be of the format '($<$c$>$) $<$x$>$ $<$y$>$ moveto show' where $<$c$>$ is a nucleotide
(this can be any character actually), and $<$x$>$ and $<$y$>$ are the coordinates specifying
the location of the nucleotide on the page, they should be positive real
numbers. The best way to determine what these coordinates should be is manually
by trial and error, by inspecting the resulting structure as you add each
nucleotide. Note that \mono{esl-ssdraw} will color an 8 point by 8 point cell for
each position, so nucleotides should be placed about 8 points apart from
each other.  

Each page of a  \monoi{postscript\_template\_file} may or may not include
a single 'text positiontext' section. This section can include n $>$= 1 lines,
each specifying text to be placed next to specific positions of the structure,
for example, to number them. Each line must be of the format '($<$s$>$) $<$x$>$ $<$y$>$ moveto
show' where $<$s$>$ is a string of text to place at coordinates ($<$x$>$,$<$y$>$) of the postscript
page.  Currently, the best way to determine what these coordinates is manually
by trial and error, by inspecting the resulting diagram as you add each
line.  

Each page of a  \monoi{postscript\_template\_file} may or may not include a
single 'lines positionticks' section. This section can include n $>$= 1 lines,
each specifying the location of a tick mark on the diagram. Each line must
be of the format '$<$x1$>$ $<$y1$>$ $<$x2$>$ $<$y2$>$ moveto show'. A tick mark (line of width 2.0)
will be drawn from point ($<$x1$>$,$<$y1$>$) to point ($<$x2$>$,$<$y2$>$) on each page of \monoi{postscript\_output\_file.}
Currently, the best way to determine what these coordinates should be is
manually by trial and error, by inspecting the resulting diagram as you
add each line.  

Each page of a  \monoi{postscript\_template\_file} may or may not
include a single 'lines bpconnects' section. This section must include $<$nbp$>$
lines, where $<$nbp$>$ is the number of basepairs in the consensus structure
of the input \monoi{msafile} annotated as \#=GC SS\_cons. Each line should connect
two basepaired positions in the consensus structure diagram. Each line must
be of the format '$<$x1$>$ $<$y1$>$ $<$x2$>$ $<$y2$>$ moveto show'. A line will be drawn from point
($<$x1$>$,$<$y1$>$) to point ($<$x2$>$,$<$y2$>$) on each page of \monoi{postscript\_output\_file.} Currently,
the best way to determine what these coordinates should be is manually
by trial and error, by inspecting the resulting diagram as you add each
line.     
\subsection*{Required Memory}
 

The memory required by  \mono{esl-ssdraw} will be equal
to roughly the larger of 2 Mb and  the size of the first alignment in \monoi{msafile}.
If the  \mono{{-}{-}small } option is used, the memory required will be independent
of the alignment size. To use  \mono{{-}{-}small} the alignment must be in Pfam format,
a non-interleaved (1 line/seq) version of Stockholm format.   If the  \mono{{-}{-}indi}
option is used, the required memory may exceed the size of the alignment
by up to ten-fold, and the output \mono{postscript\_output\_file } may be up to 50
times larger than the \mono{msafile.}  
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h } ] Print brief help;  includes version
number and summary of all options, including expert options.  
\item [\monob{-d } ] Draw the
default set of alignment summary diagrams: consensus sequence, information
content, mutual information, insert frequency, average insert length, deletion
frequency, and average posterior probability (if posterior probability
annotation exists in the alignment). These diagrams are also drawn by default
(if zero command line options are used), but using the \mono{-d } option allows
the user to add additional pages, such as individual aligned sequences
with \mono{{-}{-}indi}.  
\item [\monob{{-}{-}mask}\monoi{ $<$f$>$} ] Read the mask from file \monoi{$<$f$>$}, and draw positions differently
in  \monoi{postscript\_output\_file} depending on whether they are included or excluded
by the mask. \monoi{$<$f$>$} must contain a single line of length $<$rflen$>$ with only '0' and
'1' characters. $<$rflen$>$ is the number of nongap characters in the reference
(\#=GC RF) annotation of the first alignment in  \monoi{msafile} A '0' at position
$<$x$>$ of the mask indicates position $<$x$>$ is excluded by the mask, and a '1' indicates
that position $<$x$>$ is included by the mask.  
\item [\monob{{-}{-}small} ] Operate in memory saving
mode. Without \mono{{-}{-}indi}, required RAM will be independent of the size of the
alignment in  \monoi{msafile}. With \mono{{-}{-}indi},\mono{} the required RAM will be roughly ten times
the size of the alignment in  \monoi{msafile}. For  \mono{{-}{-}small} to work, the alignment
must be in Pfam Stockholm (non-interleaved 1 line/seq) format.  
\item [\monob{{-}{-}rf} ] Add a
page to  \monoi{postscript\_output\_file}  showing the reference sequence from the
\#=GC RF annotation in  \monoi{msafile.}  By default, basepaired nucleotides will
be colored based on what type of basepair they are. To turn this off, use
\mono{{-}{-}no-bp.} This page is drawn by default (if zero command-line options are used).
 
\item [\monob{{-}{-}info} ] Add a page to \monoi{postscript\_output\_file} with consensus (nongap RF) positions
colored based on their information content from the alignment.  Information
content is calculated as 2.0 - H, where H = sum\_x p\_x log\_2 p\_x for x in
\{A,C,G,U\}.  This page is drawn by default (if zero command-line options are
used).  
\item [\monob{{-}{-}mutinfo} ] Add a page to \monoi{postscript\_output\_file} with basepaired consensus
(nongap RF) positions colored based on the amount of mutual information
they have in the alignment. Mutual information is sum\_\{x,y\} p\_\{x,y\} log\_2
((p\_x * p\_y) / p\_\{x,y\}, where x and y are the four possible bases A,C,G,U.
p\_x is the fractions of aligned sequences that have nucleotide x of in
the left half (5' half) of the basepair. p\_y is the fraction of aligned sequences
that have nucleotide y in the position corresponding to the right half
(3' half) of the basepair. And p\_\{x,y\} is the fraction of aligned sequences
that  have basepair x:y. For all p\_x, p\_y and p\{x,y\} only sequences that
 that have a nongap nucleotide at both the left and right half of the basepair
are counted.  This page is drawn by default (if zero command-line options
are used).  
\item [\monob{{-}{-}ifreq} ] Add a page to \monoi{postscript\_output\_file} with each consensus
(nongap RF) position colored based on the fraction of sequences that span
each position that have at least 1 inserted nucleotide after the position.
 A sequence s spans consensus position x that is actual alignment position
a if s has at least one nongap nucleotide aligned to a position b $<$= a and
at least one nongap nucleotide aligned to a consensus position c $>$= a. This
page is drawn by default (if zero command-line options are used).  
\item [\monob{{-}{-}iavglen}
] Add a page to \monoi{postscript\_output\_file} with each consensus (nongap RF) position
colored based on average length of insertions that occur after it. The average
is calculated as the total number of inserted nucleotides after position
x, divided by the number of sequences that have at least 1 inserted nucleotide
after position x (so the minimum possible average insert length is 1.0).
 
\item [\monob{{-}{-}dall} ] Add a page to \monoi{postscript\_output\_file} with each consensus (nongap
RF) position colored based on the fraction of sequences that have a gap
(delete) at the position. This page is drawn by default (if zero command-line
options are used).  
\item [\monob{{-}{-}dint} ] Add a page to \monoi{postscript\_output\_file} with each
consensus (nongap RF) position colored based on the fraction of sequences
that have an internal gap (delete) at the position. An internal gap in a
sequence is one that occurs after (5' of) the sequence's first aligned nucleotide
and after (3' of) the sequence's final aligned nucleotide. This page is drawn
by default (if zero command-line options are used).  
\item [\monob{{-}{-}prob} ] Add a page to \monoi{postscript\_output\_file}
with positions colored based on average posterior probability (PP). The
alignment must contain \#=GR PP annotation for all sequences. PP annotation
is converted to numerical PP values as follows: '*' = 0.975, '9' = 0.90, '8' =
0.80, '7' = 0.70, '6' = 0.60, '5' = 0.50, '4' = 0.40, '3' = 0.30, '2' = 0.20, '1' = 0.10, '0' =
0.025. This page is drawn by default (if zero command-line options are used).
 
\item [\monob{{-}{-}span} ] Add a page to \monoi{postscript\_output\_file} with consensus (nongap RF) positions
colored based on the fraction of sequences that 'span' the position.  A sequence
s spans consensus position x that is actual alignment position a if s has
at least one nongap nucleotide aligned to a position b $<$= a and at least
one nongap nucleotide aligned to a consensus position c $>$= a. This page is
drawn by default (if zero command-line options are used).   
\end{wideitem}

\subsection*{Options for Drawing
Individual Aligned Sequences}
 \begin{wideitem}
\item [\monob{{-}{-}indi} ] Add a page displaying the aligned nucleotides
in their corresponding consensus positions of the structure diagram for
each aligned sequence in the alignment.  By default, basepaired nucleotides
will be colored based on what type of basepair they are. To turn this off,
use \mono{{-}{-}no-bp.} If posterior probability information (\#=GR PP) exists in the alignment,
one additional page per sequence will be drawn displaying the posterior
probabilities.  
\item [\monob{-f} ] With  \mono{{-}{-}indi}, force  \mono{esl-ssdraw} to create a diagram, even
if it is predicted to be large ($>$ 100 Mb). By default, if the predicted size
exceeds 100 Mb,  \mono{esl-ssdraw} will fail with a warning.    
\end{wideitem}

\subsection*{Options for Omitting
Parts of the Diagrams}
 \begin{wideitem}
\item [\monob{{-}{-}no-leg} ] Omit the legend on all pages of  \monoi{postscript\_output\_file}.
 
\item [\monob{{-}{-}no-head} ] Omit the header on all pages of  \monoi{postscript\_output\_file}.  
\item [\monob{{-}{-}no-foot}
] Omit the footer on all pages of  \monoi{postscript\_output\_file}.    
\end{wideitem}

\subsection*{Options for
Simple Two-color Mask Diagrams}
 \begin{wideitem}
\item [\monob{{-}{-}mask-col } ] With \mono{{-}{-}mask}, \monoi{postscript\_output\_file}
will contain exactly 1 page showing positions included by the mask as 
black squares, and positions excluded as pink squares.  
\item [\monob{{-}{-}mask-diff}\monoi{ $<$f$>$} ] With
\mono{{-}{-}mask}\monoi{ $<$f2$>$} and \mono{mask-col}, \monoi{postscript\_output\_file} will contain one additional
page comparing the mask from  \monoi{$<$f$>$} and the mask from \monoi{$<$f2$>$}. Positions will be
colored based on whether they are included by one mask and not the other,
excluded by both masks, and included by both masks.   
\end{wideitem}

\subsection*{Expert Options for
Controlling Individual Sequence Diagrams}
 \begin{wideitem}
\item [\monob{{-}{-}no-pp} ] When used in combination
with  \mono{{-}{-}indi}, do not draw posterior probability structure diagrams for each
sequence, even if the alignment has PP annotation.  
\item [\monob{{-}{-}no-bp} ] Do not color basepaired
nucleotides based on their basepair type.  
\item [\monob{{-}{-}no-ol} ] When used in combination
with  \mono{{-}{-}indi}, do not outline nucleotides that differ from the majority rule
consensus nucleotide given the alignment.  
\item [\monob{{-}{-}no-ntpp} ] When used in combination
with  \mono{{-}{-}indi}, do not draw nucleotides on the individual sequence posterior
probability diagrams.   
\end{wideitem}

\subsection*{Expert Options Related to Consensus Sequence Definition}

\begin{wideitem}
\item [\monob{{-}{-}no-cnt} ] Do not draw consensus nucleotides on alignment statistic diagrams
(such as information content diagrams). By default, the consensus nucleotide
is defined as the most frequent nucleotide in the alignment at the corresponding
position. Consensus nucleotides that occur in at least \monoi{$<$x$>$} fraction of the
aligned sequences (that do not contain a gap at the position) are capitalized.
By default  \monoi{$<$x$>$} is 0.75, but can be changed with the  \mono{{-}{-}cthresh}\monoi{ $<$x$>$} option.   
\item [\monob{{-}{-}cthresh}\monoi{
$<$x$>$} ] Specify the threshold for capitalizing consensus nucleotides defined
by the majority rule (i.e. when  \mono{{-}{-}cambig} is not enabled) as  \monoi{$<$x$>$}.  
\item [\monob{{-}{-}cambig} ] Change
how consensus nucleotides are calculated from majority rule to the least
ambiguous IUPAC nucleotide that represents at least \monoi{$<$x$>$} fraction of the nongap
nucleotides at each consensus position.  By default  \monoi{$<$x$>$} is 0.9, but can be
changed with the  \mono{{-}{-}athresh}\monoi{ $<$x$>$} option.   
\item [\monob{{-}{-}athresh}\monoi{ $<$x$>$} ] With \mono{{-}{-}cambig}, specify the
threshold for defining consensus nucleotides is the least ambiguous IUPAC
nucleotide that represents at least \monoi{$<$x$>$} fraction of the nongap nucleotides
at each position.   
\end{wideitem}

\subsection*{Expert Options Controlling Style of Masking Positions}

\begin{wideitem}
\item [\monob{{-}{-}mask-u} ] With  \mono{{-}{-}mask}, change the style of masked columns to squares.  
\item [\monob{{-}{-}mask-x}
] With  \mono{{-}{-}mask},\mono{} change the style of masked columns to x's.  
\item [\monob{{-}{-}mask-a} ] With  \mono{{-}{-}mask}
and \mono{{-}{-}mask-u} or \mono{{-}{-}mask-x} draw the alternative style of square or 'x' masks.   
\end{wideitem}

\subsection*{Expert
Options Related to Input Files}
 \begin{wideitem}
\item [\monob{{-}{-}dfile}\monoi{ $<$f$>$} ] Read the 'draw file' \monoi{$<$f$>$} which specifies
numerical values for each consensus position in one or more postscript
pages.  For each page, the draw file must include $<$rflen$>$+3 lines ($<$rflen$>$ is
defined in the DESCRIPTION section). The first three lines are special. The
following $<$rflen$>$ 'value lines' each must contain a single number, the numerical
value for the corresponding position.  The first of the three special lines
defines the 'description' for the page. This should be text that describes
what the numerical values refer to for the page. The maximum allowable length
is roughly 50 characters (the exact maximum length depends on the template
file and the program will report an informative error message upon execution
if it is exceeded). The second special line defines the 'legend header' line
that which will appear immediately above the legend. It has a maximum allowable
length of about 30 characters.  The third special line per page must contain
exactly 7 numbers, which must be in increasing order, each separated by
a space.  These numbers define the numerical ranges for the six different
colors used to draw the consensus positions on the page.  The first number
defines the minimum value for the first color (blue) and must be less than
or equal to the minimum value from the value lines. The second number defines
the minimum value for the second color (turquoise). The third, fourth, fifth
and sixth numbers define the minimum values for the third, fourth, fifth
and sixth colors (light green, yellow, orange, red), and the seventh final
number defines the maximum value for red and must be equal to or greater
than the maximum value from the value lines.  After the $<$rflen$>$ value lines,
there must exist a special line with only '//', signifying the end of a page.
The draw file  \monoi{$<$f$>$} must end with this special '//' line, even if it only includes
a single page. A draw file specifying $<$n$>$ pages should include exactly $<$n$>$ *
($<$rflen$>$ + 4) lines.  
\item [\monob{{-}{-}efile}\monoi{ $<$f$>$} ] Read the 'expert draw file' \monoi{$<$f$>$} which specifies
the colors and nucleotides to draw on each consensus position in one or
more postscript pages. Unlike with the  \mono{{-}{-}dfile} option, no legend will be
drawn when \mono{{-}{-}efile } is used. For each page, the draw file must include $<$rflen$>$
lines, each with four or five tab-delimited tokens. The first four tokens
on line $<$x$>$ specify the color to paint position $<$x$>$ and must be real numbers
between 0 and 1. The four numbers specify the cyan, magenta, yellow and
black values, respectively, in the CMYK color scheme for the postscript
file. The fifth token on line $<$x$>$ specifies which nucleotide to write on position
$<$x$>$ (on top of the colored background). If the fifth token does not exist,
no nucleotide will be written.  After the $<$rflen$>$ lines, there must exist
a special line with only '//', signifying the end of a page. The expert draw
file  \monoi{$<$f$>$} must end with this special '//' line, even if it only includes a
single page. A expert draw file specifying $<$n$>$ pages should include exactly
$<$n$>$ * ($<$rflen$>$ + 1) lines.  
\item [\monob{{-}{-}ifile}\monoi{ $<$f$>$} ] Read insert information from the file \monoi{$<$f$>$},
which may have been created with INFERNAL's \textsf{\mono{cmalign}(1)} program. The insert
information in  \monoi{msafile} will be ignored and the information from \monoi{$<$f$>$} will
supersede it. Inserts are columns that are gaps in the reference (\#=GC RF)
annotation.      
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-translate}}{esl-translate} - translate DNA sequence in six frames into individual}
ORFs  
\subsection*{Synopsis}
\noindent
\monob{esl-translate} [\monoi{options}] \monoi{seqfile}   
\subsection*{Description}
 

Given a  \monoi{seqfile}
 containing DNA or RNA sequences, \mono{esl-translate} outputs a six-frame translation
of them as individual open reading frames in FASTA format.  

By default,
only open reading frames greater than 20aa are reported.  This minimum ORF
length can be changed with the \mono{-l } option.  

By default, no specific initiation
codon is required, and any amino acid can start an open reading frame. This
is so \mono{esl-translate} may be used on sequence fragments, eukaryotic genes
with introns, or other cases where we do not want to assume that ORFs are
complete coding regions. This behavior can be changed. With the  \mono{-m } option,
ORFs start with an initiator AUG Met. With the \mono{-M} option, ORFs start with
any of the initiation codons allowed by the genetic code. For example, the
"standard" code (NCBI transl\_table 1)  allows AUG, CUG, and UUG as initiators.
When \mono{-m} or \mono{-M} are used, an initiator is always translated to Met (even if
the initiator is something like UUG or CUG that doesn't encode Met as an
elongator).  

If \monoi{seqfile} is - (a single dash), input is read from the stdin
pipe. This (combined with the output being a standard FASTA file) allows
\mono{esl-translate } to be used in command line incantations. If \monoi{seqfile} ends in
.gz, it is assumed to be a gzip-compressed file, and  Easel will try to read
it as a stream from \mono{gunzip -c}.    
\subsection*{Output Format}
 

The output FASTA name/description
line contains information about the source and coordinates of each ORF.
Each ORF is named  \mono{orf1,} etc., with numbering starting from 1, in order
of their start position on the top strand followed by the bottom strand.
 The rest of the FASTA name/desc line contains 4 additional fields, followed
by the description of the source sequence:  \begin{wideitem}
\item [\monob{source}=\monoi{$<$s$>$} ] \monoi{$<$s$>$} is the name of the
source DNA/RNA sequence.  
\item [\monob{coords}=\monoi{start}..\monoi{end} ] Coords, 1..L, for the translated
ORF in a source DNA sequence of length L. If start is greater than end,
the ORF is on the bottom (reverse complement) strand. The start is the first
nucleotide of the first codon; the end is the last nucleotide of the last
codon. The stop codon is not included in the coordinates (unlike in CDS
annotation in GenBank, for example.)  
\item [\monob{length}=\monoi{$<$n$>$} ] Length of the ORF in amino
acids.  
\item [\monob{frame}=\monoi{$<$n$>$} ] Which frame the ORF is in. Frames 1..3 are the top strand;
4..6 are the bottom strand. Frame 1 starts at nucleotide 1. Frame 4 starts
at nucleotide L.    
\end{wideitem}

\subsection*{Alternative Genetic Codes}
 

By default, the "standard"
genetic code is used (NCBI transl\_table 1).  Any NCBI genetic code transl\_table
can be selected with the \mono{-c } option, as follows:  \begin{wideitem}
\item [\monob{1 } ] Standard 
\item [\monob{2 } ] Vertebrate
mitochondrial 
\item [\monob{3} ] Yeast mitochondrial 
\item [\monob{4 } ] Mold, protozoan, coelenterate mitochondrial;
Mycoplasma/Spiroplasma 
\item [\monob{5 } ] Invertebrate mitochondrial 
\item [\monob{6 } ] Ciliate, dasycladacean,
Hexamita nuclear 
\item [\monob{9 } ] Echinoderm and flatworm mitochondrial 
\item [\monob{10 } ] Euplotid
nuclear 
\item [\monob{11} ] Bacterial, archaeal; and plant plastid 
\item [\monob{12 } ] Alternative yeast

\item [\monob{13 } ] Ascidian mitochondrial 
\item [\monob{14 } ] Alternative flatworm mitochondrial 
\item [\monob{16 } ] Chlorophycean
mitochondrial 
\item [\monob{21 } ] Trematode mitochondrial 
\item [\monob{22 } ] Scenedesmus obliquus mitochondrial

\item [\monob{23 } ] Thraustochytrium mitochondrial 
\item [\monob{24 } ] Pterobranchia mitochondrial 
\item [\monob{25 }
] Candidate Division SR1 and Gracilibacteria   
\end{wideitem}


As of this writing, more information
about the genetic codes in the NCBI translation tables is at  \monoi{http://www.ncbi.nlm.nih.gov/Taxonomy/}
 at a link titled \monoi{Genetic} codes.  
\subsection*{Iupac Degeneracy Codes in Dna}
 

DNA sequences
may contain IUPAC degeneracy codes, such as N, R, Y, etc. If all codons
consistent with a degenerate codon translate to the same amino acid (or
to a stop), that translation is done; otherwise, the codon is translated
as X (even if one or more compatible codons are stops). For example, in
the standard code, UAR translates to * (stop), GGN translates to G (glycine),
NNN translates to X, and UGR translates to X (it could be either a UGA
stop or a UGG Trp).  

Degenerate initiation codons are handled essentially
the same. If all codons consistent with the degenerate codon are legal initiators,
then the codon is allowed to initiate a new ORF. Stop codons are never a
legal initiator (not only with  \mono{-m } or \mono{-M} but also with the default of allowing
any amino acid to initiate), so degenerate codons consistent with a stop
cannot be initiators. For example, NNN cannot initiate an ORF, nor can UGR
{-}{-} even though they translate to X. This means that we don't translate long
stretches of N's as long ORFs of X's, which is probably a feature, given
the prevalence of artificial runs of N's in genome  sequence assemblies.
 

Degenerate DNA codons are not translated to degenerate amino acids other
than X, even when that is possible. For example, SAR and MUH are decoded
as X, not Z (Q$|$E) and J (I$|$L). The extra complexity needed for a degenerate
to degenerate translation doesn't seem worthwhile.   
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Print brief
help. Includes version number and summary of all options.  Also includes
a list of the available NCBI transl\_tables and their numerical codes, for
the \mono{-c } option.  
\item [\monob{-c}\monoi{ $<$id$>$} ] Choose alternative genetic code  \monoi{$<$id$>$} where  \monoi{$<$id$>$} is the
numerical code of one of the NCBI transl\_tables.  
\item [\monob{-l}\monoi{ $<$n$>$} ] Set the minimum reported
ORF length to  \monoi{$<$n$>$} aa.  
\item [\monob{-m} ] Require ORFs to start with an initiator codon AUG
(Met).  
\item [\monob{-M} ] Require ORFs to start with an initiator codon, as specified by
the allowed initiator codons in the NCBI transl\_table. In the default Standard
code, AUG, CUG, and UUG are allowed as initiators. An  initiation codon
is always translated as Met, even if it does not normally encode Met as
an elongator.  
\item [\monob{-W} ] Use a memory-efficient windowed sequence reader. The default
is to read entire DNA sequences into memory, which may become memory limited
for some very large eukaryotic chromosomes. The windowed reader cannot 
reverse complement a nonrewindable input stream, so  either \monoi{seqfile} must
be a file, or you must use \monoi{{-}{-}watson} to limit translation to the top strand.
  
\item [\monob{{-}{-}informat}\monoi{ $<$s$>$} ] Assert that input \monoi{seqfile} is in format \monoi{$<$s$>$}, bypassing format
autodetection. Common choices for  \monoi{$<$s$>$}  include: \mono{fasta}, \mono{embl}, \mono{genbank.} Alignment
formats also work; common choices include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast},
\mono{clustal}, \mono{phylip}. For more information, and for codes for some less common
formats, see main documentation. The string \monoi{$<$s$>$} is case-insensitive (\mono{fasta}
or \mono{FASTA} both work).   
\item [\monob{{-}{-}watson} ] Only translate the top strand.  
\item [\monob{{-}{-}crick} ] Only
translate the bottom strand.      
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{esl-weight}}{esl-weight} - calculate sequence weights in MSA(s)  }
\subsection*{Synopsis}
\noindent
\monob{esl-weight}
[\monoi{options}] \monoi{msafile}  
\subsection*{Description}
 

\mono{esl-weight} calculates individual sequence
weights for each alignment in  \monoi{msafile}  and outputs a new  multiple sequence
alignment file in Stockholm format with the weights annotated in Stockholm-format
 \mono{\#=GS }\monoi{seqname}\mono{ WT }\monoi{weight} lines. The default weighting algorithm is the Gerstein/Sonnhammer/Chothia
algorithm.  

If  \monoi{msafile} is - (a single dash), MSA input is read from stdin.
    
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h } ] Print brief help;  includes version number and summary of
all options, including expert options.  
\item [\monob{-g} ] Use the Gerstein/Sonnhammer/Chothia
weighting algorithm; this is the default.  
\item [\monob{-p} ] Use the Henikoff position-based
weighting algorithm. This is faster and more memory efficient than the default.
 
\item [\monob{-b} ] "BLOSUM weights": use approximately the same rule used in constructing
the BLOSUM score matrices. This involves single-linkage clustering at some
fractional identity threshold (default 0.62; see  \mono{{-}{-}id } option), then for
each cluster, splitting a total weight of one uniformly amongst all sequences
in the cluster.   
\end{wideitem}

\subsection*{Expert Options}
 \begin{wideitem}
\item [\monob{{-}{-}id}\monoi{ $<$x$>$} ] Sets the fractional identity threshold
used by the BLOSUM weighting rule (option  \mono{-b}; required), to a number 0$<$=x$<$=1.
Default is 0.62.  
\item [\monob{{-}{-}amino} ] Assert that the  \monoi{msafile}  contains protein sequences.
  
\item [\monob{{-}{-}dna} ] Assert that the  \monoi{msafile}  contains DNA sequences.   
\item [\monob{{-}{-}rna} ] Assert that
the  \monoi{msafile}  contains RNA sequences.   
\end{wideitem}

\newpage
