%% Manual pages chapter automatically generated. Do not edit.
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmclient}}{hmmclient} - submit searches to a server for execution   }
\subsection*{Synopsis}
\noindent
\monob{hmmclient}
[\monoi{options}] \monoi{query}  \monoi{hmmfile}  \monoi{or}  \monoi{seqfile}   
\subsection*{Description}
 

\mono{Hmmclient} is a command-line
interface to submit searches to a running \mono{hmmserver} It is designed to mimic
the behavior and output formats of the  \mono{hmmsearch} , \mono{phmmer} ,  \mono{hmmscan} ,
and \mono{jackhmmer } programs as closely as possible.    

A typical usage of hmmclient
is " \mono{hmmclient} -s $<$ \monoi{name} \monoi{of} \monoi{the} \monoi{machine} \monoi{running} \monoi{the} \monoi{server's} \monoi{master} \monoi{node} $>$
{-}{-}db $<$  \monoi{number}  \monoi{of}  \monoi{the}  \monoi{database}  \monoi{to}  \monoi{be}  \monoi{searched} $>$ $<$ \monoi{filename}  \monoi{of}  \monoi{the}  \monoi{query}
 \monoi{sequence}  \monoi{or}  \monoi{hmm} $>$,  which would cause  \mono{hmmclient } to send the specified
query to the server, wait for results to come back, and then display them
in the same manner as our command-line programs.  \mono{Hmmclient } accepts the
same set of options to control search parameters and format its output
as our command-line programs, which are described below.  

The query object
may be read from standard input.  To instruct \mono{hmmclient} to do this, provide
a dash ("-") as the last input to  \mono{hmmclient} instead of a file name.   
\subsection*{Options}

\begin{wideitem}
\item [\monob{-h} ] Help; print a brief reminder of command line usage and all available
options.  
\end{wideitem}

\subsection*{Options that Control the Connection to the Server}
 \begin{wideitem}
\item [\monob{-s}\monoi{ $<$servername$>$}
] Specifies the name of the machine running the server that  \mono{hmmclient } should
send the search to.  
\item [\monob{{-}{-}cport}\monoi{ $<$portnumber$>$}\mono{} ] Specifies the port on the server that
 \mono{hmmclient } should connect to.  Defaults to 51371.  
\item [\monob{{-}{-}db}\monoi{ $<$n$>$} ] Specifies the number
of the database to be searched, defaults to one.  
\item [\monob{{-}{-}db\_ranges}\monoi{ $<$rangelist$>$} ] Instructs
the server to search a subset of the items in the database instead of the
entire database (default).   \monoi{rangelist} must be a list of one or more numerical
ranges separated by commas, i.e.:  start1..end1,start2..end2,etc.  
\item [\monob{{-}{-}jack}\monoi{ $<$maxrounds$>$}
] Instructs the server to perform an iterative, jackhmmer-style search, with
at most \monoi{maxrounds} of iterative search.  If this option is specified, both
the query object  and the target database must be sequences, not HMMs,
or an error will occur.  
\item [\monob{{-}{-}shutdown} ] Send a shutdown command to the server
instead of a search request.  
\item [\monob{{-}{-}password}\monoi{ $<$password$>$} ] Specifies a password to
be sent along with the shutdown command.  May only be used if  \mono{{-}{-}shutdown
} is also used.     
\end{wideitem}

\subsection*{Options for Controlling Output}
 \begin{wideitem}
\item [\monob{-o}\monoi{ $<$f$>$} ] Direct the main human-readable
output to a file \monoi{$<$f$>$}  instead of the default stdout.  
\item [\monob{-A}\monoi{ $<$f$>$} ] Save a multiple
alignment of all significant hits (those satisfying \monoi{inclusion thresholds})
to the file  \monoi{$<$f$>$}.  
\item [\monob{{-}{-}tblout}\monoi{ $<$f$>$} ] Save a simple tabular (space-delimited) file summarizing
the per-target output, with one data line per homologous target sequence
found.  
\item [\monob{{-}{-}domtblout}\monoi{ $<$f$>$} ] Save a simple tabular (space-delimited) file summarizing
the per-domain output, with one data line per homologous domain detected
in a query sequence for each homologous model.  
\item [\monob{{-}{-}pfamtblout}\monoi{ $<$f$>$} ] Save a table
of hits and domains to the specified file in Pfam format.  
\item [\monob{{-}{-}chkhmm}\monoi{ prefix}
] When doing an iterative search, at the start of each iteration, checkpoint
the query HMM, saving it to a file named \monoi{prefix}\mono{-}\monoi{n}\mono{.hmm} where \monoi{n} is the iteration
number (from 1..N). Is only valid when {-}{-}jack is specfied.  
\item [\monob{{-}{-}chkali}\monoi{ prefix} ] When
doing an iterative search, at the end of each iteration, checkpoint an
alignment of all domains satisfying inclusion thresholds (e.g. what will
become the query HMM for the next iteration),  saving it to a file named
\monoi{prefix}\mono{-}\monoi{n}\mono{.sto} in Stockholm format, where \monoi{n} is the iteration number (from 1..N).
Is only valid when {-}{-}jack is specfied.  
\item [\monob{{-}{-}acc} ] Use accessions instead of names
in the main output, where available for profiles and/or sequences.  
\item [\monob{{-}{-}noali}
] Omit the alignment section from the main output. This can greatly reduce
the output volume.  
\item [\monob{{-}{-}notextw} ] Unlimit the length of each line in the main
output. The default is a limit of 120 characters per line, which helps in
displaying the output cleanly on terminals and in editors, but can truncate
target profile description lines.  
\item [\monob{{-}{-}textw}\monoi{ $<$n$>$} ] Set the main output's line length
limit to \monoi{$<$n$>$} characters per line. The default is 120.  
\end{wideitem}

\subsection*{Options Controlling
Single Sequence Scoring}
 These options control how an HMM is generated from
a single sequence, either when both the query object and target database
are sequences, or in the first round of an iterative search.  \begin{wideitem}
\item [\monob{{-}{-}popen}\monoi{ $<$x$>$} ] Set
the gap open probability for a single sequence query model to  \monoi{$<$x$>$}. The default
is 0.02.  \monoi{$<$x$>$}  must be $>$= 0 and $<$ 0.5.  
\item [\monob{{-}{-}pextend}\monoi{ $<$x$>$} ] Set the gap extend probability
for a single sequence query model to  \monoi{$<$x$>$}. The default is 0.4.  \monoi{$<$x$>$}  must be $>$=
0 and $<$ 1.0.  
\item [\monob{{-}{-}mx}\monoi{ $<$s$>$} ] Obtain residue alignment probabilities from the built-in
substitution matrix named \monoi{$<$s$>$}.\monoi{} Several standard matrices are built-in, and
do not need to be read from files.  The matrix name \monoi{$<$s$>$}  can be PAM30, PAM70,
PAM120, PAM240, BLOSUM45, BLOSUM50, BLOSUM62, BLOSUM80, or BLOSUM90. Only
one of the \mono{{-}{-}mx } and \mono{{-}{-}mxfile} options may be used.  
\item [\monob{{-}{-}mxfile}\monoi{ mxfile} ] Obtain residue
alignment probabilities from the substitution matrix in file \monoi{mxfile} on
the machine running the server program.   The default score matrix is BLOSUM62
(this matrix is internal to HMMER and does not have to be available as
a file).  The format of a substitution matrix \monoi{mxfile} is the standard format
accepted by BLAST, FASTA, and other sequence  analysis software. See \mono{ftp.ncbi.nlm.nih.gov/blast/matrices/}
for example files. (The only exception: we require matrices to be square,
so for DNA, use files like NCBI's NUC.4.4, not NUC.4.2.)  
\end{wideitem}

\subsection*{Options Controlling
Reporting Thresholds}
 Reporting thresholds control which hits are reported
in output files (the main output, \mono{{-}{-}tblout}, and  \mono{{-}{-}domtblout}). Sequence hits
and domain hits are ranked by statistical significance (E-value) and output
is generated in two sections called per-target and per-domain output. In per-target
output, by default, all sequence hits with an E-value $<$= 10 are reported.
In the per-domain output, for each target that has passed per-target reporting
thresholds, all domains satisfying per-domain reporting thresholds are reported.
By default, these are domains with conditional E-values of $<$= 10. The following
options allow you to change the default E-value reporting thresholds, or
to use bit score thresholds instead.   \begin{wideitem}
\item [\monob{-E}\monoi{ $<$x$>$} ] In the per-target output, report
target sequences with an E-value of $<$= \monoi{$<$x$>$}.\monoi{} The default is 10.0, meaning that
on average, about 10 false positives will be reported per query, so you
can see the top of the noise and decide for yourself if it's really noise.
 
\item [\monob{-T}\monoi{ $<$x$>$} ] Instead of thresholding per-profile output on E-value, instead report
target sequences with a bit score of $>$= \monoi{$<$x$>$}.  
\item [\monob{{-}{-}domE}\monoi{ $<$x$>$} ] In the per-domain output,
for target sequences that have already satisfied the per-profile reporting
threshold, report individual domains with a conditional E-value of $<$= \monoi{$<$x$>$}.\monoi{} The
default is 10.0.  A conditional E-value means the expected number of additional
false positive domains in the smaller search space of those comparisons
that already satisfied the per-target reporting threshold (and thus must
have at least one homologous domain already).   
\item [\monob{{-}{-}domT}\monoi{ $<$x$>$} ] Instead of thresholding
per-domain output on E-value, instead report domains with a bit score of
$>$= \monoi{$<$x$>$}.     
\end{wideitem}

\subsection*{Options for Inclusion Thresholds}
 Inclusion thresholds are stricter
than reporting thresholds. Inclusion thresholds control which hits are considered
to be reliable enough to be included in an output alignment or a subsequent
search round, or marked as significant ("!") as opposed to questionable
("?") in domain output.  \begin{wideitem}
\item [\monob{{-}{-}incE}\monoi{ $<$x$>$} ] Use an E-value of $<$= \monoi{$<$x$>$} as the per-target inclusion
threshold. The default is 0.01, meaning that on average, about 1 false positive
would be expected in every 100 searches with different query sequences.
 
\item [\monob{{-}{-}incT}\monoi{ $<$x$>$} ] Instead of using E-values for setting the inclusion threshold, instead
use a bit score of $>$=  \monoi{$<$x$>$} as the per-target inclusion threshold. By default
this option is unset.  
\item [\monob{{-}{-}incdomE}\monoi{ $<$x$>$} ] Use a conditional E-value of $<$= \monoi{$<$x$>$}  as the
per-domain inclusion threshold, in targets that have already satisfied the
overall per-target inclusion threshold. The default is 0.01.  
\item [\monob{{-}{-}incdomT}\monoi{ $<$x$>$} ] Instead
of using E-values, use a bit score of $>$= \monoi{$<$x$>$} as the per-domain inclusion threshold.
   
\end{wideitem}

\subsection*{Options for Model-specific Score Thresholding}
 Curated profile databases
may define specific bit score thresholds for each profile, superseding
any thresholding based on statistical significance alone.  To use these
options, the profile must contain the appropriate (GA, TC, and/or NC) optional
score threshold annotation; this is picked up by  \mono{hmmbuild} from Stockholm
format alignment files. Each thresholding option has two scores: the per-sequence
threshold $<$x1$>$ and the per-domain threshold $<$x2$>$ These act as if \mono{-T}\monoi{ $<$x1$>$} \mono{{-}{-}incT}\monoi{ $<$x1$>$}
\mono{{-}{-}domT}\monoi{ $<$x2$>$} \mono{{-}{-}incdomT}\monoi{ $<$x2$>$} has been applied specifically using each model's curated
thresholds.  \begin{wideitem}
\item [\monob{{-}{-}cut\_ga} ] Use the GA (gathering) bit scores in the model to set
per-sequence (GA1) and per-domain (GA2) reporting and inclusion thresholds.
GA thresholds are generally considered to be the reliable curated thresholds
defining family membership; for example, in Pfam, these thresholds define
what gets included in Pfam Full alignments based on searches with Pfam
Seed models.  
\item [\monob{{-}{-}cut\_nc} ] Use the NC (noise cutoff) bit score thresholds in the
model to set per-sequence (NC1) and per-domain (NC2) reporting and inclusion
thresholds. NC thresholds are generally considered to be the score of the
highest-scoring known false positive.  
\item [\monob{{-}{-}cut\_tc} ] Use the TC (trusted cutoff)
bit score thresholds in the model to set per-sequence (TC1) and per-domain
(TC2) reporting and inclusion thresholds. TC thresholds are generally considered
to be the score of the lowest-scoring known true positive that is above
all known false positives.      
\end{wideitem}

\subsection*{Options Controlling the Acceleration Pipeline}

HMMER3 searches are accelerated in a three-step filter pipeline: the MSV
filter, the Viterbi filter, and the Forward filter. The first filter is
the fastest and most approximate; the last is the full Forward scoring
algorithm. There is also a bias filter step between MSV and Viterbi. Targets
that pass all the steps in the acceleration pipeline are then subjected
to postprocessing {-}{-} domain identification and scoring using the Forward/Backward
algorithm.  Changing filter thresholds only removes or includes targets
from consideration; changing filter thresholds does not alter bit scores,
E-values, or alignments, all of which are determined solely in postprocessing.
 \begin{wideitem}
\item [\monob{{-}{-}max} ] Turn off all filters, including the bias filter, and run full Forward/Backward
postprocessing on every target. This increases sensitivity somewhat, at
a large cost in speed.  
\item [\monob{{-}{-}F1}\monoi{ $<$x$>$} ] Set the P-value threshold for the MSV filter
step.  The default is 0.02, meaning that roughly 2\% of the highest scoring
nonhomologous targets are expected to pass the filter.  
\item [\monob{{-}{-}F2}\monoi{ $<$x$>$} ] Set the P-value
threshold for the Viterbi filter step. The default is 0.001.   
\item [\monob{{-}{-}F3}\monoi{ $<$x$>$} ] Set the
P-value threshold for the Forward filter step. The default is 1e-5.  
\item [\monob{{-}{-}nobias}
] Turn off the bias filter. This increases sensitivity somewhat, but can come
at a high cost in speed, especially if the query has biased residue composition
(such as a repetitive sequence region, or if it is a membrane protein with
large regions of hydrophobicity). Without the bias filter, too many sequences
may pass the filter with biased queries, leading to slower than expected
performance as the computationally intensive Forward/Backward algorithms
shoulder an abnormally heavy load.    
\end{wideitem}

\subsection*{Options Controlling Profile Construction}
This
option is only valid when performing an iterative search. \begin{wideitem}
\item [\monob{{-}{-}fragthresh}\monoi{ $<$x$>$} ] We
only want to count terminal gaps as deletions if the aligned sequence is
known to be full-length, not if it is a fragment (for instance, because
only part of it was sequenced). HMMER uses a simple rule to infer fragments:
if the sequence length L is less than  or equal to a fraction \monoi{$<$x$>$}  times
the alignment length in columns, then the sequence is handled as a fragment.
The default is 0.5. Setting \mono{{-}{-}fragthresh 0} will define no (nonempty) sequence
as a fragment; you might want to do this if you know you've got a carefully
curated alignment of full-length sequences. Setting \mono{{-}{-}fragthresh 1} will define
all sequences as fragments; you might want to do this if you know your
alignment is entirely composed of fragments, such as translated short reads
in metagenomic shotgun data.  
\end{wideitem}

\subsection*{Options Controlling Relative Weights}
These
options are only valid when performing an iterative search.  Whenever a
profile is built from a multiple alignment, HMMER uses an ad hoc sequence
weighting algorithm to downweight closely related sequences and upweight
distantly related ones. This has the effect of making models less biased
by uneven phylogenetic representation.  These options control which algorithm
gets used.  \begin{wideitem}
\item [\monob{{-}{-}wpb} ] Use the Henikoff position-based sequence weighting scheme
[Henikoff and Henikoff, J. Mol. Biol. 243:574, 1994].  This is the default.
 
\item [\monob{{-}{-}wgsc } ] Use the Gerstein/Sonnhammer/Chothia weighting algorithm [Gerstein
et al, J. Mol. Biol. 235:1067, 1994].  
\item [\monob{{-}{-}wblosum} ] Use the same clustering scheme
that was used to weight data in calculating BLOSUM substitution matrices
[Henikoff and Henikoff, Proc. Natl. Acad. Sci 89:10915, 1992]. Sequences are
single-linkage clustered at an identity threshold (default 0.62; see \mono{{-}{-}wid})
and within each cluster of c sequences, each sequence gets relative weight
1/c.  
\item [\monob{{-}{-}wnone} ] No relative weights. All sequences are assigned uniform weight.
  
\item [\monob{{-}{-}wid}\monoi{ $<$x$>$} ] Sets the identity threshold used by single-linkage clustering when
 using  \mono{{-}{-}wblosum}.\mono{} Invalid with any other weighting scheme. Default is 0.62.
 
\end{wideitem}

\subsection*{Options Controlling Effective Sequence Number}
 After relative weights are
determined, they are normalized to sum to a total effective sequence number,
 \monoi{eff\_nseq}.\monoi{} This number may be the actual number of sequences in the alignment,
but it is almost always smaller than that. The default entropy weighting
method  (\mono{{-}{-}eent}) reduces the effective sequence number to reduce the information
content (relative entropy, or average expected score on true homologs)
per consensus position. The target relative entropy is controlled by a two-parameter
function, where the two parameters are settable with \mono{{-}{-}ere} and  \mono{{-}{-}esigma}.  \begin{wideitem}
\item [\monob{{-}{-}eent}
] Adjust effective sequence number to achieve a specific relative entropy
per position (see \mono{{-}{-}ere}). This is the default.  
\item [\monob{{-}{-}eentexp} ] Adjust the effective
sequence number to reach the relative entropy target using exponential
scaling.  
\item [\monob{{-}{-}eclust} ] Set effective sequence number to the number of single-linkage
clusters at a specific identity threshold (see  \mono{{-}{-}eid}). This option is not
recommended; it's for experiments evaluating how much better \mono{{-}{-}eent} is.  
\item [\monob{{-}{-}enone}
] Turn off effective sequence number determination and just use the actual
number of sequences. One reason you might want to do this is to try to maximize
the relative entropy/position of your model, which may be useful for short
models.  
\item [\monob{{-}{-}eset}\monoi{ $<$x$>$} ] Explicitly set the effective sequence number for all models
to  \monoi{$<$x$>$}.  
\item [\monob{{-}{-}ere}\monoi{ $<$x$>$} ] Set the minimum relative entropy/position target to  \monoi{$<$x$>$}. Requires
\mono{{-}{-}eent}.\mono{} Default depends on the sequence alphabet; for protein sequences, it
is 0.59 bits/position.  
\item [\monob{{-}{-}esigma}\monoi{ $<$x$>$} ] Sets the minimum relative entropy contributed
by an entire model alignment, over its whole length. This has the effect
of making short models have  higher relative entropy per position than
 \mono{{-}{-}ere } alone would give. The default is 45.0 bits.  
\item [\monob{{-}{-}eid}\monoi{ $<$x$>$} ] Sets the fractional
pairwise identity cutoff used by  single linkage clustering with the \mono{{-}{-}eclust
} option. The default is 0.62.  
\end{wideitem}

\subsection*{Options Controlling Priors}
These options are
only valid when performing an iterative search or when both the query object
and target database are sequences.  In profile construction, by default,
weighted counts are converted to mean posterior probability parameter estimates
using mixture Dirichlet priors.  Default mixture Dirichlet prior parameters
for protein models and for nucleic acid (RNA and DNA) models are built
in. The following options allow you to override the default priors.  \begin{wideitem}
\item [\monob{{-}{-}pnone}
] Don't use any priors. Probability parameters will simply be the observed
frequencies, after relative sequence weighting.   
\item [\monob{{-}{-}plaplace} ] Use a Laplace
+1 prior in place of the default mixture Dirichlet prior.    
\end{wideitem}

\subsection*{Options Controlling
E-value Calibration}
These options are only valid when performing an iterative
search.  Estimating the location parameters for the expected score distributions
for MSV filter scores, Viterbi filter scores, and Forward scores requires
three short random sequence simulations.  \begin{wideitem}
\item [\monob{{-}{-}EmL}\monoi{ $<$n$>$} ] Sets the sequence length
in simulation that estimates the location parameter mu for MSV filter E-values.
Default is 200.  
\item [\monob{{-}{-}EmN}\monoi{ $<$n$>$} ] Sets the number of sequences in simulation that estimates
the location parameter mu for MSV filter E-values. Default is 200.  
\item [\monob{{-}{-}EvL}\monoi{ $<$n$>$}
] Sets the sequence length in simulation that estimates the location parameter
mu for Viterbi filter E-values. Default is 200.  
\item [\monob{{-}{-}EvN}\monoi{ $<$n$>$} ] Sets the number of
sequences in simulation that estimates the location parameter mu for Viterbi
filter E-values. Default is 200.  
\item [\monob{{-}{-}EfL}\monoi{ $<$n$>$} ] Sets the sequence length in simulation
that estimates the location parameter tau for Forward E-values. Default is
100.  
\item [\monob{{-}{-}EfN}\monoi{ $<$n$>$} ] Sets the number of sequences in simulation that estimates the
location parameter tau for Forward E-values. Default is 200.  
\item [\monob{{-}{-}Eft}\monoi{ $<$x$>$} ] Sets the
tail mass fraction to fit in the simulation that estimates the location
parameter tau for Forward evalues. Default is 0.04.  
\end{wideitem}

\subsection*{Other Options}
 \begin{wideitem}
\item [\monob{{-}{-}nonull2}
] Turn off the null2 score corrections for biased composition.  
\item [\monob{-Z}\monoi{ $<$x$>$} ] Assert
that the total number of targets in your searches is \monoi{$<$x$>$}, for the purposes
of per-sequence E-value calculations, rather than the actual number of targets
seen.   
\item [\monob{{-}{-}domZ}\monoi{ $<$x$>$} ] Assert that the total number of targets in your searches
is \monoi{$<$x$>$}, for the purposes of per-domain conditional E-value calculations, rather
than the number of targets that passed the reporting thresholds.  
\item [\monob{{-}{-}seed}\monoi{ $<$n$>$}
] Set the random number seed to  \monoi{$<$n$>$}. Some steps in postprocessing require Monte
Carlo simulation.  The default is to use a fixed seed (42), so that results
are exactly reproducible. Any other positive integer will give different
(but also reproducible) results. A choice of 0 uses a randomly chosen seed.
 
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmserver}}{hmmserver} - server that accelerates homology searches   }
\subsection*{Synopsis}
\noindent
\monob{hmmserver}
[\monoi{options}] \monoi{list}  \monoi{of}  \monoi{data}  \monoi{files}  \monoi{to}  \monoi{be}  \monoi{read}  \monoi{into}  \monoi{memory}   
\subsection*{Description}



The  \mono{hmmserver} command starts a service that accepts search requests from
clients and distributes the work of searches across one or more computers
to improve performance. Once started, the server listens on the specified
port for search requests, processes them as they arrive, and returns the
results to clients.  The server userguide (documentation/userguide/Server\_Userguide.pdf)
describes the format that \mono{hmmserver} expects for search requests and replies.
 

\mono{Hmmserver} must be run as an MPI (message-passing interface) process with
at least two ranks, one for the master process and one or more worker processes.
 See your local system's documentation for details on how to start an MPI
process, as the details vary depending on cluster configuration and the
implementation of MPI used.    

  
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help; print a brief reminder of command line usage and all
available options.  
\item [\monob{{-}{-}cport}\monoi{ $<$n$>$} ] Port to use for communication between clients
and the master node of the server.  The default is 51371.  
\item [\monob{{-}{-}ccncts}\monoi{ $<$n$>$} ] Maximum
number of client connections to accept. The default is 16.  
\item [\monob{{-}{-}num\_dbs}\monoi{ $<$n$>$} ] Number
of database files to read into RAM.  Must match the number of data files
passed as arguments to  \mono{hmmserver} , and defaults to one.  
\item [\monob{{-}{-}num\_shards}\monoi{ $<$n$>$} ] Number
of shards (pieces) to divide each database into to reduce memory usage.
Defaults to one, and must be less than the number of ranks in the MPI process
to allow at least one worker rank per shard, plus one rank for the master
node.  In general, having fewer shards will increase performance by allowing
better load-balancing at the cost of making the server use more memory on
each worker node.  Specifying a number of shards that is not an integer
divisor of the number of worker ranks will lead to suboptimal performance
as some shards will have fewer worker nodes operating on them than others.
   
\item [\monob{{-}{-}cpu}\monoi{ $<$n$>$} ] Number of worker threads to start on each worker node.  If 0 (the
default), queries the system's hardware to determine how many threads it
can support simultaneously, and starts that many minus one worker threads,
to leave one thread for the worker node's master thread.  The master node
always starts a fixed (small) number of threads, independent of the value
passed via this flag.  
\item [\monob{{-}{-}stall} ] Instructs the server to stall immediately after
startup so that a debugger can be attached to each node.  When this option
is set, the server sets the variable "stalling" to TRUE and enters a loop
that iterates until stalling is set to false by the debugger.  
\item [\monob{{-}{-}password}\monoi{
$<$password$>$} ] Specifies a password that a client must send along with the shutdown
command in order to shut down the server.     
\end{wideitem}

\newpage
