%% Manual pages chapter automatically generated. Do not edit.
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmc2}}{hmmc2} - example client for the HMMER daemon   }
\subsection*{Synopsis}
\noindent
\monob{hmmc2} [\monoi{options}]
  
\subsection*{Description}
 

\mono{Hmmc2} is a text client for the hmmpgmd or hmmpgmd\_shard daemons.
 When run, it opens a connection to a daemon at the specified IP address
and port, and then enters an interactive loop waiting for the user to input
commands to be sent to the daemon. See the User's Guide for the HMMER Daemon
for a discussion of hmmpgmd's command format.  

  
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-i $<$IP address$>$} ] Specify the IP address of the daemon that hmmc2
should connect to.  Defaults to 127.0.0.1 if not provided   
\item [\monob{-p $<$port number$>$} ] Specify
the port number that the daemon is listening on.  Defaults to 51371 if not
provided   
\item [\monob{-S} ] Print the scores of any hits found during searches.   
\item [\monob{-A} ] Print
the alignment of any hits found during searches.  This is a superset of
the "-S" flag, so providing both is redundant.    
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmpgmd}}{hmmpgmd} - daemon for database search web services   }
\subsection*{Synopsis}
\noindent
\monob{hmmpgmd}
[\monoi{options}]   
\subsection*{Description}
 

The \mono{hmmpgmd } program is the daemon that we use
internally for the hmmer.org web server.  It essentially stands in front
of the search programs \mono{phmmer},\mono{} \mono{hmmsearch}, and  \mono{hmmscan}.\mono{}  

To use \mono{hmmpgmd},\mono{}
first an instance must be started up as a  master  server, and provided
with at least one  sequence database (using the  \mono{{-}{-}seqdb} flag) and/or an
 HMM database (using the \mono{{-}{-}hmmdb} flag).  A sequence database must be in hmmpgmd
format, which may be produced using  \mono{esl-reformat}. An HMM database is of
the form produced by  \mono{hmmbuild}. The input database(s) will be loaded into
memory by the  master. When the master has finished loading the database(s),
it  prints the line: "Data loaded into memory. Master is ready."   

After
the master is ready, one or more instances of hmmpgmd may be started as
workers. These workers may be (and typically are) on different machines
from the master, but must have access to the  same database file(s) provided
to the master, with the same path. As  with the master, each worker loads
the database(s) into memory, and  indicates completion by printing: "Data
loaded into memory. Worker is ready."   

The master process and workers are
expected to remain running. One or more clients then connect to the master
and submit possibly many queries. The master distributes the work of a query
among the workers, collects results, and merges them before responding
to the client. Two example client programs are included in the HMMER src
 directory - the C program \mono{hmmc2} and the perl script \mono{hmmpgmd\_client\_example.pl}.
These are intended as examples only, and should be extended as  necessary
to meet your needs.   

A query is submitted to the master from the client
as a character string. Queries may be the sort that would normally be handled
by  \mono{phmmer} (protein sequence vs protein database), \mono{hmmsearch} (protein HMM
query vs protein database), or \mono{hmmscan} (protein query vs protein HMM database).
 

  The general form of a client query is to start with a single line of
the form  \mono{@[options]},\mono{} followed by multiple lines of text representing either
the query HMM  or fasta-formatted sequence. The final line of each query
is the separator  \mono{//}.   

For example, to perform a  \mono{phmmer} type search of
a sequence against a sequence database  file, the first line is of the
form  \mono{@{-}{-}seqdb 1}, then the fasta-formatted query sequence starting with the
header line \mono{$>$sequence-name}, followed by one or more lines of sequence, and
finally the closing \mono{//}.  

To perform an \mono{hmmsearch } type search, the query
sequence is replaced by the full text of a HMMER-format query HMM.   

To perform
an \mono{hmmscan } type search, the text matches that of the  \mono{phmmer} type search,
except that the first line changes to  \mono{@{-}{-}hmmdb 1}.  

In the hmmpgmd-formatted
sequence database file, each sequence can be associated with one or more
sub-databases. The  \mono{{-}{-}seqdb} flag indicates which of these sub-databases will
be queried.  The HMM database format does not support sub-databases.    

  
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help; print a brief reminder of command line usage and all
available options.  
\item [\monob{{-}{-}master} ] Run as the master server.  
\item [\monob{{-}{-}worker}\monoi{ $<$s$>$} ] Run as a worker,
connecting to the master server that is running on IP address \monoi{$<$s$>$}.  
\item [\monob{{-}{-}cport}\monoi{
$<$n$>$} ] Port to use for communication between clients and the master server. 
The default is 51371.  
\item [\monob{{-}{-}wport}\monoi{ $<$n$>$} ] Port to use for communication between workers
and the master server.  The default is 51372.  
\item [\monob{{-}{-}ccncts}\monoi{ $<$n$>$} ] Maximum number of
client connections to accept. The default is 16.  
\item [\monob{{-}{-}wcncts}\monoi{ $<$n$>$} ] Maximum number
of worker connections to accept. The default is 32.  
\item [\monob{{-}{-}pid}\monoi{ $<$f$>$} ] Name of file into
which the process id will be written.   
\item [\monob{{-}{-}seqdb}\monoi{ $<$f$>$} ] Name of the file (in \mono{hmmpgmd}
format) containing protein sequences. The contents of this file will be
cached for searches.   
\item [\monob{{-}{-}hmmdb}\monoi{ $<$f$>$} ] Name of the file containing protein HMMs.
The contents of this file  will be cached for searches.  
\item [\monob{{-}{-}cpu}\monoi{ $<$n$>$} ] Number of
parallel threads to use (for  \mono{{-}{-}worker} ).   
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmpgmd\_shard}}{hmmpgmd\_shard} - sharded daemon for database search web services   }

\subsection*{Synopsis}
\noindent
\monob{hmmpgmd\_shard} [\monoi{options}]   
\subsection*{Description}
 

The  \mono{hmmpgmd\_shard } program
provides a sharded version of the  \mono{hmmpgmd } program that we use internally
to implement high-performance HMMER services that can be accessed via the
internet.  See the  \mono{hmmpgmd} man page for a discussion of how the base  \mono{hmmpgmd}
program is used.  This man page discusses differences between  \mono{hmmpgmd\_shard}
and  \mono{hmmpgmd.   } The base  \mono{hmmpgmd} program loads the entirety of its database
file into RAM on every worker node, in spite of the fact that each worker
node searches a predictable fraction of the database(s) contained in that
file when performing searches.  This wastes RAM, particularly when many
worker nodes are used to accelerate searches of large databases.  

\mono{Hmmpgmd\_shard
} addresses this by dividing protein sequence database files into shards.
 Each worker node loads only 1/Nth of the database file, where N is the
number of worker nodes attached to the master.  HMM database files are not
sharded, meaning that every worker node will load the entire database file
into RAM.  Current HMM databases are much smaller than current protein sequence
databases, and easily fit into the RAM of modern servers even without sharding.
 

\mono{Hmmpgmd\_shard } is used in the same manner as  \mono{hmmpgmd} , except that it
takes one additional argument:  \mono{{-}{-}num\_shards}\monoi{ $<$n$>$} , which specifies the number
of shards that protein databases will be divided into, and defaults to
1 if unspecified.  This argument is only valid for the master node of a
 \mono{hmmpgmd} system (i.e., when  \mono{{-}{-}master} is passed to the  \mono{hmmpgmd} program), and
must be equal to the number of worker nodes that will connect to the master
node.   \mono{Hmmpgmd\_shard } will signal an error if more than  \mono{num\_shards} worker
nodes attempt to connect to the master node or if a search is started when
fewer than  \mono{num\_shards}\monoi{} workers are connected to the master.  
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help;
print a brief reminder of command line usage and all available options.
 
\item [\monob{{-}{-}master} ] Run as the master server.  
\item [\monob{{-}{-}worker}\monoi{ $<$s$>$} ] Run as a worker, connecting
to the master server that is running on IP address \monoi{$<$s$>$}.  
\item [\monob{{-}{-}cport}\monoi{ $<$n$>$} ] Port to use
for communication between clients and the master server.  The default is
51371.  
\item [\monob{{-}{-}wport}\monoi{ $<$n$>$} ] Port to use for communication between workers and the master
server.  The default is 51372.  
\item [\monob{{-}{-}ccncts}\monoi{ $<$n$>$} ] Maximum number of client connections
to accept. The default is 16.  
\item [\monob{{-}{-}wcncts}\monoi{ $<$n$>$} ] Maximum number of worker connections
to accept. The default is 32.  
\item [\monob{{-}{-}pid}\monoi{ $<$f$>$} ] Name of file into which the process
id will be written.   
\item [\monob{{-}{-}seqdb}\monoi{ $<$f$>$} ] Name of the file (in \mono{hmmpgmd} format) containing
protein sequences. The contents of this file will be cached for searches.
  
\item [\monob{{-}{-}hmmdb}\monoi{ $<$f$>$} ] Name of the file containing protein HMMs. The contents of this
file  will be cached for searches.  
\item [\monob{{-}{-}cpu}\monoi{ $<$n$>$} ] Number of parallel threads to
use (for  \mono{{-}{-}worker} ).  
\item [\monob{{-}{-}num\_shards}\monoi{ $<$n$>$} ] Number of shards to divide cached sequence
database(s) into.  HMM databases are not sharded, due to their small size.
This option is only valid when the  \mono{{-}{-}master } option is present, and defaults
to 1 if not specified. \mono{Hmmpgmd\_shard } requires that the number of shards
be equal to the number of worker nodes, and will give errors if more than
 \mono{num\_shards}\monoi{} workers attempt to connect to the master node or if a search
is started with fewer than  \mono{num\_shards}\monoi{} workers connected to the master.
 
\end{wideitem}

\newpage
