%% Manual pages chapter automatically generated. Do not edit.
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{alimask}}{alimask} - calculate and add column mask to a multiple sequence alignment}
 
\subsection*{Synopsis}
\noindent
\monob{alimask} [\monoi{options}] \monoi{msafile} \monoi{postmsafile}   
\subsection*{Description}
 

\mono{alimask} is
used to apply a mask line to a multiple sequence alignment, based on provided
alignment or model coordinates. When  \mono{hmmbuild} receives a masked alignment
as input, it produces a profile model in which the emission probabilities
at masked positions are set to match the background frequency, rather than
being set based on observed frequencies in the alignment.  Position-specific
insertion and deletion rates are not  altered, even in masked regions. 
\mono{alimask} autodetects input format, and produces masked alignments  in Stockholm
format.   \monoi{msafile}  may contain only one sequence alignment.  

A common motivation
for masking a region in an alignment is that the region contains a simple
tandem repeat that is  observed to cause an unacceptably high rate of false
positive hits.   

In the simplest case, a mask range is given in coordinates
 relative to the input alignment, using \mono{{-}{-}alirange}\monoi{ $<$s$>$}\mono{.}\monoi{} However it is more often
the case that the region to be  masked has been identified in coordinates
relative to  the profile model (e.g. based on recognizing a simple  repeat
pattern in false hit alignments or in the HMM logo).  Not all alignment
columns are converted to match state  positions in the profile (see the
 \mono{{-}{-}symfrac} flag for  \mono{hmmbuild} for discussion), so model positions do not
necessarily match  up to alignment column positions.  To remove the burden
of converting model positions to  alignment positions,  \mono{alimask} accepts
the mask range input in model coordinates as well,  using \mono{{-}{-}modelrange}\monoi{ $<$s$>$}\mono{.}\monoi{}
When using this flag,  \monoi{alimask} determines which alignment positions would
be identified by \mono{hmmbuild} as match states, a process that requires that
all \mono{hmmbuild } flags impacting that decision be supplied to  \mono{alimask}. It
is for this reason that many of the  \mono{hmmbuild } flags are also used by \mono{alimask}.\mono{}
   
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help; print a brief reminder of command line usage and all
available options.  
\item [\monob{-o}\monoi{ $<$f$>$} ] Direct the summary output to file \monoi{$<$f$>$}, rather than
to stdout.   
\end{wideitem}

\subsection*{Options for Specifying Mask Range}
 A single mask range is given
as a dash-separated pair, like \mono{{-}{-}modelrange 10-20} and multiple ranges may be
submitted as a comma-separated list, \mono{{-}{-}modelrange 10-20,30-42}.   \begin{wideitem}
\item [\monob{{-}{-}modelrange}\monoi{ $<$s$>$}
] Supply the given range(s) in model coordinates.   
\item [\monob{{-}{-}alirange}\monoi{ $<$s$>$} ] Supply the
given range(s) in alignment coordinates.   
\item [\monob{{-}{-}appendmask } ] Add to the existing
mask found with the alignment. The default is to overwrite any existing
mask.   
\item [\monob{{-}{-}model2ali}\monoi{ $<$s$>$} ] Print model range(s) and the corresponding alignment
range(s). No masked alignment is produced. The output is a single line for
each input range, of the form 

   i..j -$>$ m..n

with i \& j representing model range values, and m \& n representing alignment
range values.  
\item [\monob{{-}{-}ali2model}\monoi{ $<$s$>$} ] Print alignment range(s) and the corresponding
model range(s). No masked alignment is produced. Because some alignment positions
may not map to model positions, the range(s) produced will begin with the
first alignment position between $<$from$>$ and $<$to$>$ (inclusive) that maps to a
model position, and end with the final alignment position in that range
that maps to a model position. The output is a single line for each input
range, of the form 

  i..j -$>$ m..n

with i \& j representing alignment range values, and m \& n representing
model range values. If no alignment positions in the range $<$from$>$..$<$to$>$ map to
a model position, the output prints the input $<$from$>$ and $<$to$>$ mapping to nothing,
with the format: 

 i..j  -$>$   -..-  (no map) .

  
\end{wideitem}

\subsection*{Options for Specifying the Alphabet}
 \begin{wideitem}
\item [\monob{{-}{-}amino} ] Assert that sequences in  \monoi{msafile}
are protein, bypassing alphabet autodetection.  
\item [\monob{{-}{-}dna} ] Assert that sequences
in \monoi{msafile} are DNA, bypassing alphabet autodetection.  
\item [\monob{{-}{-}rna} ] Assert that sequences
in  \monoi{msafile} are RNA, bypassing alphabet autodetection.    
\end{wideitem}

\subsection*{Options Controlling
Profile Construction}
 These options control how consensus columns are defined
in an alignment.  \begin{wideitem}
\item [\monob{{-}{-}fast } ] Define consensus columns as those that have a fraction
$>$=  \mono{symfrac} of residues as opposed to gaps. (See below for the \mono{{-}{-}symfrac} option.)
This is the default.  
\item [\monob{{-}{-}hand} ] Define consensus columns in next profile using
reference annotation to the multiple alignment.  This allows you to define
any consensus columns you like.  
\item [\monob{{-}{-}symfrac}\monoi{ $<$x$>$} ] Define the residue fraction threshold
necessary to define a consensus column when using the  \mono{{-}{-}fast } option. The
default is 0.5. The symbol fraction in each column is calculated after taking
relative sequence weighting into account, and ignoring gap characters corresponding
to ends of sequence fragments (as opposed to internal insertions/deletions).
Setting this to 0.0 means that every alignment column will be assigned as
consensus, which may be useful in some cases. Setting it to 1.0 means that
only columns that include 0 gaps (internal insertions/deletions) will be
assigned as consensus.  
\item [\monob{{-}{-}fragthresh}\monoi{ $<$x$>$} ] We only want to count terminal gaps
as deletions if the aligned sequence is known to be full-length, not if
it is a fragment (for instance, because only part of it was sequenced).
HMMER uses a simple rule to infer fragments: if the sequence length L is
less than  or equal to a fraction \monoi{$<$x$>$}  times the alignment length in columns,
then the sequence is handled as a fragment. The default is 0.5. Setting \mono{{-}{-}fragthresh
0} will define no (nonempty) sequence as a fragment; you might want to do
this if you know you've got a carefully curated alignment of full-length
sequences. Setting \mono{{-}{-}fragthresh 1} will define all sequences as fragments;
you might want to do this if you know your alignment is entirely composed
of fragments, such as translated short reads in metagenomic shotgun data.
  
\end{wideitem}

\subsection*{Options Controlling Relative Weights}
 HMMER uses an ad hoc sequence weighting
algorithm to downweight closely related sequences and upweight distantly
related ones. This has the effect of making models less biased by uneven
phylogenetic representation. For example, two identical sequences would
typically each receive half the weight that one sequence would.  These options
control which algorithm gets used.  \begin{wideitem}
\item [\monob{{-}{-}wpb} ] Use the Henikoff position-based sequence
weighting scheme [Henikoff and Henikoff, J. Mol. Biol. 243:574, 1994].  This
is the default.  
\item [\monob{{-}{-}wgsc } ] Use the Gerstein/Sonnhammer/Chothia weighting algorithm
[Gerstein et al, J. Mol. Biol. 235:1067, 1994].  
\item [\monob{{-}{-}wblosum} ] Use the same clustering
scheme that was used to weight data in calculating BLOSUM substitution
matrices [Henikoff and Henikoff, Proc. Natl. Acad. Sci 89:10915, 1992]. Sequences
are single-linkage clustered at an identity threshold (default 0.62; see
\mono{{-}{-}wid}) and within each cluster of c sequences, each sequence gets relative
weight 1/c.  
\item [\monob{{-}{-}wnone} ] No relative weights. All sequences are assigned uniform
weight.   
\item [\monob{{-}{-}wid}\monoi{ $<$x$>$} ] Sets the identity threshold used by single-linkage clustering
when  using  \mono{{-}{-}wblosum}.\mono{} Invalid with any other weighting scheme. Default is
0.62.      
\end{wideitem}

\subsection*{Other Options}
 \begin{wideitem}
\item [\monob{{-}{-}informat}\monoi{ $<$s$>$} ] Assert that input \monoi{msafile} is in alignment
format \monoi{$<$s$>$}, bypassing format autodetection. Common choices for  \monoi{$<$s$>$}  include:
\mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal}, \mono{phylip}. For more information, and
for codes for some less common formats, see main documentation. The string
\monoi{$<$s$>$} is case-insensitive (\mono{a2m} or \mono{A2M} both work).   
\item [\monob{{-}{-}outformat}\monoi{ $<$s$>$} ] Write the output
\monoi{postmsafile} in alignment format \monoi{$<$s$>$}. Common choices for  \monoi{$<$s$>$}  include: \mono{stockholm},\mono{}
\mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal}, \mono{phylip}. The string \monoi{$<$s$>$} is case-insensitive (\mono{a2m}
or \mono{A2M} both work). Default is \mono{stockholm}.   
\item [\monob{{-}{-}seed}\monoi{ $<$n$>$} ] Seed the random number
generator with \monoi{$<$n$>$}, an integer $>$= 0.  If  \monoi{$<$n$>$}  is nonzero, any stochastic simulations
will be reproducible; the same command will give the same results. If  \monoi{$<$n$>$}
is 0, the random number generator is seeded arbitrarily, and stochastic
simulations will vary from run to run of the same command. The default seed
is 42.    
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmalign}}{hmmalign} - align sequences to a profile   }
\subsection*{Synopsis}
\noindent
\monob{hmmalign} [\monoi{options}]
\monoi{hmmfile} \monoi{seqfile}  
\subsection*{Description}
 

Perform a multiple sequence alignment of all
the sequences in \monoi{seqfile} by aligning them individually to the profile HMM
in \monoi{hmmfile.} The new alignment is output to stdout.  

The \monoi{hmmfile} should contain
only a single profile. If it contains more, only the first profile in the
file will be used.   

Either  \monoi{hmmfile} or  \monoi{seqfile}  (but not both) may be
'-' (dash), which means reading this input from stdin rather than a file. 
  

The sequences in  \monoi{seqfile} are aligned in unihit local alignment mode.
 Therefore they should already be known to contain only a single domain
(or a fragment of one). The optimal alignment may assign some residues as
nonhomologous (N and C states), in which case these residues are still
included in the resulting alignment, but shoved to the outer edges. To trim
these unaligned nonhomologous residues from the result, see the \mono{{-}{-}trim} option.
  
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help; print a brief reminder of command line usage and all
available options.  
\item [\monob{-o}\monoi{ $<$f$>$} ] Direct the output alignment to file \monoi{$<$f$>$,} rather than
to stdout.  
\item [\monob{{-}{-}mapali}\monoi{ $<$f$>$} ] Merge the existing alignment in file  \monoi{$<$f$>$} into the result,
where  \monoi{$<$f$>$}  is exactly the same alignment that was used to build the model
in   \monoi{hmmfile.} This is done using a map of alignment columns to consensus
 profile positions that is stored in the \monoi{hmmfile.} The multiple alignment
in  \monoi{$<$f$>$} will be exactly reproduced in its consensus columns (as defined by
the profile), but the displayed alignment in insert columns may be altered,
because insertions relative to a profile are considered by convention to
be unaligned data.   
\item [\monob{{-}{-}trim} ] Trim nonhomologous residues (assigned to N and
C states in the optimal alignments) from the resulting multiple alignment
output.   
\item [\monob{{-}{-}amino} ] Assert that sequences in  \monoi{seqfile} are protein, bypassing
alphabet autodetection.  
\item [\monob{{-}{-}dna} ] Assert that sequences in \monoi{seqfile} are DNA, bypassing
alphabet autodetection.  
\item [\monob{{-}{-}rna} ] Assert that sequences in  \monoi{seqfile} are RNA,
bypassing alphabet autodetection.   
\item [\monob{{-}{-}informat}\monoi{ $<$s$>$} ] Assert that input \monoi{seqfile}
is in format \monoi{$<$s$>$}, bypassing format autodetection. Common choices for  \monoi{$<$s$>$}  include:
\mono{fasta}, \mono{embl}, \mono{genbank.} Alignment formats also work; common choices include:
\mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal}, \mono{phylip}. For more information, and
for codes for some less common formats, see main documentation. The string
\monoi{$<$s$>$} is case-insensitive (\mono{fasta} or \mono{FASTA} both work).  
\item [\monob{{-}{-}outformat}\monoi{ $<$s$>$} ] Write the
output alignment in format \monoi{$<$s$>$}. Common choices for  \monoi{$<$s$>$}  include: \mono{stockholm},\mono{}
\mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal}, \mono{phylip}. The string \monoi{$<$s$>$} is case-insensitive (\mono{a2m}
or \mono{A2M} both work). Default is \mono{stockholm}.    
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmbuild}}{hmmbuild} - construct profiles from multiple sequence alignments  }
\subsection*{Synopsis}
\noindent
\monob{hmmbuild}
[\monoi{options}] \monoi{hmmfile} \monoi{msafile}   
\subsection*{Description}
 For each multiple sequence alignment
in  \monoi{msafile} build a profile HMM  and save it to a new file \monoi{hmmfile}.   

\monoi{msafile}
 may be '-' (dash), which means reading this input from stdin rather than
a file.    

\monoi{hmmfile} may not be '-' (stdout), because sending the HMM file to
stdout would conflict with the other text output of the program.     
\subsection*{Options}

\begin{wideitem}
\item [\monob{-h} ] Help; print a brief reminder of command line usage and all available
options.  
\item [\monob{-n}\monoi{ $<$s$>$} ] Name the new profile  \monoi{$<$s$>$}. The default is to use the name of
the alignment (if one is present in  the  \monoi{msafile}, or, failing that, the
name of the \monoi{hmmfile}. If  \monoi{msafile} contains more than one alignment,  \mono{-n} doesn't
work, and every alignment must have a name  annotated in the  \monoi{msafile} (as
in Stockholm \#=GF ID annotation).   
\item [\monob{-o}\monoi{ $<$f$>$} ] Direct the summary output to file
\monoi{$<$f$>$}, rather than to stdout.  
\item [\monob{-O}\monoi{ $<$f$>$} ] After each model is constructed, resave annotated,
possibly modified source alignments to a file \monoi{$<$f$>$} in Stockholm format. The
alignments are annotated with a reference annotation line indicating which
columns were assigned as consensus, and sequences are annotated with what
relative sequence weights were assigned. Some residues of the alignment
may have been shifted to accommodate restrictions of the Plan7 profile
architecture, which disallows transitions between insert and delete states.
  
\end{wideitem}

\subsection*{Options for Specifying the Alphabet}
 \begin{wideitem}
\item [\monob{{-}{-}amino} ] Assert that sequences in  \monoi{msafile}
are protein, bypassing alphabet autodetection.  
\item [\monob{{-}{-}dna} ] Assert that sequences
in \monoi{msafile} are DNA, bypassing alphabet autodetection.  
\item [\monob{{-}{-}rna} ] Assert that sequences
in  \monoi{msafile} are RNA, bypassing alphabet autodetection.  
\end{wideitem}

\subsection*{Options Controlling
Profile Construction}
 These options control how consensus columns are defined
in an alignment.  \begin{wideitem}
\item [\monob{{-}{-}fast } ] Define consensus columns as those that have a fraction
$>$=  \mono{symfrac} of residues as opposed to gaps. (See below for the \mono{{-}{-}symfrac} option.)
This is the default.  
\item [\monob{{-}{-}hand} ] Define consensus columns in next profile using
reference annotation to the multiple alignment.  This allows you to define
any consensus columns you like.  
\item [\monob{{-}{-}symfrac}\monoi{ $<$x$>$} ] Define the residue fraction threshold
necessary to define a consensus column when using the  \mono{{-}{-}fast } option. The
default is 0.5. The symbol fraction in each column is calculated after taking
relative sequence weighting into account, and ignoring gap characters corresponding
to ends of sequence fragments (as opposed to internal insertions/deletions).
Setting this to 0.0 means that every alignment column will be assigned as
consensus, which may be useful in some cases. Setting it to 1.0 means that
only columns that include 0 gaps (internal insertions/deletions) will be
assigned as consensus.  
\item [\monob{{-}{-}fragthresh}\monoi{ $<$x$>$} ] We only want to count terminal gaps
as deletions if the aligned sequence is known to be full-length, not if
it is a fragment (for instance, because only part of it was sequenced).
HMMER uses a simple rule to infer fragments: if the range of a sequence
in the alignment  (the number of alignment columns between the first and
last positions  of the sequence) is less than or equal to a fraction \monoi{$<$x$>$}
 times the alignment length in columns, then the sequence is handled as
a fragment. The default is 0.5. Setting \mono{{-}{-}fragthresh 0} will define no (nonempty)
sequence as a fragment; you might want to do this if you know you've got
a carefully curated alignment of full-length sequences. Setting \mono{{-}{-}fragthresh
1} will define all sequences as fragments; you might want to do this if
you know your alignment is entirely composed of fragments, such as translated
short reads in metagenomic shotgun data.   
\end{wideitem}

\subsection*{Options Controlling Relative
Weights}
 HMMER uses an ad hoc sequence weighting algorithm to downweight
closely related sequences and upweight distantly related ones. This has
the effect of making models less biased by uneven phylogenetic representation.
For example, two identical sequences would typically each receive half
the weight that one sequence would.  These options control which algorithm
gets used.  \begin{wideitem}
\item [\monob{{-}{-}wpb} ] Use the Henikoff position-based sequence weighting scheme
[Henikoff and Henikoff, J. Mol. Biol. 243:574, 1994].  This is the default.
 
\item [\monob{{-}{-}wgsc } ] Use the Gerstein/Sonnhammer/Chothia weighting algorithm [Gerstein
et al, J. Mol. Biol. 235:1067, 1994].  
\item [\monob{{-}{-}wblosum} ] Use the same clustering scheme
that was used to weight data in calculating BLOSUM substitution matrices
[Henikoff and Henikoff, Proc. Natl. Acad. Sci 89:10915, 1992]. Sequences are
single-linkage clustered at an identity threshold (default 0.62; see \mono{{-}{-}wid})
and within each cluster of c sequences, each sequence gets relative weight
1/c.  
\item [\monob{{-}{-}wnone} ] No relative weights. All sequences are assigned uniform weight.
  
\item [\monob{{-}{-}wid}\monoi{ $<$x$>$} ] Sets the identity threshold used by single-linkage clustering when
 using  \mono{{-}{-}wblosum}.\mono{} Invalid with any other weighting scheme. Default is 0.62.
    
\end{wideitem}

\subsection*{Options Controlling Effective Sequence Number}
 After relative weights
are determined, they are normalized to sum to a total effective sequence
number,  \monoi{eff\_nseq}.\monoi{} This number may be the actual number of sequences in
the alignment, but it is almost always smaller than that. The default entropy
weighting method  (\mono{{-}{-}eent}) reduces the effective sequence number to reduce
the information content (relative entropy, or average expected score on
true homologs) per consensus position. The target relative entropy is controlled
by a two-parameter function, where the two parameters are settable with
\mono{{-}{-}ere} and  \mono{{-}{-}esigma}.  \begin{wideitem}
\item [\monob{{-}{-}eent} ] Adjust effective sequence number to achieve a specific
relative entropy per position (see \mono{{-}{-}ere}). This is the default.  
\item [\monob{{-}{-}eclust} ] Set
effective sequence number to the number of single-linkage clusters at a
specific identity threshold (see  \mono{{-}{-}eid}). This option is not recommended;
it's for experiments evaluating how much better \mono{{-}{-}eent} is.  
\item [\monob{{-}{-}enone} ] Turn off
effective sequence number determination and just use the actual number
of sequences. One reason you might want to do this is to try to maximize
the relative entropy/position of your model, which may be useful for short
models.  
\item [\monob{{-}{-}eset}\monoi{ $<$x$>$} ] Explicitly set the effective sequence number for all models
to  \monoi{$<$x$>$}.  
\item [\monob{{-}{-}ere}\monoi{ $<$x$>$} ] Set the minimum relative entropy/position target to  \monoi{$<$x$>$}. Requires
\mono{{-}{-}eent}.\mono{} Default depends on the sequence alphabet. For protein sequences, it
is 0.59 bits/position; for nucleotide  sequences, it is 0.45 bits/position.
 
\item [\monob{{-}{-}esigma}\monoi{ $<$x$>$} ] Sets the minimum relative entropy contributed by an entire model
alignment, over its whole length. This has the effect of making short models
have  higher relative entropy per position than  \mono{{-}{-}ere } alone would give.
The default is 45.0 bits.  
\item [\monob{{-}{-}eid}\monoi{ $<$x$>$} ] Sets the fractional pairwise identity cutoff
used by  single linkage clustering with the \mono{{-}{-}eclust } option. The default
is 0.62.   
\end{wideitem}

\subsection*{Options Controlling Priors}
 By default, weighted counts are converted
to mean posterior probability parameter estimates using mixture Dirichlet
priors. Default mixture Dirichlet prior parameters for protein models and
for nucleic acid (RNA and DNA) models are built in. The following options
allow you to override the default priors.  \begin{wideitem}
\item [\monob{{-}{-}pnone} ] Don't use any priors. Probability
parameters will simply be the observed frequencies, after relative sequence
weighting.   
\item [\monob{{-}{-}plaplace} ] Use a Laplace +1 prior in place of the default mixture
Dirichlet prior.     
\end{wideitem}

\subsection*{Options Controlling Single Sequence Scoring}
 By default,
if a query is a single sequence from a file in  \monoi{fasta} format, \mono{hmmbuild
} constructs a search model from that sequence and a standard 20x20 substitution
matrix for residue probabilities, along with two additional parameters
for position-independent gap open and gap extend probabilities. These options
allow the default single-sequence scoring parameters to be changed, and
for single-sequence scoring options to be applied to a single sequence coming
from an aligned format.  \begin{wideitem}
\item [\monob{{-}{-}singlemx}\monoi{} ] If a single sequence query comes from
a multiple sequence alignment file,  such as in  \monoi{stockholm} format, the
search model is by default constructed as is typically done  for multiple
sequence alignments. This option forces  \mono{hmmbuild } to use the single-sequence
method with substitution score matrix.  
\item [\monob{{-}{-}mx}\monoi{ $<$s$>$} ] Obtain residue alignment probabilities
from the built-in substitution matrix named \monoi{$<$s$>$}.\monoi{} Several standard matrices
are built-in, and do not need to be read from files.  The matrix name \monoi{$<$s$>$} 
can be PAM30, PAM70, PAM120, PAM240, BLOSUM45, BLOSUM50, BLOSUM62, BLOSUM80,
BLOSUM90, or DNA1. Only one of the \mono{{-}{-}mx } and \mono{{-}{-}mxfile} options may be used.  
\item [\monob{{-}{-}mxfile}\monoi{
$<$mxfile$>$} ] Obtain residue alignment probabilities from the substitution matrix
in file \monoi{$<$mxfile$>$}. The default score matrix is BLOSUM62 for protein sequences,
and  DNA1 for nucleotide sequences (these matrices are internal to HMMER
and do not need to be available as a file).  The format of a substitution
matrix \monoi{$<$mxfile$>$} is the standard format accepted by BLAST, FASTA, and other
sequence  analysis software.  See ftp.ncbi.nlm.nih.gov/blast/matrices/ for example
files. (The only exception: we require matrices to be square, so for DNA,
use files like NCBI's NUC.4.4, not NUC.4.2.)  
\item [\monob{{-}{-}popen}\monoi{ $<$x$>$} ] Set the gap open probability
for a single sequence query model to  \monoi{$<$x$>$}. The default is 0.02.  \monoi{$<$x$>$}  must be
$>$= 0 and $<$ 0.5.  
\item [\monob{{-}{-}pextend}\monoi{ $<$x$>$} ] Set the gap extend probability for a single sequence
query model to  \monoi{$<$x$>$}. The default is 0.4.  \monoi{$<$x$>$}  must be $>$= 0 and $<$ 1.0.   
\end{wideitem}

\subsection*{Options Controlling
E-value Calibration}
 The location parameters for the expected score distributions
for MSV filter scores, Viterbi filter scores, and Forward scores require
three short random sequence simulations.  \begin{wideitem}
\item [\monob{{-}{-}EmL}\monoi{ $<$n$>$} ] Sets the sequence length
in simulation that estimates the location parameter mu for MSV filter E-values.
Default is 200.  
\item [\monob{{-}{-}EmN}\monoi{ $<$n$>$} ] Sets the number of sequences in simulation that estimates
the location parameter mu for MSV filter E-values. Default is 200.  
\item [\monob{{-}{-}EvL}\monoi{ $<$n$>$}
] Sets the sequence length in simulation that estimates the location parameter
mu for Viterbi filter E-values. Default is 200.  
\item [\monob{{-}{-}EvN}\monoi{ $<$n$>$} ] Sets the number of
sequences in simulation that estimates the location parameter mu for Viterbi
filter E-values. Default is 200.  
\item [\monob{{-}{-}EfL}\monoi{ $<$n$>$} ] Sets the sequence length in simulation
that estimates the location parameter tau for Forward E-values. Default is
100.  
\item [\monob{{-}{-}EfN}\monoi{ $<$n$>$} ] Sets the number of sequences in simulation that estimates the
location parameter tau for Forward E-values. Default is 200.  
\item [\monob{{-}{-}Eft}\monoi{ $<$x$>$} ] Sets the
tail mass fraction to fit in the simulation that estimates the location
parameter tau for Forward evalues. Default is 0.04.   
\end{wideitem}

\subsection*{Other Options}
 \begin{wideitem}
\item [\monob{{-}{-}cpu}\monoi{ $<$n$>$}
] Set the number of parallel worker threads to  \monoi{$<$n$>$}. On multicore machines,
the default is 2. You can also control this number by setting an environment
variable,  \monoi{HMMER\_NCPU}. There is also a master thread, so the actual number
of threads that HMMER spawns is \monoi{$<$n$>$}+1.  This option is not available if HMMER
was compiled with POSIX threads support turned off.    
\item [\monob{{-}{-}informat}\monoi{ $<$s$>$} ] Assert
that input \monoi{msafile} is in alignment format \monoi{$<$s$>$}, bypassing format autodetection.
Common choices for  \monoi{$<$s$>$}  include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal},
\mono{phylip}. For more information, and for codes for some less common formats,
see main documentation. The string \monoi{$<$s$>$} is case-insensitive (\mono{a2m} or \mono{A2M} both
work).   
\item [\monob{{-}{-}seed}\monoi{ $<$n$>$} ] Seed the random number generator with \monoi{$<$n$>$}, an integer $>$= 0.
 If  \monoi{$<$n$>$}  is nonzero, any stochastic simulations will be reproducible; the
same command will give the same results. If  \monoi{$<$n$>$} is 0, the random number generator
is seeded arbitrarily, and stochastic simulations will vary from run to
run of the same command. The default seed is 42.   
\item [\monob{{-}{-}w\_beta}\monoi{ $<$x$>$} ] Window length
tail mass. The upper bound, \monoi{W}, on the length at which nhmmer expects to
find an instance of the  model is set such that the fraction of all sequences
generated by the model with length  \monoi{$>$= W} is less than   \monoi{$<$x$>$}.\monoi{} The default is
1e-7.     
\item [\monob{{-}{-}w\_length}\monoi{ $<$n$>$} ] Override the model instance length upper bound, \monoi{W}, which
is otherwise controlled by \mono{{-}{-}w\_beta}.\mono{} It should be larger than the model length.
The value of  \monoi{W} is used deep in the acceleration pipeline, and modest changes
are not expected to impact results (though larger values of  \monoi{W} do lead
to longer run time).    
\item [\monob{{-}{-}mpi} ] Run as a parallel MPI program. Each alignment
is assigned to a MPI worker node for construction. (Therefore, the maximum
parallelization  cannot exceed the number of alignments in the input \monoi{msafile}.)
This is useful when building large profile libraries. This option is only
available if optional MPI capability was enabled at compile-time.   
\item [\monob{{-}{-}stall}
] For debugging MPI parallelization: arrest program execution immediately
after start, and wait for a debugger to attach to the running process and
release the arrest.   
\item [\monob{{-}{-}maxinsertlen}\monoi{ $<$n$>$} ] Restrict insert length parameterization
such that the expected insert length at each position of the model is no
more than \monoi{$<$n$>$}.\monoi{}  

    
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmc2}}{hmmc2} - example client for the HMMER daemon   }
\subsection*{Synopsis}
\noindent
\monob{hmmc2} [\monoi{options}]
  
\subsection*{Description}
 

\mono{Hmmc2} is a text client for the hmmpgmd or hmmpgmd\_shard daemons.
 When run, it opens a connection to a daemon at the specified IP address
and port, and then enters an interactive loop waiting for the user to input
commands to be sent to the daemon. See the User's Guide for the HMMER Daemon
for a discussion of hmmpgmd's command format.  

  
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-i $<$IP address$>$} ] Specify the IP address of the daemon that hmmc2
should connect to.  Defaults to 127.0.0.1 if not provided   
\item [\monob{-p $<$port number$>$} ] Specify
the port number that the daemon is listening on.  Defaults to 51371 if not
provided   
\item [\monob{-S} ] Print the scores of any hits found during searches.   
\item [\monob{-A} ] Print
the alignment of any hits found during searches.  This is a superset of
the "-S" flag, so providing both is redundant.    
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmconvert}}{hmmconvert} - convert profile file to various formats   }
\subsection*{Synopsis}
\noindent
\monob{hmmconvert}
[\monoi{options}] \monoi{hmmfile}   
\subsection*{Description}
 

The \mono{hmmconvert } utility converts an input
profile file to different HMMER formats.  

By default, the input profile
can be in any HMMER format, including old/obsolete formats from HMMER2,
ASCII or binary; the output profile is a current HMMER3 ASCII format.  

\monoi{hmmfile}
may be '-' (dash), which means reading this input from stdin rather than a
file.   
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help; print a brief reminder of command line usage and
all available options.  
\item [\monob{-a} ] Output profiles in ASCII text format. This is the
default.  
\item [\monob{-b} ] Output profiles in binary format.   
\item [\monob{-2} ] Output in legacy HMMER2
ASCII text format, in ls (glocal) mode. This allows HMMER3 models to be
converted back to a close approximation of HMMER2, for comparative studies.
 
\item [\monob{{-}{-}outfmt}\monoi{ $<$s$>$} ] Output in a HMMER3 ASCII text format other then the most current
one. Valid choices for  \monoi{$<$s$>$} are \mono{3/a} through \mono{3/f}. The current format is \mono{3/f},
and this is the default. The format \mono{3/b } was used in the official HMMER3
release, and the others were used in the various testing versions.   
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmemit}}{hmmemit} - sample sequences from a profile   }
\subsection*{Synopsis}
\noindent
\monob{hmmemit} [\monoi{options}]
\monoi{hmmfile}   
\subsection*{Description}
 

The  \mono{hmmemit} program  samples (emits) sequences from
the profile HMM(s) in \monoi{hmmfile}, and writes them to output. Sampling sequences
may be useful for a variety of purposes, including creating synthetic true
positives for benchmarks or tests.  

The default is to sample one unaligned
sequence from the core probability model, which means that each sequence
consists of one full-length domain.  Alternatively, with the \mono{-c} option, you
can emit a simple majority-rule consensus sequence; or with the \mono{-a } option,
you can emit an alignment (in which case, you probably also want to set
 \mono{-N } to something other than its default of 1 sequence per model).  

As another
option, with the \mono{-p} option you can sample a sequence from a fully configured
HMMER search profile. This means sampling a `homologous sequence' by HMMER's
definition, including nonhomologous flanking sequences, local alignments,
and multiple domains per sequence, depending on the length model and alignment
mode chosen for the profile.  

The \monoi{hmmfile}  may contain a library of HMMs,
in which case each HMM will be used in turn.  

\monoi{hmmfile}  may be '-' (dash), which
means reading this input from stdin rather than a file.     
\subsection*{Common Options}

\begin{wideitem}
\item [\monob{-h} ] Help; print a brief reminder of command line usage and all available
options.   
\item [\monob{-o}\monoi{ $<$f$>$} ] Direct the output sequences to file \monoi{$<$f$>$}, rather than to stdout.
 
\item [\monob{-N}\monoi{ $<$n$>$} ] Sample \monoi{$<$n$>$} sequences per model, rather than just one.    
\end{wideitem}

\subsection*{Options Controlling
What to Emit}
 The default is to sample \mono{N} sequences from the core model. Alternatively,
you may choose one (and only one) of the following alternatives.   \begin{wideitem}
\item [\monob{-a} ] Emit
an alignment for each HMM in the  \monoi{hmmfile} rather than sampling unaligned
sequences one at a time.  
\item [\monob{-c} ] Emit a plurality-rule consensus sequence, instead
of sampling a sequence from the profile HMM's probability distribution. The
consensus sequence is formed by selecting the maximum probability residue
at each match state.  
\item [\monob{-C} ] Emit a fancier plurality-rule consensus sequence
than the \mono{-c} option. If the maximum probability residue has p $<$  \mono{minl} show
it as a lower case 'any' residue (n or x); if p $>$=  \mono{minl } and $<$  \mono{minu } show
it as a lower case residue; and if p $>$=  \mono{minu} show it as an upper case residue.
The default settings of  \mono{minu} and  \mono{minl } are both 0.0, which means \mono{-C } gives
the same output as  \mono{-c } unless you also set  \mono{minu} and \mono{minl } to what you
want.  
\item [\monob{-p} ] Sample unaligned sequences from the implicit search profile, not
from the core model.  The core model consists only of the homologous states
(between the begin and end states of a HMMER Plan7 model). The profile includes
the nonhomologous N, C, and J states, local/glocal and uni/multihit algorithm
configuration, and the target length model. Therefore sequences sampled
from a profile may include nonhomologous as well as homologous sequences,
and may contain more than one homologous sequence segment. By default, the
profile is in multihit local mode, and the target sequence length is configured
for L=400.      
\end{wideitem}

\subsection*{Options Controlling Emission from Profiles}
 These options
require that you have set the \mono{-p} option.  \begin{wideitem}
\item [\monob{-L}\monoi{ $<$n$>$} ] Configure the profile's target
sequence length model to generate a mean length of approximately $<$n$>$ rather
than the default of 400.  
\item [\monob{{-}{-}local} ] Configure the profile for multihit local
alignment.  
\item [\monob{{-}{-}unilocal} ] Configure the profile for unihit local alignment (Smith/Waterman).
 
\item [\monob{{-}{-}glocal} ] Configure the profile for multihit glocal alignment.  
\item [\monob{{-}{-}uniglocal}
] Configure the profile for unihit glocal alignment.   
\end{wideitem}

\subsection*{Options Controlling
Fancy Consensus Emission}
 These options require that you have set the \mono{-C}
option.  \begin{wideitem}
\item [\monob{{-}{-}minl}\monoi{ $<$x$>$} ] Sets the  \mono{minl} threshold for showing weakly conserved residues
as lower case. (0 $<$= x $<$= 1)  
\item [\monob{{-}{-}minu}\monoi{ $<$x$>$} ] Sets the  \mono{minu } threshold for showing
strongly conserved residues as upper case. (0 $<$= x $<$= 1)    
\end{wideitem}

\subsection*{Other Options}

\begin{wideitem}
\item [\monob{{-}{-}seed}\monoi{ $<$n$>$} ] Seed the random number generator with \monoi{$<$n$>$}, an integer $>$= 0.  If  \monoi{$<$n$>$} 
is nonzero, any stochastic simulations will be reproducible; the same command
will give the same results. If  \monoi{$<$n$>$} is 0, the random number generator is seeded
arbitrarily, and stochastic simulations will vary from run to run of the
same command. The default is 0: use an arbitrary seed, so different \mono{hmmemit}
runs will generate different samples.      
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmfetch}}{hmmfetch} - retrieve profiles from a file  }
\subsection*{Synopsis}
 

\noindent
\monob{hmmfetch} [\monoi{options}] \monoi{hmmfile key}

 (retrieve HMM named \monoi{key})

\noindent
\monob{hmmfetch -f }[\monoi{options}] \monoi{hmmfile keyfile}

 (retrieve all HMMs listed in \monoi{keyfile})

\noindent
\monob{hmmfetch {-}{-}index }[\monoi{options}] \monoi{hmmfile}

 (index \monoi{hmmfile} for fetching)

 
\subsection*{Description}
 

Quickly retrieves one or more profile HMMs from an \monoi{hmmfile}
(a large Pfam database, for example).   

For maximum speed, the  \monoi{hmmfile}
should be indexed first, using \mono{hmmfetch {-}{-}index}. The index is a binary file
named \monoi{hmmfile}.ssi. However, this is optional, and retrieval will still work
from unindexed files, albeit much more slowly.  

The default mode is to retrieve
a single profile by name or accession, called the \monoi{key}. For example:  

    \user{\% hmmfetch Pfam-A.hmm Caudal\_act}

    \user{\% hmmfetch Pfam-A.hmm PF00045}

 

With the \mono{-f} option, a  \monoi{keyfile}  containing a list of one or more keys is
read instead.  The first whitespace-delimited field on each non-blank non-comment
line of the \monoi{keyfile}  is used as a  \monoi{key}, and any remaining data on the line
is ignored. This allows a variety of whitespace delimited datafiles to be
used as a \monoi{keyfile}.  

When using \mono{-f } and a \monoi{keyfile}, if  \mono{hmmfile } has been indexed,
the keys are retrieved in the order they occur in the  \monoi{keyfile}, but if
 \mono{hmmfile } isn't indexed, keys are retrieved in the order they occur in the
 \mono{hmmfile}.\mono{} This is a side effect of an implementation that allows multiple
keys to be retrieved even if the \mono{hmmfile } is a nonrewindable stream, like
a standard input pipe.  

In normal use (without \mono{{-}{-}index} or  \mono{-f} options), \monoi{hmmfile}
 may be '-' (dash), which means reading input from stdin rather than a file.
  With the \mono{{-}{-}index} option,  \monoi{hmmfile} may not be '-'; it does not make sense to
index a standard input stream. With the  \mono{-f } option,   either  \monoi{hmmfile}  or
 \monoi{keyfile}  (but not both) may be '-'. It is often particularly useful to read
\monoi{keyfile} from standard input, because this allows use to use arbitrary command
line invocations to create a list of HMM names or accessions, then fetch
them all to a new file, just with one command.  

By default, fetched HMMs
are printed to standard output in HMMER3 format.   
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help; print
a brief reminder of command line usage and all available options.  
\item [\monob{-f} ] The
second commandline argument is a  \monoi{keyfile} instead of a single  \monoi{key}. The
first field on each line of the \monoi{keyfile}  is used as a retrieval  \monoi{key} (an
HMM name or accession).  Blank lines and comment lines (that start with
a \# character) are ignored.   
\item [\monob{-o}\monoi{ $<$f$>$} ] Output HMM(s) to file \monoi{$<$f$>$} instead of to
standard output.  
\item [\monob{-O} ] Output one retrieved HMM (by \monoi{key}) to a file named \monoi{key}.
This is a convenience for saving some typing: instead of  

 \user{\% hmmfetch -o RRM\_1 hmmfile RRM\_1}

you can just type 

 \user{\% hmmfetch -O hmmfile RRM\_1}

The \mono{-O } option only works if you're retrieving a single profile; it is incompatible
with  \mono{-f.}     
\item [\monob{{-}{-}index} ] Instead of retrieving one or more profiles from \monoi{hmmfile},
index the \monoi{hmmfile} for future retrievals. This creates a \monoi{hmmfile}.ssi binary
index file.    
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmlogo}}{hmmlogo} - produce a conservation logo graphic from a profile   }
\subsection*{Synopsis}
\noindent
\monob{hmmlogo}
[\monoi{options}] \monoi{hmmfile}   
\subsection*{Description}
 

\mono{hmmlogo } computes letter height and indel
parameters that can be used to  produce a profile HMM logo. This tool is
essentially a  command-line interface for much of the data underlying the
Skylign  logo server (skylign.org).  By default,  \mono{hmmlogo} prints out a table
of per-position letter heights (dependent on the  requested height method),
then prints a table of per-position gap probabilities.   In a typical logo,
the total height of a stack of letters for one position depends on the
information content of the position, and  that stack height is subdivided
according to the emission  probabilities of the letters of the alphabet.
 

  
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help; print a brief reminder of command line usage and all
available options.   
\item [\monob{{-}{-}height\_relent\_all} ] Total height = relative entropy (aka
information content); all letters  are given a positive height.  (default)
 
\item [\monob{{-}{-}height\_relent\_abovebg} ] Total height = relative entropy (aka information
content); only letters  with above-background probability are given positive
height.  
\item [\monob{{-}{-}height\_score} ] Total height = sums of scores of positive-scoring letters;
letter height depends on the score of that letter at that position. Only
 letters with above-background probability (positive score) are  given positive
height. (Note that only letter height is meaningful - stack height has no
inherent meaning).  
\item [\monob{{-}{-}no\_indel} ] Don't print out the indel probability table.
  
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmpgmd}}{hmmpgmd} - daemon for database search web services   }
\subsection*{Synopsis}
\noindent
\monob{hmmpgmd}
[\monoi{options}]   
\subsection*{Description}
 

The \mono{hmmpgmd } program is the daemon that we use
internally for the hmmer.org web server.  It essentially stands in front
of the search programs \mono{phmmer},\mono{} \mono{hmmsearch}, and  \mono{hmmscan}.\mono{}  

To use \mono{hmmpgmd},\mono{}
first an instance must be started up as a  master  server, and provided
with at least one  sequence database (using the  \mono{{-}{-}seqdb} flag) and/or an
 HMM database (using the \mono{{-}{-}hmmdb} flag).  A sequence database must be in hmmpgmd
format, which may be produced using  \mono{esl-reformat}. An HMM database is of
the form produced by  \mono{hmmbuild}. The input database(s) will be loaded into
memory by the  master. When the master has finished loading the database(s),
it  prints the line: "Data loaded into memory. Master is ready."   

After
the master is ready, one or more instances of hmmpgmd may be started as
workers. These workers may be (and typically are) on different machines
from the master, but must have access to the  same database file(s) provided
to the master, with the same path. As  with the master, each worker loads
the database(s) into memory, and  indicates completion by printing: "Data
loaded into memory. Worker is ready."   

The master process and workers are
expected to remain running. One or more clients then connect to the master
and submit possibly many queries. The master distributes the work of a query
among the workers, collects results, and merges them before responding
to the client. Two example client programs are included in the HMMER src
 directory - the C program \mono{hmmc2} and the perl script \mono{hmmpgmd\_client\_example.pl}.
These are intended as examples only, and should be extended as  necessary
to meet your needs.   

A query is submitted to the master from the client
as a character string. Queries may be the sort that would normally be handled
by  \mono{phmmer} (protein sequence vs protein database), \mono{hmmsearch} (protein HMM
query vs protein database), or \mono{hmmscan} (protein query vs protein HMM database).
 

  The general form of a client query is to start with a single line of
the form  \mono{@[options]},\mono{} followed by multiple lines of text representing either
the query HMM  or fasta-formatted sequence. The final line of each query
is the separator  \mono{//}.   

For example, to perform a  \mono{phmmer} type search of
a sequence against a sequence database  file, the first line is of the
form  \mono{@{-}{-}seqdb 1}, then the fasta-formatted query sequence starting with the
header line \mono{$>$sequence-name}, followed by one or more lines of sequence, and
finally the closing \mono{//}.  

To perform an \mono{hmmsearch } type search, the query
sequence is replaced by the full text of a HMMER-format query HMM.   

To perform
an \mono{hmmscan } type search, the text matches that of the  \mono{phmmer} type search,
except that the first line changes to  \mono{@{-}{-}hmmdb 1}.  

In the hmmpgmd-formatted
sequence database file, each sequence can be associated with one or more
sub-databases. The  \mono{{-}{-}seqdb} flag indicates which of these sub-databases will
be queried.  The HMM database format does not support sub-databases.    

  
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help; print a brief reminder of command line usage and all
available options.  
\item [\monob{{-}{-}master} ] Run as the master server.  
\item [\monob{{-}{-}worker}\monoi{ $<$s$>$} ] Run as a worker,
connecting to the master server that is running on IP address \monoi{$<$s$>$}.  
\item [\monob{{-}{-}cport}\monoi{
$<$n$>$} ] Port to use for communication between clients and the master server. 
The default is 51371.  
\item [\monob{{-}{-}wport}\monoi{ $<$n$>$} ] Port to use for communication between workers
and the master server.  The default is 51372.  
\item [\monob{{-}{-}ccncts}\monoi{ $<$n$>$} ] Maximum number of
client connections to accept. The default is 16.  
\item [\monob{{-}{-}wcncts}\monoi{ $<$n$>$} ] Maximum number
of worker connections to accept. The default is 32.  
\item [\monob{{-}{-}pid}\monoi{ $<$f$>$} ] Name of file into
which the process id will be written.   
\item [\monob{{-}{-}seqdb}\monoi{ $<$f$>$} ] Name of the file (in \mono{hmmpgmd}
format) containing protein sequences. The contents of this file will be
cached for searches.   
\item [\monob{{-}{-}hmmdb}\monoi{ $<$f$>$} ] Name of the file containing protein HMMs.
The contents of this file  will be cached for searches.  
\item [\monob{{-}{-}cpu}\monoi{ $<$n$>$} ] Number of
parallel threads to use (for  \mono{{-}{-}worker} ).   
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmpgmd\_shard}}{hmmpgmd\_shard} - sharded daemon for database search web services   }

\subsection*{Synopsis}
\noindent
\monob{hmmpgmd\_shard} [\monoi{options}]   
\subsection*{Description}
 

The  \mono{hmmpgmd\_shard } program
provides a sharded version of the  \mono{hmmpgmd } program that we use internally
to implement high-performance HMMER services that can be accessed via the
internet.  See the  \mono{hmmpgmd} man page for a discussion of how the base  \mono{hmmpgmd}
program is used.  This man page discusses differences between  \mono{hmmpgmd\_shard}
and  \mono{hmmpgmd.   } The base  \mono{hmmpgmd} program loads the entirety of its database
file into RAM on every worker node, in spite of the fact that each worker
node searches a predictable fraction of the database(s) contained in that
file when performing searches.  This wastes RAM, particularly when many
worker nodes are used to accelerate searches of large databases.  

\mono{Hmmpgmd\_shard
} addresses this by dividing protein sequence database files into shards.
 Each worker node loads only 1/Nth of the database file, where N is the
number of worker nodes attached to the master.  HMM database files are not
sharded, meaning that every worker node will load the entire database file
into RAM.  Current HMM databases are much smaller than current protein sequence
databases, and easily fit into the RAM of modern servers even without sharding.
 

\mono{Hmmpgmd\_shard } is used in the same manner as  \mono{hmmpgmd} , except that it
takes one additional argument:  \mono{{-}{-}num\_shards}\monoi{ $<$n$>$} , which specifies the number
of shards that protein databases will be divided into, and defaults to
1 if unspecified.  This argument is only valid for the master node of a
 \mono{hmmpgmd} system (i.e., when  \mono{{-}{-}master} is passed to the  \mono{hmmpgmd} program), and
must be equal to the number of worker nodes that will connect to the master
node.   \mono{Hmmpgmd\_shard } will signal an error if more than  \mono{num\_shards} worker
nodes attempt to connect to the master node or if a search is started when
fewer than  \mono{num\_shards}\monoi{} workers are connected to the master.  
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help;
print a brief reminder of command line usage and all available options.
 
\item [\monob{{-}{-}master} ] Run as the master server.  
\item [\monob{{-}{-}worker}\monoi{ $<$s$>$} ] Run as a worker, connecting
to the master server that is running on IP address \monoi{$<$s$>$}.  
\item [\monob{{-}{-}cport}\monoi{ $<$n$>$} ] Port to use
for communication between clients and the master server.  The default is
51371.  
\item [\monob{{-}{-}wport}\monoi{ $<$n$>$} ] Port to use for communication between workers and the master
server.  The default is 51372.  
\item [\monob{{-}{-}ccncts}\monoi{ $<$n$>$} ] Maximum number of client connections
to accept. The default is 16.  
\item [\monob{{-}{-}wcncts}\monoi{ $<$n$>$} ] Maximum number of worker connections
to accept. The default is 32.  
\item [\monob{{-}{-}pid}\monoi{ $<$f$>$} ] Name of file into which the process
id will be written.   
\item [\monob{{-}{-}seqdb}\monoi{ $<$f$>$} ] Name of the file (in \mono{hmmpgmd} format) containing
protein sequences. The contents of this file will be cached for searches.
  
\item [\monob{{-}{-}hmmdb}\monoi{ $<$f$>$} ] Name of the file containing protein HMMs. The contents of this
file  will be cached for searches.  
\item [\monob{{-}{-}cpu}\monoi{ $<$n$>$} ] Number of parallel threads to
use (for  \mono{{-}{-}worker} ).  
\item [\monob{{-}{-}num\_shards}\monoi{ $<$n$>$} ] Number of shards to divide cached sequence
database(s) into.  HMM databases are not sharded, due to their small size.
This option is only valid when the  \mono{{-}{-}master } option is present, and defaults
to 1 if not specified. \mono{Hmmpgmd\_shard } requires that the number of shards
be equal to the number of worker nodes, and will give errors if more than
 \mono{num\_shards}\monoi{} workers attempt to connect to the master node or if a search
is started with fewer than  \mono{num\_shards}\monoi{} workers connected to the master.
 
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmpress}}{hmmpress} - prepare a profile database for hmmscan   }
\subsection*{Synopsis}
\noindent
 \monob{hmmpress}
[\monoi{options}] \monoi{hmmfile}   
\subsection*{Description}
 

Constructs binary compressed datafiles
for  \mono{hmmscan}, starting from a profile database \monoi{hmmfile} in standard HMMER3
format. The  \mono{hmmpress} step is required for \mono{hmmscan} to work.  

Four files are
created: \monoi{hmmfile}\mono{.h3m,} \monoi{hmmfile}\mono{.h3i,} \monoi{hmmfile}\mono{.h3f,} and \monoi{hmmfile}\mono{.h3p.} The  \monoi{hmmfile}\mono{.h3m}
file contains the profile HMMs and their annotation in a binary format.
The  \monoi{hmmfile}\mono{.h3i} file is an SSI index for the \monoi{hmmfile}\mono{.h3m} file. The \monoi{hmmfile}\mono{.h3f}
file contains precomputed data structures for the fast heuristic filter
(the MSV filter). The \monoi{hmmfile}\mono{.h3p} file contains precomputed data structures
for the rest of each profile.  

\monoi{hmmfile} may not be '-' (dash); running \mono{hmmpress}
on a standard input stream rather than a file is not allowed.   
\subsection*{Options}

\begin{wideitem}
\item [\monob{-h} ] Help; print a brief reminder of command line usage and all available
options.  
\item [\monob{-f} ] Force; overwrites any previous hmmpress'ed datafiles. The default
is to bitch about any existing files and ask you to delete them first. 
   
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmscan}}{hmmscan} - search sequence(s) against a profile database   }
\subsection*{Synopsis}
\noindent
\monob{hmmscan}
[\monoi{options}] \monoi{hmmdb} \monoi{seqfile}    
\subsection*{Description}
 

\mono{hmmscan } is used to search protein
sequences against collections  of protein profiles. For each sequence in
 \monoi{seqfile}, use that query sequence to search the target database of profiles
in \monoi{hmmdb}, and output ranked lists of the profiles with the most significant
matches to the sequence.  

The  \monoi{seqfile}  may contain more than one query
sequence. Each will be searched in turn against \monoi{hmmdb.}  

The \monoi{hmmdb} needs to
be press'ed using  \mono{hmmpress} before it can be searched with  \mono{hmmscan}.\mono{} This
creates four binary files, suffixed \mono{.h3\{fimp\}}.  

The query \monoi{seqfile}  may be
'-' (a dash character), in which case the query sequences are read from a
stdin pipe instead of from a file. The \monoi{hmmdb}  cannot be read from a stdin
stream, because it needs to have those four auxiliary binary files generated
by  \mono{hmmpress}.  

The output format is designed to be human-readable, but is
often so voluminous that reading it is impractical, and parsing it is a
pain. The \mono{{-}{-}tblout } and  \mono{{-}{-}domtblout } options save output in simple tabular
formats that are concise and easier to parse.  The  \mono{-o} option allows redirecting
the main output, including throwing it away in /dev/null.    
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help;
print a brief reminder of command line usage and all available options.
   
\end{wideitem}

\subsection*{Options for Controlling Output}
 \begin{wideitem}
\item [\monob{-o}\monoi{ $<$f$>$} ] Direct the main human-readable output
to a file \monoi{$<$f$>$}  instead of the default stdout.  
\item [\monob{{-}{-}tblout}\monoi{ $<$f$>$} ] Save a simple tabular
(space-delimited) file summarizing the per-target output, with one data line
per homologous target model found.  
\item [\monob{{-}{-}domtblout}\monoi{ $<$f$>$} ] Save a simple tabular (space-delimited)
file summarizing the per-domain output, with one data line per homologous
domain detected in a query sequence for each homologous model.  
\item [\monob{{-}{-}pfamtblout}\monoi{
$<$f$>$} ] Save an especially succinct tabular (space-delimited) file  summarizing
the per-target output, with one data line per  homologous target model found.
  
\item [\monob{{-}{-}acc} ] Use accessions instead of names in the main output, where available
for profiles and/or sequences.  
\item [\monob{{-}{-}noali} ] Omit the alignment section from the
main output. This can greatly reduce the output volume.  
\item [\monob{{-}{-}notextw} ] Unlimit
the length of each line in the main output. The default is a limit of 120
characters per line, which helps in displaying the output cleanly on terminals
and in editors, but can truncate target profile description lines.  
\item [\monob{{-}{-}textw}\monoi{
$<$n$>$} ] Set the main output's line length limit to \monoi{$<$n$>$} characters per line. The default
is 120.    
\end{wideitem}

\subsection*{Options for Reporting Thresholds}
 Reporting thresholds control
which hits are reported in output files (the main output, \mono{{-}{-}tblout}, and 
\mono{{-}{-}domtblout}).  \begin{wideitem}
\item [\monob{-E}\monoi{ $<$x$>$} ] In the per-target output, report target profiles with an
E-value of $<$= \monoi{$<$x$>$}.\monoi{} The default is 10.0, meaning that on average, about 10 false
positives will be reported per query, so you can see the top of the noise
and decide for yourself if it's really noise.  
\item [\monob{-T}\monoi{ $<$x$>$} ] Instead of thresholding
per-profile output on E-value, instead report target profiles with a bit
score of $>$= \monoi{$<$x$>$}.  
\item [\monob{{-}{-}domE}\monoi{ $<$x$>$} ] In the per-domain output, for target profiles that
have already satisfied the per-profile reporting threshold, report individual
domains with a conditional E-value of $<$= \monoi{$<$x$>$}.\monoi{} The default is 10.0.  A conditional
E-value means the expected number of additional false positive domains in
the smaller search space of those comparisons that already satisfied the
per-profile reporting threshold (and thus must have at least one homologous
domain already).   
\item [\monob{{-}{-}domT}\monoi{ $<$x$>$} ] Instead of thresholding per-domain output on E-value,
instead report domains with a bit score of $>$= \monoi{$<$x$>$}.     
\end{wideitem}

\subsection*{Options for Inclusion
Thresholds}
 Inclusion thresholds are stricter than reporting thresholds.
Inclusion thresholds control which hits are considered to be reliable enough
to be included in an output alignment or a subsequent search round. In 
\mono{hmmscan}, which does not have any alignment output (like  \mono{hmmsearch} or 
\mono{phmmer})\mono{} nor any iterative search steps (like  \mono{jackhmmer}), inclusion thresholds
have little effect. They only affect what domains get marked as significant
(!) or questionable (?) in domain output.   \begin{wideitem}
\item [\monob{{-}{-}incE}\monoi{ $<$x$>$} ] Use an E-value of $<$= \monoi{$<$x$>$}
as the per-target inclusion threshold. The default is 0.01, meaning that on
average, about 1 false positive would be expected in every 100 searches
with different query sequences.  
\item [\monob{{-}{-}incT}\monoi{ $<$x$>$} ] Instead of using E-values for setting
the inclusion threshold, instead use a bit score of $>$=  \monoi{$<$x$>$} as the per-target
inclusion threshold. It would be unusual to use bit score thresholds with
\monoi{hmmscan}, because you don't expect a single score threshold to work for different
profiles; different profiles have slightly different expected score distributions.
 
\item [\monob{{-}{-}incdomE}\monoi{ $<$x$>$} ] Use a conditional E-value of $<$= \monoi{$<$x$>$}  as the per-domain inclusion
threshold, in targets that have already satisfied the overall per-target
inclusion threshold. The default is 0.01.  
\item [\monob{{-}{-}incdomT}\monoi{ $<$x$>$} ] Instead of using E-values,
instead use a bit score of $>$= \monoi{$<$x$>$} as the per-domain inclusion threshold. As
with  \mono{{-}{-}incT } above, it would be unusual to use a single bit score threshold
in \mono{hmmscan}.    
\end{wideitem}

\subsection*{Options for Model-specific Score Thresholding}
 Curated profile
databases may define specific bit score thresholds for each profile, superseding
any thresholding based on statistical significance alone.  To use these
options, the profile must contain the appropriate (GA, TC, and/or NC) optional
score threshold annotation; this is picked up by  \mono{hmmbuild} from Stockholm
format alignment files. Each thresholding option has two scores: the per-sequence
threshold \monoi{$<$x1$>$} and the per-domain threshold \monoi{$<$x2$>$}. These act as if \mono{-T}\monoi{ $<$x1$>$} \mono{{-}{-}incT}\monoi{ $<$x1$>$}
\mono{{-}{-}domT}\monoi{ $<$x2$>$} \mono{{-}{-}incdomT}\monoi{ $<$x2$>$} has been applied specifically using each model's curated
thresholds.  \begin{wideitem}
\item [\monob{{-}{-}cut\_ga} ] Use the GA (gathering) bit scores in the model to set
per-sequence (GA1) and per-domain (GA2) reporting and inclusion thresholds.
GA thresholds are generally considered to be the reliable curated thresholds
defining family membership; for example, in Pfam, these thresholds define
what gets included in Pfam Full alignments based on searches with Pfam
Seed models.  
\item [\monob{{-}{-}cut\_nc} ] Use the NC (noise cutoff) bit score thresholds in the
model to set per-sequence (NC1) and per-domain (NC2) reporting and inclusion
thresholds. NC thresholds are generally considered to be the score of the
highest-scoring known false positive.  
\item [\monob{{-}{-}cut\_tc} ] Use the NC (trusted cutoff)
bit score thresholds in the model to set per-sequence (TC1) and per-domain
(TC2) reporting and inclusion thresholds. TC thresholds are generally considered
to be the score of the lowest-scoring known true positive that is above
all known false positives.      
\end{wideitem}

\subsection*{Control of the Acceleration Pipeline}
 HMMER3
searches are accelerated in a three-step filter pipeline: the MSV filter,
the Viterbi filter, and the Forward filter. The first filter is the fastest
and most approximate; the last is the full Forward scoring algorithm. There
is also a bias filter step between MSV and Viterbi. Targets that pass all
the steps in the acceleration pipeline are then subjected to postprocessing
{-}{-} domain identification and scoring using the Forward/Backward algorithm.
 Changing filter thresholds only removes or includes targets from consideration;
changing filter thresholds does not alter bit scores, E-values, or alignments,
all of which are determined solely in postprocessing.  \begin{wideitem}
\item [\monob{{-}{-}max} ] Turn off all
filters, including the bias filter, and run full Forward/Backward postprocessing
on every target. This increases sensitivity somewhat, at a large cost in
speed.  
\item [\monob{{-}{-}F1}\monoi{ $<$x$>$} ] Set the P-value threshold for the MSV filter step.  The default
is 0.02, meaning that roughly 2\% of the highest scoring nonhomologous targets
are expected to pass the filter.  
\item [\monob{{-}{-}F2}\monoi{ $<$x$>$} ] Set the P-value threshold for the
Viterbi filter step. The default is 0.001.   
\item [\monob{{-}{-}F3}\monoi{ $<$x$>$} ] Set the P-value threshold
for the Forward filter step. The default is 1e-5.  
\item [\monob{{-}{-}nobias} ] Turn off the bias
filter. This increases sensitivity somewhat, but can come at a high cost
in speed, especially if the query has biased residue composition (such
as a repetitive sequence region, or if it is a membrane protein with large
regions of hydrophobicity). Without the bias filter, too many sequences
may pass the filter with biased queries, leading to slower than expected
performance as the computationally intensive Forward/Backward algorithms
shoulder an abnormally heavy load.    
\end{wideitem}

\subsection*{Other Options}
 \begin{wideitem}
\item [\monob{{-}{-}nonull2} ] Turn off the
null2 score corrections for biased composition.  
\item [\monob{-Z}\monoi{ $<$x$>$} ] Assert that the total
number of targets in your searches is \monoi{$<$x$>$}, for the purposes of per-sequence
E-value calculations, rather than the actual number of targets seen.   
\item [\monob{{-}{-}domZ}\monoi{
$<$x$>$} ] Assert that the total number of targets in your searches is \monoi{$<$x$>$}, for the
purposes of per-domain conditional E-value calculations, rather than the
number of targets that passed the reporting thresholds.  
\item [\monob{{-}{-}seed}\monoi{ $<$n$>$} ] Set the
random number seed to  \monoi{$<$n$>$}. Some steps in postprocessing require Monte Carlo
simulation.  The default is to use a fixed seed (42), so that results are
exactly reproducible. Any other positive integer will give different (but
also reproducible) results. A choice of 0 uses an arbitrarily chosen seed.
 
\item [\monob{{-}{-}qformat}\monoi{ $<$s$>$} ] Assert that input \monoi{seqfile} is in format \monoi{$<$s$>$}, bypassing format autodetection.
Common choices for  \monoi{$<$s$>$}  include: \mono{fasta}, \mono{embl}, \mono{genbank.} Alignment formats
also work; common choices include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal},
\mono{phylip}. For more information, and for codes for some less common formats,
see main documentation. The string \monoi{$<$s$>$} is case-insensitive (\mono{fasta} or \mono{FASTA}
both work).    
\item [\monob{{-}{-}cpu}\monoi{ $<$n$>$} ] Set the number of parallel worker threads to  \monoi{$<$n$>$}. On
multicore machines, the default is 2. You can also control this number by
setting an environment variable,  \monoi{HMMER\_NCPU}. There is also a master thread,
so the actual number of threads that HMMER spawns is \monoi{$<$n$>$}+1.  This option is
not available if HMMER was compiled with POSIX threads support turned off.
  
\item [\monob{{-}{-}stall} ] For debugging the MPI master/worker version: pause after start,
to enable the developer to attach debuggers to the running master and worker(s)
processes. Send SIGCONT signal to release the pause. (Under gdb:  \mono{(gdb) signal
SIGCONT})  (Only available if optional MPI support was enabled at compile-time.)
 
\item [\monob{{-}{-}mpi} ] Run under MPI control with master/worker parallelization (using \mono{mpirun},
for example, or equivalent). Only available if optional MPI support was
enabled at compile-time.     
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmsearch}}{hmmsearch} - search profile(s) against a sequence database   }
\subsection*{Synopsis}
\noindent
\monob{hmmsearch}
[\monoi{options}] \monoi{hmmfile} \monoi{seqdb}   
\subsection*{Description}
 

\mono{hmmsearch } is used to search one
or more profiles against a sequence database. For each profile in  \monoi{hmmfile},
use that query profile to search the target database of sequences in \monoi{seqdb},
and output ranked lists of the sequences with the most significant matches
to the profile. To build profiles from multiple alignments, see \mono{hmmbuild}.
 

Either the query \monoi{hmmfile} or the target \monoi{seqdb}  may be '-' (a dash character),
in which case the query profile or target database input will be read from
a stdin pipe instead of from a file. Only one input source can come through
stdin, not both. An exception is that if the \monoi{hmmfile}  contains more than
one profile query, then \monoi{seqdb}  cannot come from stdin, because we can't
rewind the streaming target database to search it with another profile.
  

The output format is designed to be human-readable, but is often so voluminous
that reading it is impractical, and parsing it is a pain. The \mono{{-}{-}tblout } and
 \mono{{-}{-}domtblout } options save output in simple tabular formats that are concise
and easier to parse. The  \mono{-o} option allows redirecting the main output, including
throwing it away in /dev/null.    
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help; print a brief reminder
of command line usage and all available options.    
\end{wideitem}

\subsection*{Options for Controlling
Output}
 \begin{wideitem}
\item [\monob{-o}\monoi{ $<$f$>$} ] Direct the main human-readable output to a file \monoi{$<$f$>$}  instead of
the default stdout.  
\item [\monob{-A}\monoi{ $<$f$>$} ] Save a multiple alignment of all significant hits
(those satisfying \monoi{inclusion thresholds}) to the file  \monoi{$<$f$>$}.  
\item [\monob{{-}{-}tblout}\monoi{ $<$f$>$} ] Save a
simple tabular (space-delimited) file summarizing the per-target output,
with one data line per homologous target sequence found.  
\item [\monob{{-}{-}domtblout}\monoi{ $<$f$>$} ] Save
a simple tabular (space-delimited) file summarizing the per-domain output,
with one data line per homologous domain detected in a query sequence for
each homologous model.  
\item [\monob{{-}{-}acc} ] Use accessions instead of names in the main
output, where available for profiles and/or sequences.  
\item [\monob{{-}{-}noali} ] Omit the alignment
section from the main output. This can greatly reduce the output volume.
 
\item [\monob{{-}{-}notextw} ] Unlimit the length of each line in the main output. The default
is a limit of 120 characters per line, which helps in displaying the output
cleanly on terminals and in editors, but can truncate target profile description
lines.  
\item [\monob{{-}{-}textw}\monoi{ $<$n$>$} ] Set the main output's line length limit to \monoi{$<$n$>$} characters per
line. The default is 120.    
\end{wideitem}

\subsection*{Options Controlling Reporting Thresholds}
 Reporting
thresholds control which hits are reported in output files (the main output,
\mono{{-}{-}tblout}, and  \mono{{-}{-}domtblout}). Sequence hits and domain hits are ranked by statistical
significance (E-value) and output is generated in two sections called per-target
and per-domain output. In per-target output, by default, all sequence hits
with an E-value $<$= 10 are reported. In the per-domain output, for each target
that has passed per-target reporting thresholds, all domains satisfying
per-domain reporting thresholds are reported. By default, these are domains
with conditional E-values of $<$= 10. The following options allow you to change
the default E-value reporting thresholds, or to use bit score thresholds
instead.   \begin{wideitem}
\item [\monob{-E}\monoi{ $<$x$>$} ] In the per-target output, report target sequences with an
E-value of $<$= \monoi{$<$x$>$}.\monoi{} The default is 10.0, meaning that on average, about 10 false
positives will be reported per query, so you can see the top of the noise
and decide for yourself if it's really noise.  
\item [\monob{-T}\monoi{ $<$x$>$} ] Instead of thresholding
per-profile output on E-value, instead report target sequences with a bit
score of $>$= \monoi{$<$x$>$}.  
\item [\monob{{-}{-}domE}\monoi{ $<$x$>$} ] In the per-domain output, for target sequences that
have already satisfied the per-profile reporting threshold, report individual
domains with a conditional E-value of $<$= \monoi{$<$x$>$}.\monoi{} The default is 10.0.  A conditional
E-value means the expected number of additional false positive domains in
the smaller search space of those comparisons that already satisfied the
per-target reporting threshold (and thus must have at least one homologous
domain already).   
\item [\monob{{-}{-}domT}\monoi{ $<$x$>$} ] Instead of thresholding per-domain output on E-value,
instead report domains with a bit score of $>$= \monoi{$<$x$>$}.     
\end{wideitem}

\subsection*{Options for Inclusion
Thresholds}
 Inclusion thresholds are stricter than reporting thresholds.
Inclusion thresholds control which hits are considered to be reliable enough
to be included in an output alignment or a subsequent search round, or
marked as significant ("!") as opposed to questionable ("?") in domain
output.  \begin{wideitem}
\item [\monob{{-}{-}incE}\monoi{ $<$x$>$} ] Use an E-value of $<$= \monoi{$<$x$>$} as the per-target inclusion threshold.
The default is 0.01, meaning that on average, about 1 false positive would
be expected in every 100 searches with different query sequences.  
\item [\monob{{-}{-}incT}\monoi{
$<$x$>$} ] Instead of using E-values for setting the inclusion threshold, instead
use a bit score of $>$=  \monoi{$<$x$>$} as the per-target inclusion threshold. By default
this option is unset.  
\item [\monob{{-}{-}incdomE}\monoi{ $<$x$>$} ] Use a conditional E-value of $<$= \monoi{$<$x$>$}  as the
per-domain inclusion threshold, in targets that have already satisfied the
overall per-target inclusion threshold. The default is 0.01.  
\item [\monob{{-}{-}incdomT}\monoi{ $<$x$>$} ] Instead
of using E-values, use a bit score of $>$= \monoi{$<$x$>$} as the per-domain inclusion threshold.
   
\end{wideitem}

\subsection*{Options for Model-specific Score Thresholding}
 Curated profile databases
may define specific bit score thresholds for each profile, superseding
any thresholding based on statistical significance alone.  To use these
options, the profile must contain the appropriate (GA, TC, and/or NC) optional
score threshold annotation; this is picked up by  \mono{hmmbuild} from Stockholm
format alignment files. Each thresholding option has two scores: the per-sequence
threshold $<$x1$>$ and the per-domain threshold $<$x2$>$ These act as if \mono{-T}\monoi{ $<$x1$>$} \mono{{-}{-}incT}\monoi{ $<$x1$>$}
\mono{{-}{-}domT}\monoi{ $<$x2$>$} \mono{{-}{-}incdomT}\monoi{ $<$x2$>$} has been applied specifically using each model's curated
thresholds.  \begin{wideitem}
\item [\monob{{-}{-}cut\_ga} ] Use the GA (gathering) bit scores in the model to set
per-sequence (GA1) and per-domain (GA2) reporting and inclusion thresholds.
GA thresholds are generally considered to be the reliable curated thresholds
defining family membership; for example, in Pfam, these thresholds define
what gets included in Pfam Full alignments based on searches with Pfam
Seed models.  
\item [\monob{{-}{-}cut\_nc} ] Use the NC (noise cutoff) bit score thresholds in the
model to set per-sequence (NC1) and per-domain (NC2) reporting and inclusion
thresholds. NC thresholds are generally considered to be the score of the
highest-scoring known false positive.  
\item [\monob{{-}{-}cut\_tc} ] Use the TC (trusted cutoff)
bit score thresholds in the model to set per-sequence (TC1) and per-domain
(TC2) reporting and inclusion thresholds. TC thresholds are generally considered
to be the score of the lowest-scoring known true positive that is above
all known false positives.      
\end{wideitem}

\subsection*{Options Controlling the Acceleration Pipeline}

HMMER3 searches are accelerated in a three-step filter pipeline: the MSV
filter, the Viterbi filter, and the Forward filter. The first filter is
the fastest and most approximate; the last is the full Forward scoring
algorithm. There is also a bias filter step between MSV and Viterbi. Targets
that pass all the steps in the acceleration pipeline are then subjected
to postprocessing {-}{-} domain identification and scoring using the Forward/Backward
algorithm.  Changing filter thresholds only removes or includes targets
from consideration; changing filter thresholds does not alter bit scores,
E-values, or alignments, all of which are determined solely in postprocessing.
 \begin{wideitem}
\item [\monob{{-}{-}max} ] Turn off all filters, including the bias filter, and run full Forward/Backward
postprocessing on every target. This increases sensitivity somewhat, at
a large cost in speed.  
\item [\monob{{-}{-}F1}\monoi{ $<$x$>$} ] Set the P-value threshold for the MSV filter
step.  The default is 0.02, meaning that roughly 2\% of the highest scoring
nonhomologous targets are expected to pass the filter.  
\item [\monob{{-}{-}F2}\monoi{ $<$x$>$} ] Set the P-value
threshold for the Viterbi filter step. The default is 0.001.   
\item [\monob{{-}{-}F3}\monoi{ $<$x$>$} ] Set the
P-value threshold for the Forward filter step. The default is 1e-5.  
\item [\monob{{-}{-}nobias}
] Turn off the bias filter. This increases sensitivity somewhat, but can come
at a high cost in speed, especially if the query has biased residue composition
(such as a repetitive sequence region, or if it is a membrane protein with
large regions of hydrophobicity). Without the bias filter, too many sequences
may pass the filter with biased queries, leading to slower than expected
performance as the computationally intensive Forward/Backward algorithms
shoulder an abnormally heavy load.    
\end{wideitem}

\subsection*{Other Options}
 \begin{wideitem}
\item [\monob{{-}{-}nonull2} ] Turn off the
null2 score corrections for biased composition.  
\item [\monob{-Z}\monoi{ $<$x$>$} ] Assert that the total
number of targets in your searches is \monoi{$<$x$>$}, for the purposes of per-sequence
E-value calculations, rather than the actual number of targets seen.   
\item [\monob{{-}{-}domZ}\monoi{
$<$x$>$} ] Assert that the total number of targets in your searches is \monoi{$<$x$>$}, for the
purposes of per-domain conditional E-value calculations, rather than the
number of targets that passed the reporting thresholds.  
\item [\monob{{-}{-}seed}\monoi{ $<$n$>$} ] Set the
random number seed to  \monoi{$<$n$>$}. Some steps in postprocessing require Monte Carlo
simulation.  The default is to use a fixed seed (42), so that results are
exactly reproducible. Any other positive integer will give different (but
also reproducible) results. A choice of 0 uses a randomly chosen seed.  
\item [\monob{{-}{-}tformat}\monoi{
$<$s$>$} ] Assert that target sequence file \monoi{seqfile} is in format \monoi{$<$s$>$}, bypassing format
autodetection. Common choices for  \monoi{$<$s$>$}  include: \mono{fasta}, \mono{embl}, \mono{genbank.} Alignment
formats also work; common choices include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast},
\mono{clustal}, \mono{phylip}. For more information, and for codes for some less common
formats, see main documentation. The string \monoi{$<$s$>$} is case-insensitive (\mono{fasta}
or \mono{FASTA} both work).  
\item [\monob{{-}{-}cpu}\monoi{ $<$n$>$} ] Set the number of parallel worker threads to
 \monoi{$<$n$>$}. On multicore machines, the default is 2. You can also control this number
by setting an environment variable,  \monoi{HMMER\_NCPU}. There is also a master
thread, so the actual number of threads that HMMER spawns is \monoi{$<$n$>$}+1.  This
option is not available if HMMER was compiled with POSIX threads support
turned off.   
\item [\monob{{-}{-}stall} ] For debugging the MPI master/worker version: pause after
start, to enable the developer to attach debuggers to the running master
and worker(s) processes. Send SIGCONT signal to release the pause. (Under
gdb:  \mono{(gdb) signal SIGCONT}) (Only available if optional MPI support was
enabled at compile-time.)    
\item [\monob{{-}{-}mpi} ] Run under MPI control with master/worker
parallelization (using \mono{mpirun}, for example, or equivalent). Only available
if optional MPI support was enabled at compile-time.       
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmsim}}{hmmsim} - collect profile score distributions on random sequences  }
\subsection*{Synopsis}
\noindent
\monob{hmmsim}
[\monoi{options}] \monoi{hmmfile}   
\subsection*{Description}
 

The  \mono{hmmsim } program generates random sequences,
scores them with the model(s) in  \monoi{hmmfile}, and outputs various sorts of
histograms, plots, and fitted distributions for the resulting scores.  

\mono{hmmsim}
is not a mainstream part of the HMMER package and most users would have
no reason to use it. It is used to develop and test the statistical methods
used to determine P-values and E-values in HMMER3. For example, it was used
to generate most of the results in a 2008 paper on H3's local alignment
statistics (PLoS Comp Bio 4:e1000069, 2008; http://www.ploscompbiol.org/doi/pcbi.1000069).
  

Because it is a research testbed, you should not expect it to be as robust
as other programs in the package. For example, options may interact in weird
ways; we haven't tested nor tried to anticipate all different possible combinations.
 

The main task is to fit a maximum likelihood Gumbel distribution to Viterbi
scores or an maximum likelihood exponential tail to high-scoring Forward
scores, and to test that these fitted distributions obey the conjecture
that lambda ~ log\_2 for both the Viterbi Gumbel and the Forward exponential
tail.   

The output is a table of numbers, one row for each model. Four different
parametric fits to the score data are tested: (1) maximum likelihood fits
to both location (mu/tau) and slope (lambda) parameters; (2) assuming lambda=log\_2,
maximum likelihood fit to the location parameter only; (3) same but assuming
an edge-corrected lambda, using current procedures in H3 [Eddy, 2008]; \textsf{and
(4)} using both parameters determined by H3's current procedures. The standard
simple, quick and dirty statistic for goodness-of-fit is 'E@10', the calculated
E-value of the 10th ranked top hit, which we expect to be about 10.   

In
detail, the columns of the output are:  \begin{wideitem}
\item [\monob{name} ] Name of the model.  
\item [\monob{tailp} ] Fraction
of the highest scores used to fit the distribution. For Viterbi, MSV, and
Hybrid scores, this defaults to 1.0 (a Gumbel distribution is fitted to
all the data). For Forward scores, this defaults to 0.02 (an exponential
tail is fitted to the highest 2\% scores).  
\item [\monob{mu/tau} ] Location parameter for
the maximum likelihood fit to the data.  
\item [\monob{lambda} ] Slope parameter for the
maximum likelihood fit to the data.  
\item [\monob{E@10} ] The E-value calculated for the
10th ranked high score ('E@10') using the ML mu/tau and lambda. By definition,
this expected to be about 10, if E-value estimation were accurate.  
\item [\monob{mufix}
] Location parameter, for a maximum likelihood fit with a known (fixed) slope
parameter lambda of log\_2 (0.693).  
\item [\monob{E@10fix} ] The E-value calculated for the
10th ranked score using mufix and the expected lambda = log\_2 = 0.693.  

\item [\monob{mufix2} ] Location parameter, for a maximum likelihood fit with an edge-effect-corrected
lambda.  
\item [\monob{E@10fix2} ] The E-value calculated for the 10th ranked score using
mufix2 and the edge-effect-corrected lambda.  
\item [\monob{pmu} ] Location parameter as determined
by H3's estimation procedures.  
\item [\monob{plambda} ] Slope parameter as determined by
H3's estimation procedures.  
\item [\monob{pE@10} ] The E-value calculated for the 10th ranked
score using pmu, plambda.   
\end{wideitem}


At the end of this table, one more line is printed,
starting with \# and summarizing the overall CPU time used by the simulations.
 

Some of the optional output files are in xmgrace xy format. xmgrace is
powerful and freely available graph-plotting software.   
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help;
print a brief reminder of command line usage and all available options.
 
\item [\monob{-a} ] Collect expected Viterbi alignment length statistics from each simulated
sequence. This only works with Viterbi scores (the default; see \mono{{-}{-}vit}).\mono{} Two
additional fields are printed in the output table for each model: the mean
length of Viterbi alignments, and the standard deviation.  
\item [\monob{-v} ] (Verbose). Print
the scores too, one score per line.   
\item [\monob{-L}\monoi{ $<$n$>$} ] Set the length of the randomly
sampled (nonhomologous) sequences to  \monoi{$<$n$>$}. The default is 100.   
\item [\monob{-N}\monoi{ $<$n$>$} ] Set the
number of randomly sampled sequences to  \monoi{$<$n$>$}. The default is 1000.  
\item [\monob{{-}{-}mpi} ] Run
under MPI control with master/worker parallelization (using \mono{mpirun}, for
example, or equivalent). Only available if optional MPI support was enabled
at compile-time.  It is parallelized at the level of sending one profile
at a time to an MPI worker process, so parallelization only helps if you
have more than one profile in the  \monoi{hmmfile}, and you want to have at least
as many profiles as MPI worker processes.     
\end{wideitem}

\subsection*{Options Controlling Output}

\begin{wideitem}
\item [\monob{-o}\monoi{ $<$f$>$} ] Save the main output table to a file \monoi{$<$f$>$} rather than sending it to stdout.
 
\item [\monob{{-}{-}afile}\monoi{ $<$f$>$} ] When collecting Viterbi alignment statistics (the \mono{-a } option),
for each sampled sequence, output two fields per line to a file \monoi{$<$f$>$}: the
length of the optimal alignment, and the Viterbi bit score. Requires that
the  \mono{-a} option is also used.   
\item [\monob{{-}{-}efile}\monoi{ $<$f$>$} ] Output a rank vs. E-value plot in XMGRACE
xy format to file \monoi{$<$f$>$}. The x-axis is the rank of this sequence, from highest
score to lowest; the y-axis is the E-value calculated for this sequence. E-values
are calculated using H3's default procedures (i.e. the pmu, plambda parameters
in the output table). You expect a rough match between rank and E-value if
E-values are accurately estimated.   
\item [\monob{{-}{-}ffile}\monoi{ $<$f$>$} ] Output a "filter power" file
to  \monoi{$<$f$>$}: for each model, a line with three fields: model name, number of
sequences passing the P-value threshold, and fraction of sequences passing
the P-value threshold. See \mono{{-}{-}pthresh} for setting the P-value threshold, which
defaults to 0.02 (the default MSV filter threshold in H3). The P-values are
as determined by H3's default procedures (the pmu,plambda parameters in
the output table). If all is well, you expect to see filter power equal
to the predicted P-value setting of the threshold.  
\item [\monob{{-}{-}pfile}\monoi{ $<$f$>$} ] Output cumulative
survival plots (P(S$>$x)) to file \monoi{$<$f$>$} in XMGRACE xy format. There are three plots:
(1) the observed score distribution;  (2) the maximum likelihood fitted
distribution; (3) a maximum likelihood fit to the location parameter (mu/tau)
while     assuming lambda=log\_2.

  

 
\item [\monob{{-}{-}xfile}\monoi{ $<$f$>$} ] Output the bit scores as a binary array of double-precision floats
(8 bytes per score) to file \monoi{$<$f$>$}. Programs like Easel's  \mono{esl-histplot} can read
such binary files. This is useful when generating extremely large sample
sizes.   
\end{wideitem}

\subsection*{Options Controlling Model Configuration (mode)}
 H3 only uses multihit
local alignment ( \mono{{-}{-}fs } mode), and this is where we believe the statistical
fits.  Unihit local alignment scores (Smith/Waterman;  \mono{{-}{-}sw} mode) also obey
our statistical conjectures. Glocal alignment statistics (either multihit
or unihit) are still not adequately understood nor adequately fitted.  \begin{wideitem}
\item [\monob{{-}{-}fs}
] Collect multihit local alignment scores. This is the default. "fs" comes
from HMMER2's historical terminology for multihit local alignment as 'fragment
search mode'.  
\item [\monob{{-}{-}sw} ] Collect unihit local alignment scores. The H3 J state is
disabled. "sw" comes from HMMER2's historical terminology for unihit local
alignment as 'Smith/Waterman search mode'.  
\item [\monob{{-}{-}ls} ] Collect multihit glocal alignment
scores. In glocal (global/local) alignment, the entire model must align,
to a subsequence of the target. The H3 local entry/exit transition probabilities
are disabled. 'ls' comes from HMMER2's historical terminology for multihit
local alignment as 'local search mode'.  
\item [\monob{{-}{-}s} ] Collect unihit glocal alignment
scores.  Both the H3 J state and local entry/exit transition probabilities
are disabled. 's' comes from HMMER2's historical terminology for unihit glocal
alignment.    
\end{wideitem}

\subsection*{Options Controlling Scoring Algorithm}
 \begin{wideitem}
\item [\monob{{-}{-}vit} ] Collect Viterbi
maximum likelihood alignment scores. This is the default.  
\item [\monob{{-}{-}fwd} ] Collect Forward
log-odds likelihood scores, summed over alignment ensemble.  
\item [\monob{{-}{-}hyb} ] Collect
'Hybrid' scores, as described in papers by Yu and Hwa (for instance, Bioinformatics
18:864, 2002). These involve calculating a Forward matrix and taking the
maximum cell value. The number itself is statistically somewhat unmotivated,
but the distribution is expected be a well-behaved extreme value distribution
(Gumbel).  
\item [\monob{{-}{-}msv} ] Collect MSV (multiple ungapped segment Viterbi) scores, using
H3's main acceleration heuristic.  
\item [\monob{{-}{-}fast} ] For any of the above options, use
H3's optimized production implementation (using SIMD vectorization). The
default is to use the "generic" implementation (slow and non-vectorized).
The optimized implementations sacrifice a small amount of numerical precision.
This can introduce confounding noise into statistical simulations and fits,
so when one gets super-concerned about exact details, it's better to be able
to factor that source of noise out.  
\end{wideitem}

\subsection*{Options Controlling Fitted Tail Masses
for Forward}
 In some experiments, it was useful to fit Forward scores to
a range of different tail masses, rather than just one. These options provide
a mechanism for fitting an evenly-spaced range of different tail masses.
For each different tail mass, a line is generated in the output.  \begin{wideitem}
\item [\monob{{-}{-}tmin}\monoi{ $<$x$>$}
] Set the lower bound on the tail mass distribution. (The default is 0.02 for
the default single tail mass.)  
\item [\monob{{-}{-}tmax}\monoi{ $<$x$>$} ] Set the upper bound on the tail mass
distribution. (The default is 0.02 for the default single tail mass.)  
\item [\monob{{-}{-}tpoints}\monoi{
$<$n$>$} ] Set the number of tail masses to sample, starting from \mono{{-}{-}tmin} and ending
at  \mono{{-}{-}tmax}. (The default is 1, for the default 0.02 single tail mass.)  
\item [\monob{{-}{-}tlinear}
] Sample a range of tail masses with uniform linear spacing. The default is
to use uniform logarithmic spacing.    
\end{wideitem}

\subsection*{Options Controlling H3 Parameter
Estimation Methods}
 H3 uses three short random sequence simulations to estimating
the location parameters for the expected score distributions for MSV scores,
Viterbi scores, and Forward scores. These options allow these simulations
to be modified.  \begin{wideitem}
\item [\monob{{-}{-}EmL}\monoi{ $<$n$>$} ] Sets the sequence length in simulation that estimates
the location parameter mu for MSV E-values. Default is 200.  
\item [\monob{{-}{-}EmN}\monoi{ $<$n$>$} ] Sets the
number of sequences in simulation that estimates the location parameter
mu for MSV E-values. Default is 200.  
\item [\monob{{-}{-}EvL}\monoi{ $<$n$>$} ] Sets the sequence length in simulation
that estimates the location parameter mu for Viterbi E-values. Default is
200.  
\item [\monob{{-}{-}EvN}\monoi{ $<$n$>$} ] Sets the number of sequences in simulation that estimates the
location parameter mu for Viterbi E-values. Default is 200.  
\item [\monob{{-}{-}EfL}\monoi{ $<$n$>$} ] Sets the
sequence length in simulation that estimates the location parameter tau
for Forward E-values. Default is 100.  
\item [\monob{{-}{-}EfN}\monoi{ $<$n$>$} ] Sets the number of sequences
in simulation that estimates the location parameter tau for Forward E-values.
Default is 200.  
\item [\monob{{-}{-}Eft}\monoi{ $<$x$>$} ] Sets the tail mass fraction to fit in the simulation
that estimates the location parameter tau for Forward evalues. Default is
0.04.   
\end{wideitem}

\subsection*{Debugging Options}
 \begin{wideitem}
\item [\monob{{-}{-}stall} ] For debugging the MPI master/worker version:
pause after start, to enable the developer to attach debuggers to the running
master and worker(s) processes. Send SIGCONT signal to release the pause.
(Under gdb:  \monoi{(gdb) signal SIGCONT}) (Only available if optional MPI support
was enabled at compile-time.)  
\item [\monob{{-}{-}seed}\monoi{ $<$n$>$} ] Set the random number seed to  \monoi{$<$n$>$}. The
default is 0, which makes the random number generator use an arbitrary
seed, so that different runs of  \mono{hmmsim } will almost certainly generate
a different statistical sample. For debugging, it is useful to force reproducible
results, by fixing a random number seed.    
\end{wideitem}

\subsection*{Experimental Options}
 These options
were used in a small variety of different exploratory experiments.  \begin{wideitem}
\item [\monob{{-}{-}bgflat
} ] Set the background residue distribution to a uniform distribution, both
for purposes of the null model used in calculating scores, and for generating
the random sequences. The default is to use a standard amino acid background
frequency distribution.  
\item [\monob{{-}{-}bgcomp} ] Set the background residue distribution
to the mean composition of the profile. This was used in exploring some
of the effects of biased composition.  
\item [\monob{{-}{-}x-no-lengthmodel} ] Turn the H3 target
sequence length model off. Set the self-transitions for N,C,J and the null
model to 350/351 instead; this emulates HMMER2. Not a good idea in general.
This was used to demonstrate one of the main H2 vs. H3 differences.  
\item [\monob{{-}{-}nu}\monoi{ $<$x$>$}
] Set the nu parameter for the MSV algorithm {-}{-} the expected number of ungapped
local alignments per target sequence. The default is 2.0, corresponding to
a E-$>$J transition probability of 0.5. This was used to test whether varying
nu has significant effect on result (it doesn't seem to, within reason).
This option only works if \mono{{-}{-}msv} is selected (it only affects MSV), and it
will not work with  \mono{{-}{-}fast} (because the optimized implementations are hardwired
to assume nu=2.0).  
\item [\monob{{-}{-}pthresh}\monoi{ $<$x$>$} ] Set the filter P-value threshold to use in generating
filter power files with \mono{{-}{-}ffile}. The default is 0.02 (which would be appropriate
for testing MSV scores, since this is the default MSV filter threshold
in H3's acceleration pipeline.) Other appropriate choices (matching defaults
in the acceleration pipeline) would be 0.001 for Viterbi, and 1e-5 for Forward.
     
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{hmmstat}}{hmmstat} - summary statistics for a profile file   }
\subsection*{Synopsis}
\noindent
\monob{hmmstat} [\monoi{options}]
\monoi{hmmfile}   
\subsection*{Description}
 The \mono{hmmstat} utility prints out a tabular file of
summary statistics for each profile in \monoi{hmmfile}.   

\monoi{hmmfile}  may be '-' (a dash
character), in which case profiles are read from a stdin pipe instead of
from a file.  

The columns are:  \begin{wideitem}
\item [\monob{idx} ] The index of this profile, numbering
each profile in the file starting from 1.  
\item [\monob{name} ] The name of the profile.
 
\item [\monob{accession} ] The optional accession of the profile, or "-" if there is none.
 
\item [\monob{nseq} ] The number of sequences that the profile was estimated from.  
\item [\monob{eff\_nseq}
] The effective number of sequences that the profile was estimated from,
after HMMER applied an effective sequence number calculation such as the
default entropy weighting.  
\item [\monob{M} ] The length of the model in consensus residues
(match states).  
\item [\monob{relent} ] Mean relative entropy per match state, in bits. This
is the expected (mean) score per consensus position. This is what the default
entropy-weighting method for effective sequence number estimation focuses
on, so for default HMMER3 models, you expect this value to reflect the
default target for entropy-weighting.  
\item [\monob{info} ] Mean information content per
match state, in bits. Probably not useful. Information content is a slightly
different calculation than relative entropy.   
\item [\monob{p relE} ] Mean positional relative
entropy, in bits. This is a fancier version of the per-match-state relative
entropy, taking into account the transition (insertion/deletion) probabilities;
it may be a more accurate estimation of the average score contributed per
model consensus position.  
\item [\monob{compKL} ] Kullback-Leibler divergence from the default
background frequency distribution to the average composition of the profile's
consensus match states, in bits. The higher this number, the more biased
the residue composition of the profile is. Highly biased profiles can slow
the HMMER3 acceleration pipeline, by causing too many nonhomologous sequences
to pass the filters.   
\end{wideitem}

\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help; print a brief reminder of command
line usage and all available options.   
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{jackhmmer}}{jackhmmer} - iteratively search sequence(s) against a sequence database}
 
\subsection*{Synopsis}
\noindent
\monob{jackhmmer} [\monoi{options}] \monoi{seqfile} \monoi{seqdb}  
\subsection*{Description}
 

\mono{jackhmmer} iteratively
searches each query sequence in  \monoi{seqfile} against the target sequence(s)
in \monoi{seqdb}. The first iteration is identical to a  \mono{phmmer} search. For the next
iteration, a multiple alignment of the query together with all target sequences
satisfying  inclusion thresholds is assembled, a profile is constructed
from this alignment (identical to using \mono{hmmbuild} on the alignment), and
profile search of the \monoi{seqdb} is done (identical to an \mono{hmmsearch} with the
profile).   

The query \monoi{seqfile}  may be '-' (a dash character), in which case
the query sequences are read from a stdin pipe instead of from a file. 
The \monoi{seqdb}  needs to be a 'normal' sequence file. It cannot be read from a
stdin stream, because \mono{jackhmmer} needs to do multiple passes over the database.
It cannot be a compressed (gzipped) file either, because we treat gzipped
files essentially as stdin streams, calling an external decompression program.
   

The output format is designed to be human-readable, but is often so voluminous
that reading it is impractical, and parsing it is a pain. The \mono{{-}{-}tblout } and
 \mono{{-}{-}domtblout } options save output in simple tabular formats that are concise
and easier to parse. The  \mono{-o} option allows redirecting the main output, including
throwing it away in /dev/null.   
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help; print a brief reminder
of command line usage and all available options.  
\item [\monob{-N}\monoi{ $<$n$>$} ] Set the maximum number
of iterations to  \monoi{$<$n$>$}. The default is 5. If N=1, the result is equivalent to
a \mono{phmmer} search.     
\end{wideitem}

\subsection*{Options Controlling Output}
 By default, output for each
iteration appears on stdout in a somewhat human readable, somewhat parseable
format. These options allow redirecting that output or saving additional
kinds of output to files, including checkpoint files for each iteration.
 \begin{wideitem}
\item [\monob{-o}\monoi{ $<$f$>$} ] Direct the human-readable output to a file \monoi{$<$f$>$}.  
\item [\monob{-A}\monoi{ $<$f$>$} ] After the final iteration,
save an annotated multiple alignment of all hits satisfying inclusion thresholds
(also including the original query) to \monoi{$<$f$>$} in Stockholm format.  
\item [\monob{{-}{-}tblout}\monoi{ $<$f$>$}
] After the final iteration, save a tabular summary of top sequence hits
to  \monoi{$<$f$>$} in a readily parseable, columnar, whitespace-delimited format.  
\item [\monob{{-}{-}domtblout}\monoi{
$<$f$>$} ] After the final iteration, save a tabular summary of top domain hits
to  \monoi{$<$f$>$} in a readily parseable, columnar, whitespace-delimited format.  
\item [\monob{{-}{-}chkhmm}\monoi{
prefix} ] At the start of each iteration, checkpoint the query HMM, saving
it to a file named \monoi{prefix}\mono{-}\monoi{n}\mono{.hmm} where \monoi{n} is the iteration number (from 1..N).
 
\item [\monob{{-}{-}chkali}\monoi{ prefix} ] At the end of each iteration, checkpoint an alignment of
all domains satisfying inclusion thresholds (e.g. what will become the query
HMM for the next iteration),  saving it to a file named \monoi{prefix}\mono{-}\monoi{n}\mono{.sto} in Stockholm
format, where \monoi{n} is the iteration number (from 1..N).  
\item [\monob{{-}{-}acc} ] Use accessions instead
of names in the main output, where available for profiles and/or sequences.
 
\item [\monob{{-}{-}noali} ] Omit the alignment section from the main output. This can greatly
reduce the output volume.  
\item [\monob{{-}{-}notextw} ] Unlimit the length of each line in the
main output. The default is a limit of 120 characters per line, which helps
in displaying the output cleanly on terminals and in editors, but can truncate
target profile description lines.  
\item [\monob{{-}{-}textw}\monoi{ $<$n$>$} ] Set the main output's line length
limit to \monoi{$<$n$>$} characters per line. The default is 120.       
\end{wideitem}

\subsection*{Options Controlling
Single Sequence Scoring (first Iteration)}
 By default, the first iteration
uses a search model constructed from a single query sequence. This model
is constructed using a standard 20x20 substitution matrix for residue probabilities,
and two additional parameters for position-independent gap open and gap
extend probabilities. These options allow the default single-sequence scoring
parameters to be changed.  \begin{wideitem}
\item [\monob{{-}{-}popen}\monoi{ $<$x$>$} ] Set the gap open probability for a single
sequence query model to  \monoi{$<$x$>$}. The default is 0.02.  \monoi{$<$x$>$}  must be $>$= 0 and $<$ 0.5. 

\item [\monob{{-}{-}pextend}\monoi{ $<$x$>$} ] Set the gap extend probability for a single sequence query model
to  \monoi{$<$x$>$}. The default is 0.4.  \monoi{$<$x$>$}  must be $>$= 0 and $<$ 1.0.  
\item [\monob{{-}{-}mx}\monoi{ $<$s$>$} ] Obtain residue alignment
probabilities from the built-in substitution matrix named \monoi{$<$s$>$}.\monoi{} Several standard
matrices are built-in, and do not need to be read from files.  The matrix
name \monoi{$<$s$>$}  can be PAM30, PAM70, PAM120, PAM240, BLOSUM45, BLOSUM50, BLOSUM62,
BLOSUM80, or BLOSUM90. Only one of the \mono{{-}{-}mx } and \mono{{-}{-}mxfile} options may be used.
 
\item [\monob{{-}{-}mxfile}\monoi{ mxfile} ] Obtain residue alignment probabilities from the substitution
matrix in file \monoi{mxfile}. The default score matrix is BLOSUM62 (this matrix
is internal to HMMER and does not have to be available as a file).  The
format of a substitution matrix \monoi{mxfile} is the standard format accepted
by BLAST, FASTA, and other sequence  analysis software. See \mono{ftp.ncbi.nlm.nih.gov/blast/matrices/}
for example files. (The only exception: we require matrices to be square,
so for DNA, use files like NCBI's NUC.4.4, not NUC.4.2.)   
\end{wideitem}

\subsection*{Options Controlling
Reporting Thresholds}
 Reporting thresholds control which hits are reported
in output files (the main output, \mono{{-}{-}tblout}, and  \mono{{-}{-}domtblout}). In each iteration,
sequence hits and domain hits are ranked by statistical significance (E-value)
and output is generated in two sections called per-target and per-domain
output. In per-target output, by default, all sequence hits with an E-value
$<$= 10 are reported. In the per-domain output, for each target that has passed
per-target reporting thresholds, all domains satisfying per-domain reporting
thresholds are reported. By default, these are domains with conditional
E-values of $<$= 10. The following options allow you to change the default E-value
reporting thresholds, or to use bit score thresholds instead.   \begin{wideitem}
\item [\monob{-E}\monoi{ $<$x$>$} ] Report
sequences with E-values $<$= \monoi{$<$x$>$} in per-sequence output. The default is 10.0.  
\item [\monob{-T}\monoi{
$<$x$>$} ] Use a bit score threshold for per-sequence output instead of an E-value
threshold (any setting of \mono{-E} is ignored). Report sequences with a bit score
of $>$= \monoi{$<$x$>$}. By default this option is unset.  
\item [\monob{-Z}\monoi{ $<$x$>$} ] Declare the total size of the
database to be \monoi{$<$x$>$} sequences, for purposes of E-value calculation. Normally
E-values are calculated relative to the size of the database you actually
searched (e.g. the number of sequences in  \monoi{target\_seqdb}). In some cases (for
instance, if you've split your target sequence database into multiple files
for parallelization of your search), you may know better what the actual
size of your search space is.  
\item [\monob{{-}{-}domE}\monoi{ $<$x$>$} ] Report domains with conditional E-values
$<$= \monoi{$<$x$>$} in per-domain output, in addition to the top-scoring domain per significant
sequence hit. The default is 10.0.  
\item [\monob{{-}{-}domT}\monoi{ $<$x$>$} ] Use a bit score threshold for per-domain
output instead of an E-value threshold (any setting of \mono{{-}{-}domE} is ignored).
Report domains with a bit score of $>$= \monoi{$<$x$>$} in per-domain output, in addition
to the top-scoring domain per significant sequence hit. By default this option
is unset.  
\item [\monob{{-}{-}domZ}\monoi{ $<$x$>$} ] Declare the number of significant sequences to be \monoi{$<$x$>$} sequences,
for purposes of conditional E-value calculation for additional domain significance.
Normally conditional E-values are calculated relative to the number of sequences
passing per-sequence reporting threshold.   
\end{wideitem}

\subsection*{Options Controlling Inclusion
Thresholds}
 Inclusion thresholds control which hits are included in the
multiple alignment and profile constructed for the next search iteration.
By default,  a sequence must have a per-sequence E-value of $<$= 0.001 (see \mono{-E
} option) to be included, and any additional domains in it besides the top-scoring
one must have a conditional E-value of $<$= 0.001 (see  \mono{{-}{-}domE } option). The difference
between reporting thresholds and inclusion thresholds is that inclusion
thresholds control which hits actually get used in the next iteration (or
the final output multiple alignment if the  \mono{-A } option is used), whereas
reporting thresholds control what you see in output. Reporting thresholds
are generally more loose so you can see borderline hits in the top of the
noise that might be of interest.  \begin{wideitem}
\item [\monob{{-}{-}incE}\monoi{ $<$x$>$} ] Include sequences with E-values
$<$= \monoi{$<$x$>$} in subsequent iteration or final alignment output by  \mono{-A}. The default
is 0.001.  
\item [\monob{{-}{-}incT}\monoi{ $<$x$>$} ] Use a bit score threshold for per-sequence inclusion instead
of an E-value threshold (any setting of \mono{{-}{-}incE} is ignored). Include sequences
with a bit score of $>$= \monoi{$<$x$>$}. By default this option is unset.  
\item [\monob{{-}{-}incdomE}\monoi{ $<$x$>$} ] Include
domains with conditional E-values $<$= \monoi{$<$x$>$} in subsequent iteration or final alignment
output by \mono{-A}, in addition to the top-scoring domain per significant sequence
hit.  The default is 0.001.  
\item [\monob{{-}{-}incdomT}\monoi{ $<$x$>$} ] Use a bit score threshold for per-domain
inclusion instead of an E-value threshold (any setting of \mono{{-}{-}incdomE} is ignored).
Include domains with a bit score of $>$= \monoi{$<$x$>$}. By default this option is unset.
   
\end{wideitem}

\subsection*{Options Controlling Acceleration Heuristics}
 HMMER3 searches are accelerated
in a three-step filter pipeline: the MSV filter, the Viterbi filter, and
the Forward filter. The first filter is the fastest and most approximate;
the last is the full Forward scoring algorithm, slowest but most accurate.
There is also a bias filter step between MSV and Viterbi. Targets that pass
all the steps in the acceleration pipeline are then subjected to postprocessing
{-}{-} domain identification and scoring using the Forward/Backward algorithm.
 Essentially the only free parameters that control HMMER's heuristic filters
are the P-value thresholds controlling the expected fraction of nonhomologous
sequences that pass the filters. Setting the default thresholds higher will
pass a higher proportion of nonhomologous sequence, increasing sensitivity
at the expense of speed; conversely, setting lower P-value thresholds will
pass a smaller proportion, decreasing sensitivity and increasing speed.
Setting a filter's P-value threshold to 1.0 means it will passing all sequences,
and effectively disables the filter.  Changing filter thresholds only removes
or includes targets from consideration; changing filter thresholds does
not alter bit scores, E-values, or alignments, all of which are determined
solely in postprocessing.  \begin{wideitem}
\item [\monob{{-}{-}max} ] Maximum sensitivity.  Turn off all filters,
including the bias filter, and run full Forward/Backward postprocessing
on every target. This increases sensitivity slightly, at a large cost in
speed.  
\item [\monob{{-}{-}F1}\monoi{ $<$x$>$} ] First filter threshold; set the P-value threshold for the MSV
filter step.  The default is 0.02, meaning that roughly 2\% of the highest
scoring nonhomologous targets are expected to pass the filter.  
\item [\monob{{-}{-}F2}\monoi{ $<$x$>$} ] Second
filter threshold; set the P-value threshold for the Viterbi filter step.
 The default is 0.001.  
\item [\monob{{-}{-}F3}\monoi{ $<$x$>$} ] Third filter threshold; set the P-value threshold
for the Forward filter step.  The default is 1e-5.  
\item [\monob{{-}{-}nobias} ] Turn off the bias
filter. This increases sensitivity somewhat, but can come at a high cost
in speed, especially if the query has biased residue composition (such
as a repetitive sequence region, or if it is a membrane protein with large
regions of hydrophobicity). Without the bias filter, too many sequences
may pass the filter with biased queries, leading to slower than expected
performance as the computationally intensive Forward/Backward algorithms
shoulder an abnormally heavy load.    
\end{wideitem}

\subsection*{Options Controlling Profile Construction
(later Iterations)}
 \mono{jackhmmer} always includes your original query sequence
in the alignment result at every iteration, and consensus positions are
always defined by that query sequence. That is, a  \mono{jackhmmer} profile is
always the same length as your original query, at every iteration. Therefore
\mono{jackhmmer} gives you less control over profile construction than \mono{hmmbuild}
does; it does not have the \mono{{-}{-}fast}, or \mono{{-}{-}hand}, or \mono{{-}{-}symfrac} options. The only profile
construction option available in \mono{jackhmmer} is \mono{{-}{-}fragthresh}:   \begin{wideitem}
\item [\monob{{-}{-}fragthresh}\monoi{
$<$x$>$} ] We only want to count terminal gaps as deletions if the aligned sequence
is known to be full-length, not if it is a fragment (for instance, because
only part of it was sequenced). HMMER uses a simple rule to infer fragments:
if the sequence length L is less than  or equal to a fraction \monoi{$<$x$>$}  times
the alignment length in columns, then the sequence is handled as a fragment.
The default is 0.5. Setting \mono{{-}{-}fragthresh 0} will define no (nonempty) sequence
as a fragment; you might want to do this if you know you've got a carefully
curated alignment of full-length sequences. Setting \mono{{-}{-}fragthresh 1} will define
all sequences as fragments; you might want to do this if you know your
alignment is entirely composed of fragments, such as translated short reads
in metagenomic shotgun data.    
\end{wideitem}

\subsection*{Options Controlling Relative Weights}
 Whenever
a profile is built from a multiple alignment, HMMER uses an ad hoc sequence
weighting algorithm to downweight closely related sequences and upweight
distantly related ones. This has the effect of making models less biased
by uneven phylogenetic representation. For example, two identical sequences
would typically each receive half the weight that one sequence would (and
this is why  \mono{jackhmmer } isn't concerned about always including your original
query sequence in each iteration's alignment, even if it finds it again
in the database you're searching). These options control which algorithm
gets used.  \begin{wideitem}
\item [\monob{{-}{-}wpb} ] Use the Henikoff position-based sequence weighting scheme
[Henikoff and Henikoff, J. Mol. Biol. 243:574, 1994].  This is the default.
 
\item [\monob{{-}{-}wgsc } ] Use the Gerstein/Sonnhammer/Chothia weighting algorithm [Gerstein
et al, J. Mol. Biol. 235:1067, 1994].  
\item [\monob{{-}{-}wblosum} ] Use the same clustering scheme
that was used to weight data in calculating BLOSUM substitution matrices
[Henikoff and Henikoff, Proc. Natl. Acad. Sci 89:10915, 1992]. Sequences are
single-linkage clustered at an identity threshold (default 0.62; see \mono{{-}{-}wid})
and within each cluster of c sequences, each sequence gets relative weight
1/c.  
\item [\monob{{-}{-}wnone} ] No relative weights. All sequences are assigned uniform weight.
  
\item [\monob{{-}{-}wid}\monoi{ $<$x$>$} ] Sets the identity threshold used by single-linkage clustering when
 using  \mono{{-}{-}wblosum}.\mono{} Invalid with any other weighting scheme. Default is 0.62.
     
\end{wideitem}

\subsection*{Options Controlling Effective Sequence Number}
 After relative weights
are determined, they are normalized to sum to a total effective sequence
number,  \monoi{eff\_nseq}.\monoi{} This number may be the actual number of sequences in
the alignment, but it is almost always smaller than that. The default entropy
weighting method  (\mono{{-}{-}eent}) reduces the effective sequence number to reduce
the information content (relative entropy, or average expected score on
true homologs) per consensus position. The target relative entropy is controlled
by a two-parameter function, where the two parameters are settable with
\mono{{-}{-}ere} and  \mono{{-}{-}esigma}.  \begin{wideitem}
\item [\monob{{-}{-}eent} ] Adjust effective sequence number to achieve a specific
relative entropy per position (see \mono{{-}{-}ere}). This is the default.  
\item [\monob{{-}{-}eentexp} ] Adjust
the effective sequence number to reach the relative entropy target using
exponential scaling.  
\item [\monob{{-}{-}eclust} ] Set effective sequence number to the number
of single-linkage clusters at a specific identity threshold (see  \mono{{-}{-}eid}). This
option is not recommended; it's for experiments evaluating how much better
\mono{{-}{-}eent} is.  
\item [\monob{{-}{-}enone} ] Turn off effective sequence number determination and just
use the actual number of sequences. One reason you might want to do this
is to try to maximize the relative entropy/position of your model, which
may be useful for short models.  
\item [\monob{{-}{-}eset}\monoi{ $<$x$>$} ] Explicitly set the effective sequence
number for all models to  \monoi{$<$x$>$}.  
\item [\monob{{-}{-}ere}\monoi{ $<$x$>$} ] Set the minimum relative entropy/position
target to  \monoi{$<$x$>$}. Requires \mono{{-}{-}eent}.\mono{} Default depends on the sequence alphabet; for
protein sequences, it is 0.59 bits/position.  
\item [\monob{{-}{-}esigma}\monoi{ $<$x$>$} ] Sets the minimum relative
entropy contributed by an entire model alignment, over its whole length.
This has the effect of making short models have  higher relative entropy
per position than  \mono{{-}{-}ere } alone would give. The default is 45.0 bits.  
\item [\monob{{-}{-}eid}\monoi{ $<$x$>$}
] Sets the fractional pairwise identity cutoff used by  single linkage clustering
with the \mono{{-}{-}eclust } option. The default is 0.62.    
\end{wideitem}

\subsection*{Options Controlling Priors}

In profile construction, by default, weighted counts are converted to mean
posterior probability parameter estimates using mixture Dirichlet priors.
 Default mixture Dirichlet prior parameters for protein models and for
nucleic acid (RNA and DNA) models are built in. The following options allow
you to override the default priors.  \begin{wideitem}
\item [\monob{{-}{-}pnone} ] Don't use any priors. Probability
parameters will simply be the observed frequencies, after relative sequence
weighting.   
\item [\monob{{-}{-}plaplace} ] Use a Laplace +1 prior in place of the default mixture
Dirichlet prior.    
\end{wideitem}

\subsection*{Options Controlling E-value Calibration}
 Estimating the
location parameters for the expected score distributions for MSV filter
scores, Viterbi filter scores, and Forward scores requires three short
random sequence simulations.  \begin{wideitem}
\item [\monob{{-}{-}EmL}\monoi{ $<$n$>$} ] Sets the sequence length in simulation
that estimates the location parameter mu for MSV filter E-values. Default
is 200.  
\item [\monob{{-}{-}EmN}\monoi{ $<$n$>$} ] Sets the number of sequences in simulation that estimates
the location parameter mu for MSV filter E-values. Default is 200.  
\item [\monob{{-}{-}EvL}\monoi{ $<$n$>$}
] Sets the sequence length in simulation that estimates the location parameter
mu for Viterbi filter E-values. Default is 200.  
\item [\monob{{-}{-}EvN}\monoi{ $<$n$>$} ] Sets the number of
sequences in simulation that estimates the location parameter mu for Viterbi
filter E-values. Default is 200.  
\item [\monob{{-}{-}EfL}\monoi{ $<$n$>$} ] Sets the sequence length in simulation
that estimates the location parameter tau for Forward E-values. Default is
100.  
\item [\monob{{-}{-}EfN}\monoi{ $<$n$>$} ] Sets the number of sequences in simulation that estimates the
location parameter tau for Forward E-values. Default is 200.  
\item [\monob{{-}{-}Eft}\monoi{ $<$x$>$} ] Sets the
tail mass fraction to fit in the simulation that estimates the location
parameter tau for Forward evalues. Default is 0.04.   
\end{wideitem}

\subsection*{Other Options}
 \begin{wideitem}
\item [\monob{{-}{-}nonull2}
] Turn off the null2 score corrections for biased composition.  
\item [\monob{-Z}\monoi{ $<$x$>$} ] Assert
that the total number of targets in your searches is \monoi{$<$x$>$}, for the purposes
of per-sequence E-value calculations, rather than the actual number of targets
seen.   
\item [\monob{{-}{-}domZ}\monoi{ $<$x$>$} ] Assert that the total number of targets in your searches
is \monoi{$<$x$>$}, for the purposes of per-domain conditional E-value calculations, rather
than the number of targets that passed the reporting thresholds.  
\item [\monob{{-}{-}seed}\monoi{ $<$n$>$}
] Seed the random number generator with \monoi{$<$n$>$}, an integer $>$= 0.  If  \monoi{$<$n$>$}  is $>$0, any
stochastic simulations will be reproducible; the same command will give
the same results. If  \monoi{$<$n$>$} is 0, the random number generator is seeded arbitrarily,
and stochastic simulations will vary from run to run of the same command.
The default seed is 42.   
\item [\monob{{-}{-}qformat}\monoi{ $<$s$>$} ] Assert that input query \monoi{seqfile} is in
format \monoi{$<$s$>$}, bypassing format autodetection. Common choices for  \monoi{$<$s$>$}  include:
\mono{fasta}, \mono{embl}, \mono{genbank.} Alignment formats also work; common choices include:
\mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal}, \mono{phylip}. \mono{jackhmmer} always uses a
single sequence query to start its search, so when the input \monoi{seqfile} is
an alignment, \mono{jackhmmer} reads it one unaligned query sequence at a time,
not as an alignment. For more information, and for codes for some less common
formats, see main documentation. The string \monoi{$<$s$>$} is case-insensitive (\mono{fasta}
or \mono{FASTA} both work).  
\item [\monob{{-}{-}tformat}\monoi{ $<$s$>$} ] Assert that the input target sequence \monoi{seqdb}
is in format  \monoi{$<$s$>$}. See \mono{{-}{-}qformat} above for accepted choices for \monoi{$<$s$>$}.    
\item [\monob{{-}{-}cpu}\monoi{ $<$n$>$} ] Set
the number of parallel worker threads to  \monoi{$<$n$>$}. On multicore machines, the
default is 2. You can also control this number by setting an environment
variable,  \monoi{HMMER\_NCPU}. There is also a master thread, so the actual number
of threads that HMMER spawns is \monoi{$<$n$>$}+1.  This option is not available if HMMER
was compiled with POSIX threads support turned off.    
\item [\monob{{-}{-}stall} ] For debugging
the MPI master/worker version: pause after start, to enable the developer
to attach debuggers to the running master and worker(s) processes. Send
SIGCONT signal to release the pause. (Under gdb:  \mono{(gdb) signal SIGCONT})
(Only available if optional MPI support was enabled at compile-time.)  
\item [\monob{{-}{-}mpi}
] Run under MPI control with master/worker parallelization (using \mono{mpirun},
for example, or equivalent). Only available if optional MPI support was
enabled at compile-time.      
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{makehmmerdb}}{makehmmerdb} - build nhmmer database from a sequence file   }
\subsection*{Synopsis}
\noindent
\monob{makehmmerdb}
[\monoi{options}] \monoi{seqfile} \monoi{binaryfile}   
\subsection*{Description}
 

\mono{makehmmerdb } is used to create
a binary file from a DNA sequence file. This  binary file may be used as
a target database for the DNA search tool \mono{nhmmer}.\mono{} Using default settings
in  \mono{nhmmer}, this yields a roughly 10-fold acceleration with small loss of
 sensitivity on benchmarks.    
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help; print a brief reminder of
command line usage and all available options.    
\end{wideitem}

\subsection*{Other Options}
 \begin{wideitem}
\item [\monob{{-}{-}informat}\monoi{
$<$s$>$} ] Assert that input \monoi{seqfile} is in format \monoi{$<$s$>$}, bypassing format autodetection.
Common choices for  \monoi{$<$s$>$}  include: \mono{fasta}, \mono{embl}, \mono{genbank.} Alignment formats
also work; common choices include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal},
\mono{phylip}. For more information, and for codes for some less common formats,
see main documentation. The string \monoi{$<$s$>$} is case-insensitive (\mono{fasta} or \mono{FASTA}
both work).   
\item [\monob{{-}{-}bin\_length}\monoi{ $<$n$>$} ] Bin length. The binary file depends on a data
structure called the  FM index, which organizes a permuted copy of the
sequence in bins  of length \monoi{$<$n$>$}. Longer bin length will lead to smaller files
(because data is  captured about each bin) and possibly slower query time.
The  default is 256. Much more than 512 may lead to notable reduction  in
speed.   
\item [\monob{{-}{-}sa\_freq}\monoi{ $<$n$>$} ] Suffix array sample rate. The FM index structure also
samples from  the underlying suffix array for the sequence database. More
frequent  sampling (smaller value for  \monoi{$<$n$>$}) will yield larger file size and
faster search (until file size becomes large enough to cause I/O to be
a bottleneck). The default value is 8. Must be a power of 2.   
\item [\monob{{-}{-}block\_size}\monoi{
$<$n$>$} ] The input sequence is broken into blocks of size \monoi{$<$n$>$} million letters. An
FM index is built for each block, rather than  building an FM index for
the entire sequence database. Default is  50. Larger blocks do not seem to
yield substantial speed increase.     
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{nhmmer}}{nhmmer} - search DNA queries against a DNA sequence database   }
\subsection*{Synopsis}
\noindent
\monob{nhmmer}
[\monoi{options}] \monoi{queryfile} \monoi{seqdb}   
\subsection*{Description}
 

\mono{nhmmer } is used to search one or
more nucleotide queries against a  nucleotide sequence database. For each
query in  \monoi{queryfile}, use that query to search the target database of sequences
in \monoi{seqdb}, and output a ranked list of the hits with the most significant
matches to the query. A query may be either a profile model  built using
 \mono{hmmbuild},\mono{} a sequence alignment, or a single sequence. Sequence based queries
can be in a number of formats (see \mono{{-}{-}qformat}),\mono{} and can typically be autodetected.
Note that only  Stockholm format supports queries made up of more than
one sequence  alignment.    

Either the query \monoi{queryfile}  or the target \monoi{seqdb}
 may be '-' (a dash character), in which case the query file or target database
input will be read from a $<$stdin$>$ pipe instead of from a file. Only one input
source can come through $<$stdin$>$, not both. If the \monoi{queryfile}  contains more
than one query, then \monoi{seqdb}  cannot come from stdin, because we can't rewind
the streaming target database to search it with another profile.   

If the
query is sequence-based (unaligned or aligned), a new file containing the
HMM(s) built from the input(s) in  \monoi{queryfile} may optionally be produced,
with the filename set using the  \mono{{-}{-}hmmout} flag.   

The output format is designed
to be human-readable, but is often so voluminous that reading it is impractical,
and parsing it is a pain. The \mono{{-}{-}tblout } option saves output in a simple tabular
format that is concise and easier to parse. The  \mono{-o} option allows redirecting
the main output, including throwing it away in /dev/null.    
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help;
print a brief reminder of command line usage and all available options.
   
\end{wideitem}

\subsection*{Options for Controlling Output}
 \begin{wideitem}
\item [\monob{-o}\monoi{ $<$f$>$} ] Direct the main human-readable output
to a file \monoi{$<$f$>$}  instead of the default stdout.  
\item [\monob{-A}\monoi{ $<$f$>$} ] Save a multiple alignment
of all significant hits (those satisfying "inclusion thresholds") to the
file  \monoi{$<$f$>$}.  
\item [\monob{{-}{-}tblout}\monoi{ $<$f$>$} ] Save a simple tabular (space-delimited) file summarizing
the per-target output, with one data line per homologous target sequence
found.  
\item [\monob{{-}{-}dfamtblout}\monoi{ $<$f$>$} ] Save a tabular (space-delimited) file summarizing the
 per-hit output, similar to  \mono{{-}{-}tblout} but more succinct.   
\item [\monob{{-}{-}aliscoresout}\monoi{ $<$f$>$}\mono{} ] Save
to file a list of per-position scores for each hit. This is useful, for example,
in identifying regions of high score density for use in resolving overlapping
hits from  different models.  
\item [\monob{{-}{-}hmmout}\monoi{ $<$f$>$}\mono{} ] If \monoi{queryfile} is sequence-based, write
the internally-computed HMM(s) to file \monoi{$<$f$>$}.\monoi{}   
\item [\monob{{-}{-}acc} ] Use accessions instead of
names in the main output, where available for profiles and/or sequences.
 
\item [\monob{{-}{-}noali} ] Omit the alignment section from the main output. This can greatly
reduce the output volume.  
\item [\monob{{-}{-}notextw} ] Unlimit the length of each line in the
main output. The default is a limit of 120 characters per line, which helps
in displaying the output cleanly on terminals and in editors, but can truncate
target profile description lines.  
\item [\monob{{-}{-}textw}\monoi{ $<$n$>$} ] Set the main output's line length
limit to \monoi{$<$n$>$} characters per line. The default is 120.    
\end{wideitem}

\subsection*{Options Controlling
Single Sequence Scoring}
 By default, if a query is a single sequence from
a file in  fasta format, \mono{nhmmer } uses a search model constructed from that
sequence and a standard 20x20 substitution matrix for residue probabilities,
along with two additional parameters for position-independent gap open and
gap extend probabilities. These options allow the default single-sequence
scoring parameters to be changed, and for single-sequence scoring options
to be applied to a single sequence coming from an aligned format.  \begin{wideitem}
\item [\monob{{-}{-}singlemx}\monoi{}
] If a single sequence query comes from a multiple sequence alignment file,
 such as in Stockholm format, the search model is by default constructed
as is typically done  for multiple sequence alignments. This option forces
 \mono{nhmmer } to use the single-sequence method with substitution score matrix.
 
\item [\monob{{-}{-}mxfile}\monoi{$<$mxfile} ] Obtain residue alignment probabilities from the substitution
matrix in file \monoi{mxfile}. The default score matrix is DNA1 (this matrix is
internal to HMMER and does not have to be available as a file).  The format
of a substitution matrix \monoi{mxfile} is the standard format accepted by BLAST,
FASTA, and other sequence  analysis software. See \mono{ftp.ncbi.nlm.nih.gov/blast/matrices/}
for example files. (The only exception: we require matrices to be square,
so for DNA, use files like NCBI's NUC.4.4, not NUC.4.2.)   
\item [\monob{{-}{-}popen}\monoi{ $<$x$>$} ] Set the gap
open probability for a single sequence query model to  \monoi{$<$x$>$}. The default is
0.02.  \monoi{$<$x$>$}  must be $>$= 0 and $<$ 0.5.  
\item [\monob{{-}{-}pextend}\monoi{ $<$x$>$} ] Set the gap extend probability for
a single sequence query model to  \monoi{$<$x$>$}. The default is 0.4.  \monoi{$<$x$>$}  must be $>$= 0 and
$<$ 1.0.    
\end{wideitem}

\subsection*{Options Controlling Reporting Thresholds}
 Reporting thresholds control
which hits are reported in output files (the main output, \mono{{-}{-}tblout}, and 
\mono{{-}{-}dfamtblout}). Hits are ranked by statistical significance (E-value).    \begin{wideitem}
\item [\monob{-E}\monoi{ $<$x$>$}
] Report target sequences with an E-value of $<$= \monoi{$<$x$>$}.\monoi{} The default is 10.0, meaning
that on average, about 10 false positives will be reported per query, so
you can see the top of the noise and decide for yourself if it's really
noise.  
\item [\monob{-T}\monoi{ $<$x$>$} ] Instead of thresholding output on E-value, instead report target
sequences with a bit score of $>$= \monoi{$<$x$>$}.     
\end{wideitem}

\subsection*{Options for Inclusion Thresholds}

Inclusion thresholds are stricter than reporting thresholds. Inclusion thresholds
control which hits are considered to be reliable enough to be included
in an output alignment or a subsequent search round, or marked as significant
("!") as opposed to questionable ("?") in hit output.  \begin{wideitem}
\item [\monob{{-}{-}incE}\monoi{ $<$x$>$} ] Use an E-value
of $<$= \monoi{$<$x$>$} as the inclusion threshold. The default is 0.01, meaning that on average,
about 1 false positive would be expected in every 100 searches with different
query sequences.  
\item [\monob{{-}{-}incT}\monoi{ $<$x$>$} ] Instead of using E-values for setting the inclusion
threshold,  use a bit score of $>$=  \monoi{$<$x$>$} as the inclusion threshold. By default
this option is unset.    
\end{wideitem}

\subsection*{Options for Model-specific Score Thresholding}
 Curated
profile databases may define specific bit score thresholds for each profile,
superseding any thresholding based on statistical significance alone.  To
use these options, the profile must contain the appropriate (GA, TC, and/or
NC) optional score threshold annotation; this is picked up by  \mono{hmmbuild}
from Stockholm format alignment files. For a nucleotide model, each  thresholding
option has a single per-hit threshold $<$x$>$ This acts as if \mono{-T}\monoi{ $<$x$>$} \mono{{-}{-}incT}\monoi{ $<$x$>$} has been
applied specifically using each model's curated thresholds.  \begin{wideitem}
\item [\monob{{-}{-}cut\_ga} ] Use the
GA (gathering) bit score threshold in the model to set per-hit reporting
and inclusion thresholds. GA thresholds are generally considered to be the
reliable curated thresholds defining family membership; for example, in
Dfam, these thresholds are applied when annotating a genome with a model
of a family known to be found in that organism. They may allow for minimal
expected false discovery rate.  
\item [\monob{{-}{-}cut\_nc} ] Use the NC (noise cutoff) bit score
threshold in the model to set per-hit reporting and inclusion thresholds.
NC thresholds are less stringent than GA; in the context of Pfam, they
are generally used to store the score of the  highest-scoring known false
positive.  
\item [\monob{{-}{-}cut\_tc} ] Use the TC (trusted cutoff) bit score threshold in the
model to set per-hit reporting and inclusion thresholds. TC thresholds are
more stringent than GA, and are generally considered to be the score of
the lowest-scoring known  true positive that is above all known false positives;
for example, in Dfam, these thresholds are applied when annotating a genome
with a model of a family not known to be found in that organism.     
\end{wideitem}

\subsection*{Options
Controlling the Acceleration Pipeline}
 HMMER3 searches are accelerated in
a three-step filter pipeline: the scanning-SSV filter, the Viterbi filter,
and the Forward filter. The  first filter is the fastest and most approximate;
the last is the full Forward scoring algorithm. There is also a bias filter
step between SSV and Viterbi. Targets that pass all the steps in the acceleration
pipeline are then subjected to postprocessing {-}{-} domain identification and
scoring using the Forward/Backward algorithm.  Changing filter thresholds
only removes or includes targets from consideration; changing filter thresholds
does not alter bit scores, E-values, or alignments, all of which are determined
solely in postprocessing.  \begin{wideitem}
\item [\monob{{-}{-}max} ] Turn off (nearly) all filters, including
the bias filter, and run full Forward/Backward postprocessing on most of
the target sequence.  In contrast to  \mono{phmmer} and \mono{hmmsearch}, where this flag
really does turn off the filters entirely, the \mono{{-}{-}max} flag in  \mono{nhmmer} sets
the scanning-SSV filter threshold to 0.4, not 1.0. Use of this flag increases
sensitivity somewhat, at a large cost in speed.  
\item [\monob{{-}{-}F1}\monoi{ $<$x$>$} ] Set the P-value threshold
for the SSV filter step.  The default is 0.02, meaning that roughly 2\% of
the highest scoring nonhomologous targets are expected to pass the filter.
 
\item [\monob{{-}{-}F2}\monoi{ $<$x$>$} ] Set the P-value threshold for the Viterbi filter step. The default
is 0.001.   
\item [\monob{{-}{-}F3}\monoi{ $<$x$>$} ] Set the P-value threshold for the Forward filter step. The
default is 1e-5.  
\item [\monob{{-}{-}nobias} ] Turn off the bias filter. This increases sensitivity
somewhat, but can come at a high cost in speed, especially if the query
has biased residue composition (such as a repetitive sequence region, or
if it is a membrane protein with large regions of hydrophobicity). Without
the bias filter, too many sequences may pass the filter with biased queries,
leading to slower than expected performance as the computationally intensive
Forward/Backward algorithms shoulder an abnormally heavy load.    
\end{wideitem}

\subsection*{Options
for Specifying the Alphabet}
 \begin{wideitem}
\item [\monob{{-}{-}dna} ] Assert that sequences in \monoi{msafile} are DNA,
bypassing alphabet autodetection.  
\item [\monob{{-}{-}rna} ] Assert that sequences in  \monoi{msafile}
are RNA, bypassing alphabet autodetection.    
\end{wideitem}

\subsection*{Options Controlling Seed Search
Heuristic}
 When searching with  \mono{nhmmer}, one may optionally precompute a
binary version of the target database, using \mono{makehmmerdb}, then search against
that database. Using default settings, this yields a roughly 10-fold acceleration
with small loss of sensitivity on benchmarks. This is achieved using a heuristic
method that searches for seeds (ungapped  alignments) around which full
processing is done. This is essentially   a replacement to the SSV stage.
(This method has been extensively tested,  but should still be treated
as somewhat experimental.) The following options only impact  \mono{nhmmer} if
the value of  \mono{{-}{-}tformat} is \mono{hmmerdb}.  Changing parameters for this seed-finding
step will impact both speed and  sensitivity - typically faster search leads
to lower sensitivity.   \begin{wideitem}
\item [\monob{{-}{-}seed\_max\_depth}\monoi{ $<$n$>$} ] The seed step requires that a seed
reach a specified bit score in length  no longer than  \monoi{$<$n$>$}.\monoi{} By default, this
value is 15. Longer seeds allow a greater chance of  meeting the bit score
threshold, leading to diminished filtering (greater sensitivity, slower
run time).  
\item [\monob{{-}{-}seed\_sc\_thresh}\monoi{ $<$x$>$} ] The seed must reach score  \monoi{$<$x$>$} (in bits). The
default is 15.0 bits. A higher threshold increases  filtering stringency,
leading to faster run times and lower  sensitivity.  
\item [\monob{{-}{-}seed\_sc\_density}\monoi{ $<$x$>$} ] Either
all prefixes or all suffixes of a seed must have  bit density (bits per
aligned position) of at least  \monoi{$<$x$>$}.\monoi{} The default is 0.8 bits/position. An increase
in the density  requirement leads to increased filtering stringency, thus
faster  run times and lower sensitivity.  
\item [\monob{{-}{-}seed\_drop\_max\_len}\monoi{ $<$n$>$} ] A seed may
not have a run of length \monoi{$<$n$>$} in which the score drops by  \mono{{-}{-}seed\_drop\_lim} or
more. Basically, this prunes seeds that go through long slightly-negative
seed extensions. The default is 4.  Increasing  the limit causes (slightly)
diminished filtering efficiency, thus  slower run times and higher sensitivity.
(minor tuning option)  
\item [\monob{{-}{-}seed\_drop\_lim}\monoi{ $<$x$>$} ] In a seed, there may be no run of
length  \mono{{-}{-}seed\_drop\_max\_len} in which the score drops by  \mono{{-}{-}seed\_drop\_lim}. The
default is 0.3 bits. Larger numbers mean less filtering. (minor tuning option)
 
\item [\monob{{-}{-}seed\_req\_pos}\monoi{ $<$n$>$} ] A seed must contain a run of at least  \monoi{$<$n$>$} positive-scoring
matches. The default is 5. Larger values mean increased filtering. (minor
tuning option)  
\item [\monob{{-}{-}seed\_ssv\_length}\monoi{ $<$n$>$} ] After finding a short seed, an ungapped
alignment is extended  in both directions in an attempt to meet the  \mono{{-}{-}F1}
score threshold. The window through which this ungapped alignment extends
is length  \monoi{$<$n$>$}. The default is 70.   Decreasing this value slightly reduces
run time, at a small risk of reduced sensitivity. (minor tuning option)
  
\end{wideitem}

\subsection*{Other Options}
  \begin{wideitem}
\item [\monob{{-}{-}qformat}\monoi{ $<$s$>$} ] Assert that input \monoi{queryfile} is a sequence file
(unaligned or aligned), in format \monoi{$<$s$>$}, bypassing format autodetection. Common
choices for  \monoi{$<$s$>$}  include: \mono{fasta}, \mono{embl}, \mono{genbank.} Alignment formats also work,
and will serve as the basis for automatic creation of a profile HMM used
for searching; common choices include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal},
\mono{phylip}. For more information, and for codes for some less common formats,
see main documentation.   
\item [\monob{{-}{-}qsingle\_seqs} ] Force \monoi{queryfile} to be read as individual
sequences, even if it is in an msa format. For example, if the input is
in aligned \mono{stockholm} format, the \mono{{-}{-}qsingle\_seqs}  flag will cause each sequence
in that alignment to be used as a separate query sequence.

  
\item [\monob{{-}{-}tformat}\monoi{ $<$s$>$} ] Assert that target sequence database \monoi{seqdb} is in format \monoi{$<$s$>$},
bypassing format autodetection. Common choices for  \monoi{$<$s$>$}  include: \mono{fasta}, \mono{embl},
\mono{genbank}, \mono{fmindex}. Alignment formats also work; common choices include: \mono{stockholm},\mono{}
\mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal}, \mono{phylip}. For more information, and for codes
for some less common formats, see main documentation. The string \monoi{$<$s$>$} is case-insensitive
(\mono{fasta} or \mono{FASTA} both work). The format \mono{fmindex} indicates that the database
file is a binary file produced using \mono{makehmmerdb}.\mono{}   
\item [\monob{{-}{-}nonull2} ] Turn off the
null2 score corrections for biased composition.  
\item [\monob{-Z}\monoi{ $<$x$>$} ] For the purposes of
per-hit E-value calculations, Assert that the total size of the target database
is \monoi{$<$x$>$} million nucleotides,  rather than the actual number of targets seen.
   
\item [\monob{{-}{-}seed}\monoi{ $<$n$>$} ] Set the random number seed to  \monoi{$<$n$>$}. Some steps in postprocessing
require Monte Carlo simulation.  The default is to use a fixed seed (42),
so that results are exactly reproducible. Any other positive integer will
give different (but also reproducible) results. A choice of 0 uses a randomly
chosen seed.   
\item [\monob{{-}{-}w\_beta}\monoi{ $<$x$>$} ] Window length tail mass. The upper bound, \monoi{W}, on the
length at which nhmmer expects to find an instance of the  model is set
such that the fraction of all sequences generated by the model with length
$>$= W is less than   \monoi{$<$x$>$}.\monoi{} The default is 1e-7.  This flag may be used to override
the value of  \monoi{W} established for the model by  \mono{hmmbuild}, or when the query
is sequence-based.    
\item [\monob{{-}{-}w\_length}\monoi{ $<$n$>$} ] Override the model instance length upper
bound, W, which is otherwise controlled by \mono{{-}{-}w\_beta}.\mono{} It should be larger than
the model length. The value of W is used deep in the acceleration pipeline,
and modest changes are not expected to impact results (though larger values
of W do lead to longer run time).  This flag may be used to override the
value of  W established for the model by  \mono{hmmbuild}, or when the query is
sequence-based.    
\item [\monob{{-}{-}watson } ] Only search the top strand. By default both the
query sequence and its reverse-complement are searched.  
\item [\monob{{-}{-}crick } ] Only search
the bottom (reverse-complement) strand. By  default both the query sequence
and its reverse-complement are searched.   
\item [\monob{{-}{-}cpu}\monoi{ $<$n$>$} ] Set the number of parallel
worker threads to  \monoi{$<$n$>$}. On multicore machines, the default is 2. You can also
control this number by setting an environment variable,  \monoi{HMMER\_NCPU}. There
is also a master thread, so the actual number of threads that HMMER spawns
is \monoi{$<$n$>$}+1.  This option is not available if HMMER was compiled with POSIX threads
support turned off.     
\item [\monob{{-}{-}stall} ] For debugging the MPI master/worker version:
pause after start, to enable the developer to attach debuggers to the running
master and worker(s) processes. Send SIGCONT signal to release the pause.
(Under gdb:  \mono{(gdb) signal SIGCONT}) (Only available if optional MPI support
was enabled at compile-time.)  
\item [\monob{{-}{-}mpi} ] Run under MPI control with master/worker
parallelization (using \mono{mpirun}, for example, or equivalent). Only available
if optional MPI support was enabled at compile-time.       
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{nhmmscan}}{nhmmscan} - search DNA sequence(s) against a DNA profile database  }

\subsection*{Synopsis}
\noindent
\monob{nhmmscan} [\monoi{options}] \monoi{hmmdb} \monoi{seqfile}    
\subsection*{Description}
 

\mono{nhmmscan } is used
to search nucleotide sequences against collections  of nucleotide profiles.
For each sequence in  \monoi{seqfile}, use that query sequence to search the target
database of profiles in \monoi{hmmdb}, and output ranked lists of the profiles
with the most significant matches to the sequence.  

The  \monoi{seqfile}  may contain
more than one query sequence. It can be in FASTA format, or several other
common sequence file formats (genbank, embl, and uniprot, among others),
or in alignment file formats (stockholm, aligned fasta, and others). See
the \monoi{{-}{-}qformat}  option for a complete list.  

The \monoi{hmmdb} needs to be press'ed
using  \mono{hmmpress} before it can be searched with  \mono{nhmmscan}.\mono{} This creates four
binary files, suffixed \mono{.h3\{fimp\}.}  

The query \monoi{seqfile}  may be '-' (a dash character),
in which case the query sequences are read from a stdin pipe instead of
from a file. The \monoi{hmmdb}  cannot be read from a stdin stream, because it needs
to have the four auxiliary binary files generated by  \mono{hmmpress}.  

The output
format is designed to be human-readable, but is often so voluminous that
reading it is impractical, and parsing it is a pain. The \mono{{-}{-}tblout } option
saves output in a simple tabular format that is concise and easier to parse.
 The  \mono{-o} option allows redirecting the main output, including throwing it
away in /dev/null.    
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help; print a brief reminder of command
line usage and all available options.    
\end{wideitem}

\subsection*{Options for Controlling Output}

\begin{wideitem}
\item [\monob{-o}\monoi{ $<$f$>$} ] Direct the main human-readable output to a file \monoi{$<$f$>$}  instead of the default
stdout.  
\item [\monob{{-}{-}tblout}\monoi{ $<$f$>$} ] Save a simple tabular (space-delimited) file summarizing
the per-hit output, with one data line per homologous target model  hit
found.  
\item [\monob{{-}{-}dfamtblout}\monoi{ $<$f$>$} ] Save a tabular (space-delimited) file summarizing the
 per-hit output, similar to  \mono{{-}{-}tblout} but more succinct.   
\item [\monob{{-}{-}aliscoresout}\monoi{ $<$f$>$}\mono{} ] Save
to file a list of per-position scores for each hit. This is useful, for example,
in identifying regions of high score density for use in resolving overlapping
hits from  different models.   
\item [\monob{{-}{-}acc} ] Use accessions instead of names in the
main output, where available for profiles and/or sequences.  
\item [\monob{{-}{-}noali} ] Omit
the alignment section from the main output. This can greatly reduce the
output volume.  
\item [\monob{{-}{-}notextw} ] Unlimit the length of each line in the main output.
The default is a limit of 120 characters per line, which helps in displaying
the output cleanly on terminals and in editors, but can truncate target
profile description lines.  
\item [\monob{{-}{-}textw}\monoi{ $<$n$>$} ] Set the main output's line length limit
to \monoi{$<$n$>$} characters per line. The default is 120.    
\end{wideitem}

\subsection*{Options for Reporting Thresholds}

Reporting thresholds control which hits are reported in output files (the
main output, \mono{{-}{-}tblout}, and  \mono{{-}{-}dfamtblout}). Hits are ranked by statistical significance
(E-value).   \begin{wideitem}
\item [\monob{-E}\monoi{ $<$x$>$} ] Report target profiles with an E-value of $<$= \monoi{$<$x$>$}.\monoi{} The default
is 10.0, meaning that on average, about 10 false positives will be reported
per query, so you can see the top of the noise and decide for yourself
if it's really noise.  
\item [\monob{-T}\monoi{ $<$x$>$} ] Instead of thresholding output on E-value, instead
report target profiles with a bit score of $>$= \monoi{$<$x$>$}.     
\end{wideitem}

\subsection*{Options for Inclusion
Thresholds}
 Inclusion thresholds are stricter than reporting thresholds.
Inclusion thresholds control which hits are considered to be reliable enough
to be included in an output alignment or a subsequent search round. In 
\mono{nhmmscan},\mono{} which does not have any alignment output (like  \mono{nhmmer}), inclusion
thresholds have little effect. They only affect what hits get marked as
significant (!) or questionable (?) in hit output.   \begin{wideitem}
\item [\monob{{-}{-}incE}\monoi{ $<$x$>$} ] Use an E-value
of $<$= \monoi{$<$x$>$} as the inclusion threshold. The default is 0.01, meaning that on average,
about 1 false positive would be expected in every 100 searches with different
query sequences.  
\item [\monob{{-}{-}incT}\monoi{ $<$x$>$} ] Instead of using E-values for setting the inclusion
threshold,  use a bit score of $>$=  \monoi{$<$x$>$} as the inclusion threshold. It would
be unusual to use bit score thresholds with \mono{hmmscan}, because you don't expect
a single score threshold to work for different profiles; different profiles
have slightly different expected score distributions.    
\end{wideitem}

\subsection*{Options for Model-specific
Score Thresholding}
 Curated profile databases may define specific bit score
thresholds for each profile, superseding any thresholding based on statistical
significance alone.  To use these options, the profile must contain the
appropriate (GA, TC, and/or NC) optional score threshold annotation; this
is picked up by  \mono{hmmbuild} from Stockholm format alignment files. For a nucleotide
model, each  thresholding option has a single per-hit threshold $<$x$>$ This acts
as if \mono{-T}\monoi{ $<$x$>$} \mono{{-}{-}incT}\monoi{ $<$x$>$} has been applied specifically using each model's curated
thresholds.  \begin{wideitem}
\item [\monob{{-}{-}cut\_ga} ] Use the GA (gathering) bit score threshold in the model
to set per-hit reporting and inclusion thresholds. GA thresholds are generally
considered to be the reliable curated thresholds defining family membership;
for example, in Dfam, these thresholds are applied when annotating a genome
with a model of a family known to be found in that organism. They may allow
for minimal expected false discovery rate.  
\item [\monob{{-}{-}cut\_nc} ] Use the NC (noise cutoff)
bit score threshold in the model to set per-hit reporting and inclusion
thresholds. NC thresholds are less stringent than GA; in the context of
Pfam, they are generally used to store the score of the  highest-scoring
known false positive.  
\item [\monob{{-}{-}cut\_tc} ] Use the TC (trusted cutoff) bit score threshold
in the model to set per-hit reporting and inclusion thresholds. TC thresholds
are more stringent than GA, and are generally considered to be the score
of the lowest-scoring known  true positive that is above all known false
positives; for example, in Dfam, these thresholds are applied when annotating
a genome with a model of a family not known to be found in that organism.
   
\end{wideitem}

\subsection*{Control of the Acceleration Pipeline}
 HMMER3 searches are accelerated
in a three-step filter pipeline: the scanning-SSV filter, the Viterbi filter,
and the Forward filter. The  first filter is the fastest and most approximate;
the last is the full Forward scoring algorithm. There is also a bias filter
step between SSV and Viterbi. Targets that pass all the steps in the acceleration
pipeline are then subjected to postprocessing {-}{-} domain identification and
scoring using the Forward/Backward algorithm.  Changing filter thresholds
only removes or includes targets from consideration; changing filter thresholds
does not alter bit scores, E-values, or alignments, all of which are determined
solely in postprocessing.  \begin{wideitem}
\item [\monob{{-}{-}max} ] Turn off (nearly) all filters, including
the bias filter, and run full Forward/Backward postprocessing on most of
the target sequence. In contrast to   \mono{hmmscan,} where this flag really does
turn off the filters entirely, the  \mono{{-}{-}max} flag in  \mono{nhmmscan} sets the scanning-SSV
filter threshold to 0.4, not 1.0. Use of this flag increases sensitivity somewhat,
at a large cost in speed.  
\item [\monob{{-}{-}F1}\monoi{ $<$x$>$} ] Set the P-value threshold for the MSV filter
step.  The default is 0.02, meaning that roughly 2\% of the highest scoring
nonhomologous targets are expected to pass the filter.  
\item [\monob{{-}{-}F2}\monoi{ $<$x$>$} ] Set the P-value
threshold for the Viterbi filter step. The default is 0.001.   
\item [\monob{{-}{-}F3}\monoi{ $<$x$>$} ] Set the
P-value threshold for the Forward filter step. The default is 1e-5.  
\item [\monob{{-}{-}nobias}
] Turn off the bias filter. This increases sensitivity somewhat, but can come
at a high cost in speed, especially if the query has biased residue composition
(such as a repetitive sequence region, or if it is a membrane protein with
large regions of hydrophobicity). Without the bias filter, too many sequences
may pass the filter with biased queries, leading to slower than expected
performance as the computationally intensive Forward/Backward algorithms
shoulder an abnormally heavy load.    
\end{wideitem}

\subsection*{Other Options}
 \begin{wideitem}
\item [\monob{{-}{-}nonull2} ] Turn off the
null2 score corrections for biased composition.  
\item [\monob{-Z}\monoi{ $<$x$>$} ] Assert that the total
number of targets in your searches is \monoi{$<$x$>$}, for the purposes of per-sequence
E-value calculations, rather than the actual number of targets seen.   
\item [\monob{{-}{-}seed}\monoi{
$<$n$>$} ] Set the random number seed to  \monoi{$<$n$>$}. Some steps in postprocessing require
Monte Carlo simulation.  The default is to use a fixed seed (42), so that
results are exactly reproducible. Any other positive integer will give different
(but also reproducible) results. A choice of 0 uses an arbitrarily chosen
seed.  
\item [\monob{{-}{-}qformat}\monoi{ $<$s$>$} ] Assert that input query \monoi{seqfile} is in format \monoi{$<$s$>$}, bypassing
format autodetection. Common choices for  \monoi{$<$s$>$}  include: \mono{fasta}, \mono{embl}, \mono{genbank.}
Alignment formats also work; common choices include: \mono{stockholm},\mono{} \mono{a2m}, \mono{afa},
\mono{psiblast}, \mono{clustal}, \mono{phylip}. For more information, and for codes for some
less common formats, see main documentation. The string \monoi{$<$s$>$} is case-insensitive
(\mono{fasta} or \mono{FASTA} both work).   
\item [\monob{{-}{-}w\_beta}\monoi{ $<$x$>$} ] Window length tail mass. The upper
bound, W,  on the length at which nhmmer expects to find an instance of
the  model is set such that the fraction of all sequences generated by
the model with length $>$= W is less than   \monoi{$<$x$>$}.\monoi{} The default is 1e-7.  This flag
may be used to override the value of W established for the model by  \mono{hmmbuild}.
   
\item [\monob{{-}{-}w\_length}\monoi{ $<$n$>$} ] Override the model instance length upper bound, W, which
is otherwise controlled by \mono{{-}{-}w\_beta}.\mono{} It should be larger than the model length.
The value of  W is used deep in the acceleration pipeline, and modest changes
are not expected to impact results (though larger values of W do lead to
longer run time).  This flag may be used to override the value of W established
for the model by  \mono{hmmbuild}.   
\item [\monob{{-}{-}watson } ] Only search the top strand. By default
both the query sequence and its reverse-complement are searched.  
\item [\monob{{-}{-}crick }
] Only search the bottom (reverse-complement) strand. By  default both the
query sequence and its reverse-complement are searched.   
\item [\monob{{-}{-}cpu}\monoi{ $<$n$>$} ] Set the number
of parallel worker threads to  \monoi{$<$n$>$}. On multicore machines, the default is
2. You can also control this number by setting an environment variable,
 \monoi{HMMER\_NCPU}. There is also a master thread, so the actual number of threads
that HMMER spawns is \monoi{$<$n$>$}+1.  This option is not available if HMMER was compiled
with POSIX threads support turned off.     
\item [\monob{{-}{-}stall} ] For debugging the MPI master/worker
version: pause after start, to enable the developer to attach debuggers
to the running master and worker(s) processes. Send SIGCONT signal to release
the pause. (Under gdb:  \mono{(gdb) signal SIGCONT})  (Only available if optional
MPI support was enabled at compile-time.)  
\item [\monob{{-}{-}mpi} ] Run under MPI control with
master/worker parallelization (using \mono{mpirun}, for example, or equivalent).
Only available if optional MPI support was enabled at compile-time.     
    
\end{wideitem}

\newpage
% manual page source format generated by PolyglotMan v3.2,
% available at http://polyglotman.sourceforge.net/

\def\thefootnote{\fnsymbol{footnote}}
 
\section{\texorpdfstring{\monob{phmmer}}{phmmer} - search protein sequence(s) against a protein sequence database}
  
\subsection*{Synopsis}
\noindent
\monob{phmmer} [\monoi{options}] \monoi{seqfile} \monoi{seqdb}   
\subsection*{Description}
 

\mono{phmmer} is used to
search one or more query protein sequences against a protein sequence database.
For each query sequence in  \monoi{seqfile}, use that sequence to search the target
database of sequences in \monoi{seqdb}, and output ranked lists of the sequences
with the most significant matches to the query.  

Either the query \monoi{seqfile}
or the target \monoi{seqdb}  may be '-' (a dash character), in which case the query
sequences or target database input will be read from a $<$stdin$>$ pipe instead
of from a file. Only one input source can come through $<$stdin$>$, not both. An
exception is that if the \monoi{seqfile}  contains more than one query sequence,
then \monoi{seqdb}  cannot come from $<$stdin$>$, because we can't rewind the streaming
target database to search it with another query.   

The output format is
designed to be human-readable, but is often so voluminous that reading it
is impractical, and parsing it is a pain. The \mono{{-}{-}tblout } and  \mono{{-}{-}domtblout } options
save output in simple tabular formats that are concise and easier to parse.
The  \mono{-o} option allows redirecting the main output, including throwing it
away in /dev/null.  
\subsection*{Options}
 \begin{wideitem}
\item [\monob{-h} ] Help; print a brief reminder of command line
usage and all available options.   
\end{wideitem}

\subsection*{Options for Controlling Output}
 \begin{wideitem}
\item [\monob{-o}\monoi{ $<$f$>$} ] Direct
the main human-readable output to a file \monoi{$<$f$>$}  instead of the default stdout.
 
\item [\monob{-A}\monoi{ $<$f$>$} ] Save a multiple alignment of all significant hits (those satisfying
inclusion thresholds) to the file  \monoi{$<$f$>$} in Stockholm format.  
\item [\monob{{-}{-}tblout}\monoi{ $<$f$>$} ] Save
a simple tabular (space-delimited) file summarizing the per-target output,
with one data line per homologous target sequence found.  
\item [\monob{{-}{-}domtblout}\monoi{ $<$f$>$} ] Save
a simple tabular (space-delimited) file summarizing the per-domain output,
with one data line per homologous domain detected in a query sequence for
each homologous model.  
\item [\monob{{-}{-}acc} ] Use accessions instead of names in the main
output, where available for profiles and/or sequences.  
\item [\monob{{-}{-}noali} ] Omit the alignment
section from the main output. This can greatly reduce the output volume.
 
\item [\monob{{-}{-}notextw} ] Unlimit the length of each line in the main output. The default
is a limit of 120 characters per line, which helps in displaying the output
cleanly on terminals and in editors, but can truncate target profile description
lines.  
\item [\monob{{-}{-}textw}\monoi{ $<$n$>$} ] Set the main output's line length limit to \monoi{$<$n$>$} characters per
line. The default is 120.    
\end{wideitem}

\subsection*{Options Controlling Scoring System}
 The probability
model in \mono{phmmer} is constructed by inferring residue probabilities from
a standard 20x20 substitution score matrix, plus two additional parameters
for position-independent gap open and gap extend probabilities.  \begin{wideitem}
\item [\monob{{-}{-}popen}\monoi{ $<$x$>$}
] Set the gap open probability for a single sequence query model to  \monoi{$<$x$>$}. The
default is 0.02.  \monoi{$<$x$>$}  must be $>$= 0 and $<$ 0.5.  
\item [\monob{{-}{-}pextend}\monoi{ $<$x$>$} ] Set the gap extend probability
for a single sequence query model to  \monoi{$<$x$>$}. The default is 0.4.  \monoi{$<$x$>$}  must be $>$=
0 and $<$ 1.0.  
\item [\monob{{-}{-}mx}\monoi{ $<$s$>$} ] Obtain residue alignment probabilities from the built-in
substitution matrix named \monoi{$<$s$>$}.\monoi{} Several standard matrices are built-in, and
do not need to be read from files.  The matrix name \monoi{$<$s$>$}  can be PAM30, PAM70,
PAM120, PAM240, BLOSUM45, BLOSUM50, BLOSUM62, BLOSUM80, or BLOSUM90. Only
one of the \mono{{-}{-}mx } and \mono{{-}{-}mxfile} options may be used.  
\item [\monob{{-}{-}mxfile}\monoi{ mxfile} ] Obtain residue
alignment probabilities from the substitution matrix in file \monoi{mxfile}. The
default score matrix is BLOSUM62 (this matrix is internal to HMMER and
does not have to be available as a file).  The format of a substitution
matrix \monoi{mxfile} is the standard format accepted by BLAST, FASTA, and other
sequence  analysis software. See ftp.ncbi.nlm.nih.gov/blast/matrices/ for example
files. (The only exception: we require matrices to be square, so for DNA,
use files like NCBI's NUC.4.4, not NUC.4.2.)    
\end{wideitem}

\subsection*{Options Controlling Reporting
Thresholds}
 Reporting thresholds control which hits are reported in output
files (the main output, \mono{{-}{-}tblout}, and  \mono{{-}{-}domtblout}). Sequence hits and domain
hits are ranked by statistical significance (E-value) and output is generated
in two sections called per-target and per-domain output. In per-target output,
by default, all sequence hits with an E-value $<$= 10 are reported. In the per-domain
output, for each target that has passed per-target reporting thresholds,
all domains satisfying per-domain reporting thresholds are reported. By default,
these are domains with conditional E-values of $<$= 10. The following options
allow you to change the default E-value reporting thresholds, or to use
bit score thresholds instead.   \begin{wideitem}
\item [\monob{-E}\monoi{ $<$x$>$} ] In the per-target output, report target
sequences with an E-value of $<$= \monoi{$<$x$>$}.\monoi{} The default is 10.0, meaning that on average,
about 10 false positives will be reported per query, so you can see the
top of the noise and decide for yourself if it's really noise.  
\item [\monob{-T}\monoi{ $<$x$>$} ] Instead
of thresholding per-profile output on E-value, instead report target sequences
with a bit score of $>$= \monoi{$<$x$>$}.  
\item [\monob{{-}{-}domE}\monoi{ $<$x$>$} ] In the per-domain output, for target sequences
that have already satisfied the per-profile reporting threshold, report
individual domains with a conditional E-value of $<$= \monoi{$<$x$>$}.\monoi{} The default is 10.0.
 A conditional E-value means the expected number of additional false positive
domains in the smaller search space of those comparisons that already satisfied
the per-target reporting threshold (and thus must have at least one homologous
domain already).  
\item [\monob{{-}{-}domT}\monoi{ $<$x$>$} ] Instead of thresholding per-domain output on E-value,
instead report domains with a bit score of $>$= \monoi{$<$x$>$}.  
\end{wideitem}

\subsection*{Options Controlling Inclusion
Thresholds}
 Inclusion thresholds are stricter than reporting thresholds.
They control which hits are included in any output multiple alignment (the
\mono{-A } option) and which domains are marked as significant ("!") as opposed
to questionable ("?")  in domain output.  \begin{wideitem}
\item [\monob{{-}{-}incE}\monoi{ $<$x$>$} ] Use an E-value of $<$= \monoi{$<$x$>$} as
the per-target inclusion threshold. The default is 0.01, meaning that on average,
about 1 false positive would be expected in every 100 searches with different
query sequences.  
\item [\monob{{-}{-}incT}\monoi{ $<$x$>$} ] Instead of using E-values for setting the inclusion
threshold, instead use a bit score of $>$=  \monoi{$<$x$>$} as the per-target inclusion threshold.
By default this option is unset.  
\item [\monob{{-}{-}incdomE}\monoi{ $<$x$>$} ] Use a conditional E-value of
$<$= \monoi{$<$x$>$}  as the per-domain inclusion threshold, in targets that have already
satisfied the overall per-target inclusion threshold. The default is 0.01.
 
\item [\monob{{-}{-}incdomT}\monoi{ $<$x$>$} ] Instead of using E-values, use a bit score of $>$= \monoi{$<$x$>$} as the per-domain
inclusion threshold. By default this option is unset.     
\end{wideitem}

\subsection*{Options Controlling
the Acceleration Pipeline}
 HMMER3 searches are accelerated in a three-step
filter pipeline: the MSV filter, the Viterbi filter, and the Forward filter.
The first filter is the fastest and most approximate; the last is the full
Forward scoring algorithm, slowest but most accurate. There is also a bias
filter step between MSV and Viterbi. Targets that pass all the steps in
the acceleration pipeline are then subjected to postprocessing {-}{-} domain
identification and scoring using the Forward/Backward algorithm.  Essentially
the only free parameters that control HMMER's heuristic filters are the
P-value thresholds controlling the expected fraction of nonhomologous sequences
that pass the filters. Setting the default thresholds higher will pass a
higher proportion of nonhomologous sequence, increasing sensitivity at
the expense of speed; conversely, setting lower P-value thresholds will
pass a smaller proportion, decreasing sensitivity and increasing speed.
Setting a filter's P-value threshold to 1.0 means it will passing all sequences,
and effectively disables the filter.  Changing filter thresholds only removes
or includes targets from consideration; changing filter thresholds does
not alter bit scores, E-values, or alignments, all of which are determined
solely in postprocessing.  \begin{wideitem}
\item [\monob{{-}{-}max} ] Maximum sensitivity.  Turn off all filters,
including the bias filter, and run full Forward/Backward postprocessing
on every target. This increases sensitivity slightly, at a large cost in
speed.  
\item [\monob{{-}{-}F1}\monoi{ $<$x$>$} ] First filter threshold; set the P-value threshold for the MSV
filter step.  The default is 0.02, meaning that roughly 2\% of the highest
scoring nonhomologous targets are expected to pass the filter.  
\item [\monob{{-}{-}F2}\monoi{ $<$x$>$} ] Second
filter threshold; set the P-value threshold for the Viterbi filter step.
 The default is 0.001.  
\item [\monob{{-}{-}F3}\monoi{ $<$x$>$} ] Third filter threshold; set the P-value threshold
for the Forward filter step.  The default is 1e-5.  
\item [\monob{{-}{-}nobias} ] Turn off the bias
filter. This increases sensitivity somewhat, but can come at a high cost
in speed, especially if the query has biased residue composition (such
as a repetitive sequence region, or if it is a membrane protein with large
regions of hydrophobicity). Without the bias filter, too many sequences
may pass the filter with biased queries, leading to slower than expected
performance as the computationally intensive Forward/Backward algorithms
shoulder an abnormally heavy load.     
\end{wideitem}

\subsection*{Options Controlling E-value Calibration}

Estimating the location parameters for the expected score distributions
for MSV filter scores, Viterbi filter scores, and Forward scores requires
three short random sequence simulations.  \begin{wideitem}
\item [\monob{{-}{-}EmL}\monoi{ $<$n$>$} ] Sets the sequence length
in simulation that estimates the location parameter mu for MSV filter E-values.
Default is 200.  
\item [\monob{{-}{-}EmN}\monoi{ $<$n$>$} ] Sets the number of sequences in simulation that estimates
the location parameter mu for MSV filter E-values. Default is 200.  
\item [\monob{{-}{-}EvL}\monoi{ $<$n$>$}
] Sets the sequence length in simulation that estimates the location parameter
mu for Viterbi filter E-values. Default is 200.  
\item [\monob{{-}{-}EvN}\monoi{ $<$n$>$} ] Sets the number of
sequences in simulation that estimates the location parameter mu for Viterbi
filter E-values. Default is 200.  
\item [\monob{{-}{-}EfL}\monoi{ $<$n$>$} ] Sets the sequence length in simulation
that estimates the location parameter tau for Forward E-values. Default is
100.  
\item [\monob{{-}{-}EfN}\monoi{ $<$n$>$} ] Sets the number of sequences in simulation that estimates the
location parameter tau for Forward E-values. Default is 200.  
\item [\monob{{-}{-}Eft}\monoi{ $<$x$>$} ] Sets the
tail mass fraction to fit in the simulation that estimates the location
parameter tau for Forward evalues. Default is 0.04.     
\end{wideitem}

\subsection*{Other Options}
 \begin{wideitem}
\item [\monob{{-}{-}nonull2}
] Turn off the null2 score corrections for biased composition.  
\item [\monob{-Z}\monoi{ $<$x$>$} ] Assert
that the total number of targets in your searches is \monoi{$<$x$>$}, for the purposes
of per-sequence E-value calculations, rather than the actual number of targets
seen.   
\item [\monob{{-}{-}domZ}\monoi{ $<$x$>$} ] Assert that the total number of targets in your searches
is \monoi{$<$x$>$}, for the purposes of per-domain conditional E-value calculations, rather
than the number of targets that passed the reporting thresholds.  
\item [\monob{{-}{-}seed}\monoi{ $<$n$>$}
] Seed the random number generator with \monoi{$<$n$>$}, an integer $>$= 0.  If  \monoi{$<$n$>$}  is $>$0, any
stochastic simulations will be reproducible; the same command will give
the same results. If  \monoi{$<$n$>$} is 0, the random number generator is seeded arbitrarily,
and stochastic simulations will vary from run to run of the same command.
The default seed is 42.  
\item [\monob{{-}{-}qformat}\monoi{ $<$s$>$} ] Assert that input \monoi{seqfile} is in format
\monoi{$<$s$>$}, bypassing format autodetection. Common choices for  \monoi{$<$s$>$}  include: \mono{fasta},
\mono{embl}, \mono{genbank.} Alignment formats also work; common choices include: \mono{stockholm},\mono{}
\mono{a2m}, \mono{afa}, \mono{psiblast}, \mono{clustal}, \mono{phylip}. \mono{phmmer} always uses a single sequence
query to start its search, so when the input \monoi{seqfile} is an alignment, \mono{phmmer}
reads it one unaligned query sequence at a time, not as an alignment. For
more information, and for codes for some less common formats, see main
documentation. The string \monoi{$<$s$>$} is case-insensitive (\mono{fasta} or \mono{FASTA} both work).
 \mono{{-}{-}tformat}\monoi{ $<$s$>$} Assert that target sequence database \monoi{seqdb} is in format \monoi{$<$s$>$}, bypassing
format autodetection. See \mono{{-}{-}qformat} above for list of accepted format codes
for \monoi{$<$s$>$}.   
\item [\monob{{-}{-}cpu}\monoi{ $<$n$>$} ] Set the number of parallel worker threads to  \monoi{$<$n$>$}. On multicore
machines, the default is 2. You can also control this number by setting
an environment variable,  \monoi{HMMER\_NCPU}. There is also a master thread, so
the actual number of threads that HMMER spawns is \monoi{$<$n$>$}+1.  This option is not
available if HMMER was compiled with POSIX threads support turned off. 
  
\item [\monob{{-}{-}stall} ] For debugging the MPI master/worker version: pause after start,
to enable the developer to attach debuggers to the running master and worker(s)
processes. Send SIGCONT signal to release the pause. (Under gdb:  \mono{(gdb) signal
SIGCONT}) (Only available if optional MPI support was enabled at compile-time.)
 
\item [\monob{{-}{-}mpi} ] Run under MPI control with master/worker parallelization (using \mono{mpirun},
for example, or equivalent). Only available if optional MPI support was
enabled at compile-time.      
\end{wideitem}

\newpage
